// @jetbrains/ring-ui-built/components/tabs/collapsible-tab.js
import { c } from 'react-compiler-runtime';
import * as React from 'react';
import classNames from 'classnames';
import { s as styles, T as TabLink } from '../_helpers/tab-link.js';
import { CustomItem } from './custom-item.js';
import { jsx } from 'react/jsx-runtime';
import '../link/link.js';
import '../global/data-tests.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';

const TabTitle = /*#__PURE__*/React.memo(function TabTitle(t0) {
  const $ = c(18);
  if ($[0] !== "bc923b30ae31abf97ca3494d9773d25ec50a98e0a9dced298aaa6afe7b737db1") {
    for (let $i = 0; $i < 18; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "bc923b30ae31abf97ca3494d9773d25ec50a98e0a9dced298aaa6afe7b737db1";
  }
  const {
    selected,
    child,
    handleSelect,
    collapsed: t1,
    tabIndex
  } = t0;
  const collapsed = t1 === undefined ? false : t1;
  if (child === null || typeof child !== "object" || child.type === CustomItem) {
    return child;
  }
  const {
    title,
    titleProps,
    disabled,
    href,
    className,
    activeClassName,
    collapsedClassName,
    collapsedActiveClassName
  } = child.props;
  const t2 = selected && activeClassName;
  const t3 = collapsed && collapsedClassName;
  const t4 = collapsed && selected && collapsedActiveClassName;
  let t5;
  if ($[1] !== className || $[2] !== collapsed || $[3] !== selected || $[4] !== t2 || $[5] !== t3 || $[6] !== t4) {
    t5 = classNames(styles.title, className, t2, t3, t4, {
      [styles.selected]: selected,
      [styles.collapsed]: collapsed
    });
    $[1] = className;
    $[2] = collapsed;
    $[3] = selected;
    $[4] = t2;
    $[5] = t3;
    $[6] = t4;
    $[7] = t5;
  } else {
    t5 = $[7];
  }
  const titleClasses = t5;
  let t6;
  if ($[8] !== collapsed || $[9] !== disabled || $[10] !== handleSelect || $[11] !== href || $[12] !== selected || $[13] !== tabIndex || $[14] !== title || $[15] !== titleClasses || $[16] !== titleProps) {
    t6 = /*#__PURE__*/jsx(TabLink, {
      title: title,
      isSelected: selected,
      active: true,
      href: href,
      className: titleClasses,
      disabled: disabled,
      onPlainLeftClick: handleSelect,
      tabIndex: tabIndex,
      collapsed: collapsed,
      ...titleProps
    });
    $[8] = collapsed;
    $[9] = disabled;
    $[10] = handleSelect;
    $[11] = href;
    $[12] = selected;
    $[13] = tabIndex;
    $[14] = title;
    $[15] = titleClasses;
    $[16] = titleProps;
    $[17] = t6;
  } else {
    t6 = $[17];
  }
  return t6;
});
const getTabTitles = ({
  items,
  selected = '0',
  collapsed,
  onSelect = () => undefined,
  ...props
}) => items.map((tab, index) => {
  const key = tab.props.id || String(index);
  const isSelected = selected === key;
  return /*#__PURE__*/jsx(TabTitle, {
    handleSelect: onSelect(key),
    selected: isSelected,
    child: tab,
    collapsed: collapsed,
    ...props
  }, key);
});

export { getTabTitles as default };
