// @jetbrains/ring-ui-built/components/tabs/collapsible-more.js
import { c } from 'react-compiler-runtime';
import { memo } from 'react';
import classNames from 'classnames';
import { Directions } from '../popup/popup.consts.js';
import PopupMenu, { ListProps } from '../popup-menu/popup-menu.js';
import Dropdown from '../dropdown/dropdown.js';
import { A as Anchor } from '../_helpers/anchor.js';
import getTabTitles from './collapsible-tab.js';
import { CustomItem } from './custom-item.js';
import { s as styles } from '../_helpers/tab-link.js';
import { jsx } from 'react/jsx-runtime';
import '../popup/popup.js';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/data-tests.js';
import '../tab-trap/tab-trap.js';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../list/list.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'util-deprecate';
import 'memoize-one';
import '../global/memoize.js';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../avatar/avatar.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../global/typescript-utils.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../global/controls-height.js';
import '../_helpers/button.classes.js';

const AnchorLink = t0 => {
  const $ = c(13);
  if ($[0] !== "75a0b2879bebaad3f8d82b99df93f8f96cbaefa77f4f888abdbc3d190e1a6bc8") {
    for (let $i = 0; $i < 13; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "75a0b2879bebaad3f8d82b99df93f8f96cbaefa77f4f888abdbc3d190e1a6bc8";
  }
  let hasActiveChildren;
  let moreActiveClassName;
  let moreClassName;
  let restProps;
  if ($[1] !== t0) {
    ({
      hasActiveChildren,
      moreClassName,
      moreActiveClassName,
      ...restProps
    } = t0);
    $[1] = t0;
    $[2] = hasActiveChildren;
    $[3] = moreActiveClassName;
    $[4] = moreClassName;
    $[5] = restProps;
  } else {
    hasActiveChildren = $[2];
    moreActiveClassName = $[3];
    moreClassName = $[4];
    restProps = $[5];
  }
  const t1 = hasActiveChildren && styles.selected;
  const t2 = hasActiveChildren && moreActiveClassName;
  let t3;
  if ($[6] !== moreClassName || $[7] !== t1 || $[8] !== t2) {
    t3 = classNames(styles.title, t1, t2, moreClassName);
    $[6] = moreClassName;
    $[7] = t1;
    $[8] = t2;
    $[9] = t3;
  } else {
    t3 = $[9];
  }
  const classnames = t3;
  let t4;
  if ($[10] !== classnames || $[11] !== restProps) {
    t4 = /*#__PURE__*/jsx(Anchor, {
      title: "More",
      className: classnames,
      ...restProps,
      children: "More"
    });
    $[10] = classnames;
    $[11] = restProps;
    $[12] = t4;
  } else {
    t4 = $[12];
  }
  return t4;
};
const morePopupDirections = [Directions.BOTTOM_CENTER, Directions.BOTTOM_LEFT, Directions.BOTTOM_RIGHT];
const MoreButton = /*#__PURE__*/memo(t0 => {
  const $ = c(32);
  if ($[0] !== "75a0b2879bebaad3f8d82b99df93f8f96cbaefa77f4f888abdbc3d190e1a6bc8") {
    for (let $i = 0; $i < 32; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "75a0b2879bebaad3f8d82b99df93f8f96cbaefa77f4f888abdbc3d190e1a6bc8";
  }
  const {
    items,
    selected,
    onSelect,
    moreClassName,
    moreActiveClassName,
    morePopupClassName,
    morePopupItemClassName,
    morePopupBeforeEnd
  } = t0;
  let t1;
  if ($[1] !== onSelect) {
    t1 = onSelect ? listItem => {
      if (listItem.disabled === true || listItem.custom === true) {
        return;
      }
      const cb = onSelect(String(listItem.key));
      cb();
    } : undefined;
    $[1] = onSelect;
    $[2] = t1;
  } else {
    t1 = $[2];
  }
  const onSelectHandler = t1;
  let t2;
  if ($[3] !== items || $[4] !== selected) {
    let t3;
    if ($[6] !== selected) {
      t3 = item => item.props.alwaysHidden && item.props.id === selected;
      $[6] = selected;
      $[7] = t3;
    } else {
      t3 = $[7];
    }
    t2 = items.some(t3);
    $[3] = items;
    $[4] = selected;
    $[5] = t2;
  } else {
    t2 = $[5];
  }
  const hasActiveChild = t2;
  let data;
  if ($[8] !== items || $[9] !== morePopupBeforeEnd || $[10] !== morePopupItemClassName || $[11] !== selected) {
    let t3;
    if ($[13] !== morePopupItemClassName) {
      t3 = tab => {
        const disabled = tab.props.disabled === true;
        const custom = tab.props.child.type === CustomItem;
        return {
          template: tab,
          key: tab.key,
          rgItemType: ListProps.Type.CUSTOM,
          className: morePopupItemClassName,
          disabled,
          custom
        };
      };
      $[13] = morePopupItemClassName;
      $[14] = t3;
    } else {
      t3 = $[14];
    }
    data = getTabTitles({
      items,
      selected,
      collapsed: true
    }).map(t3);
    if (morePopupBeforeEnd) {
      data.push({
        template: morePopupBeforeEnd,
        key: "before-end-content",
        className: styles.morePopupBeforeEnd,
        rgItemType: ListProps.Type.CUSTOM
      });
    }
    $[8] = items;
    $[9] = morePopupBeforeEnd;
    $[10] = morePopupItemClassName;
    $[11] = selected;
    $[12] = data;
  } else {
    data = $[12];
  }
  let t3;
  if ($[15] !== hasActiveChild || $[16] !== moreActiveClassName || $[17] !== moreClassName) {
    t3 = /*#__PURE__*/jsx(AnchorLink, {
      moreClassName: moreClassName,
      moreActiveClassName: moreActiveClassName,
      hasActiveChildren: hasActiveChild
    });
    $[15] = hasActiveChild;
    $[16] = moreActiveClassName;
    $[17] = moreClassName;
    $[18] = t3;
  } else {
    t3 = $[18];
  }
  const popupAnchor = t3;
  let t4;
  if ($[19] !== data || $[20] !== morePopupClassName || $[21] !== onSelectHandler) {
    t4 = /*#__PURE__*/jsx(PopupMenu, {
      directions: morePopupDirections,
      className: morePopupClassName,
      onSelect: onSelectHandler,
      data: data
    });
    $[19] = data;
    $[20] = morePopupClassName;
    $[21] = onSelectHandler;
    $[22] = t4;
  } else {
    t4 = $[22];
  }
  const popup = t4;
  if (items.length === 0) {
    return null;
  }
  const t5 = hasActiveChild && moreActiveClassName;
  let t6;
  if ($[23] !== moreClassName || $[24] !== t5) {
    t6 = classNames(styles.title, moreClassName, t5);
    $[23] = moreClassName;
    $[24] = t5;
    $[25] = t6;
  } else {
    t6 = $[25];
  }
  let t7;
  if ($[26] !== popup || $[27] !== popupAnchor) {
    t7 = /*#__PURE__*/jsx(Dropdown, {
      hoverMode: true,
      anchor: popupAnchor,
      children: popup
    });
    $[26] = popup;
    $[27] = popupAnchor;
    $[28] = t7;
  } else {
    t7 = $[28];
  }
  let t8;
  if ($[29] !== t6 || $[30] !== t7) {
    t8 = /*#__PURE__*/jsx("div", {
      className: t6,
      children: t7
    });
    $[29] = t6;
    $[30] = t7;
    $[31] = t8;
  } else {
    t8 = $[31];
  }
  return t8;
});
MoreButton.displayName = 'MoreButton';
const FakeMoreButton = /*#__PURE__*/memo(t0 => {
  const $ = c(6);
  if ($[0] !== "75a0b2879bebaad3f8d82b99df93f8f96cbaefa77f4f888abdbc3d190e1a6bc8") {
    for (let $i = 0; $i < 6; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "75a0b2879bebaad3f8d82b99df93f8f96cbaefa77f4f888abdbc3d190e1a6bc8";
  }
  const {
    moreClassName,
    moreActiveClassName,
    hasActiveChildren
  } = t0;
  let t1;
  if ($[1] === Symbol.for("react.memo_cache_sentinel")) {
    t1 = classNames(styles.moreButton, styles.title);
    $[1] = t1;
  } else {
    t1 = $[1];
  }
  let t2;
  if ($[2] !== hasActiveChildren || $[3] !== moreActiveClassName || $[4] !== moreClassName) {
    t2 = /*#__PURE__*/jsx("div", {
      className: t1,
      children: /*#__PURE__*/jsx(AnchorLink, {
        moreClassName: moreClassName,
        moreActiveClassName: moreActiveClassName,
        hasActiveChildren: hasActiveChildren,
        tabIndex: -1,
        disabled: true
      })
    });
    $[2] = hasActiveChildren;
    $[3] = moreActiveClassName;
    $[4] = moreClassName;
    $[5] = t2;
  } else {
    t2 = $[5];
  }
  return t2;
});
FakeMoreButton.displayName = 'FakeMoreButton';

export { AnchorLink, FakeMoreButton, MoreButton };
