// @jetbrains/ring-ui-built/components/tabs/dumb-tabs.js
import * as React from 'react';
import { PureComponent } from 'react';
import classNames from 'classnames';
import memoize from '../global/memoize.js';
import joinDataTestAttributes from '../global/data-tests.js';
import { s as styles, T as TabLink } from '../_helpers/tab-link.js';
import CollapsibleTabs from './collapsible-tabs.js';
import { CustomItem } from './custom-item.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'react-compiler-runtime';
import '../link/link.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import 'fastdom';
import './collapsible-more.js';
import '../popup/popup.consts.js';
import '../popup-menu/popup-menu.js';
import '../popup/popup.js';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../list/list.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'util-deprecate';
import 'memoize-one';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../avatar/avatar.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../dropdown/dropdown.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../global/controls-height.js';
import '../_helpers/button.classes.js';
import './collapsible-tab.js';

class Tabs extends PureComponent {
  handleSelect = memoize(key => () => this.props.onSelect?.(key));
  getTabTitle = (child, i) => {
    if (child === null || typeof child !== 'object' || child.type === CustomItem) {
      return child;
    }
    const {
      selected,
      onSelect
    } = this.props;
    const {
      title,
      titleProps,
      id,
      disabled,
      href,
      className,
      activeClassName
    } = child.props;
    const key = id || String(i);
    const isSelected = key === selected;
    const titleClasses = classNames(styles.title, className, isSelected && activeClassName, {
      [styles.selected]: isSelected
    });
    return /*#__PURE__*/jsx(TabLink, {
      title: title,
      isSelected: isSelected,
      href: href,
      className: titleClasses,
      disabled: disabled,
      onPlainLeftClick: onSelect ? this.handleSelect(key) : undefined,
      ...titleProps
    }, key);
  };
  render() {
    const {
      className,
      tabContainerClassName,
      children,
      selected,
      autoCollapse,
      'data-test': dataTest,
      onSelect,
      ...restProps
    } = this.props;
    const classes = classNames(styles.tabs, className);
    const childrenArray = React.Children.toArray(children).filter(Boolean);
    return /*#__PURE__*/jsxs("div", {
      className: classes,
      "data-test": joinDataTestAttributes('ring-dumb-tabs', dataTest),
      children: [autoCollapse === true ? /*#__PURE__*/jsx(CollapsibleTabs, {
        ...restProps,
        onSelect: onSelect ? this.handleSelect : undefined,
        selected: selected,
        children: childrenArray
      }) : /*#__PURE__*/jsx("div", {
        className: styles.titles,
        children: childrenArray.map(this.getTabTitle)
      }), /*#__PURE__*/jsx("div", {
        className: classNames(tabContainerClassName),
        children: childrenArray.find(({
          props
        }, i) => (props.id || String(i)) === selected)
      })]
    });
  }
}

export { CustomItem, Tabs as default };
