// @jetbrains/ring-ui-built/components/login-dialog/login-dialog.js
import { Component } from 'react';
import Dialog from '../dialog/dialog.js';
import 'classnames';
import { jsxs, jsx } from 'react/jsx-runtime';
import '../_helpers/island.js';
import 'react-compiler-runtime';
import ContentWrapper from '../island/content.js';
import LoaderScreen from '../loader-screen/loader-screen.js';
import { HUB_AUTH_PAGE_OPENED } from '../auth/background-flow.js';
import 'react-dom';
import '@jetbrains/icons/close';
import '../island/island.js';
import '../global/data-tests.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import '../island/header.js';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '../loader/loader.js';
import '../_helpers/loader-core.js';
import '../auth/response-parser.js';
import '../global/url.js';

var styles = {"iFrame":"ring-login-dialog-iFrame","dialogContent":"ring-login-dialog-dialogContent","nonOpaqueLoader":"ring-login-dialog-nonOpaqueLoader","fallbackLinkContainer":"ring-login-dialog-fallbackLinkContainer"};

const HUB_AUTH_PAGE_LOGIN_STARTED = 'HUB_AUTH_PAGE_LOGIN_STARTED';
const HUB_AUTH_PAGE_LOGIN_DIMENSIONS = 'HUB_AUTH_PAGE_LOGIN_DIMENSIONS';
const DEFAULT_HEIGHT = 517;
const DEFAULT_WIDTH = 333;
const DEFAULT_SHOW_FALLBACK_TIMEOUT = 5000;
/**
 * @name Login Dialog
 */
class LoginDialog extends Component {
  static defaultProps = {
    show: false,
    url: 'about:blank',
    renderFallbackLink: () => null,
    showFallbackTimeout: DEFAULT_SHOW_FALLBACK_TIMEOUT
  };
  state = {
    loading: true,
    loggingIn: false,
    showFallbackLink: false,
    height: DEFAULT_HEIGHT,
    width: DEFAULT_WIDTH
  };
  componentDidMount() {
    window.addEventListener('message', this.onMessage);
    this.startFallbackCountdown();
  }
  componentWillUnmount() {
    window.removeEventListener('message', this.onMessage);
  }
  showFallbackTimout;
  startFallbackCountdown() {
    this.showFallbackTimout = window.setTimeout(() => this.setState({
      showFallbackLink: true
    }), this.props.showFallbackTimeout);
  }
  onMessage = event => {
    const {
      data
    } = event;
    if (!data) {
      return;
    }
    if (data === HUB_AUTH_PAGE_OPENED) {
      clearTimeout(this.showFallbackTimout);
      this.setState({
        loading: false,
        loggingIn: false
      });
      return;
    }
    if (data === HUB_AUTH_PAGE_LOGIN_STARTED) {
      this.setState({
        loading: true,
        loggingIn: true
      });
      this.startFallbackCountdown();
      return;
    }
    if (data.message === HUB_AUTH_PAGE_LOGIN_DIMENSIONS) {
      this.setState({
        height: data.height,
        width: data.width
      });
    }
  };
  render() {
    const {
      show,
      className,
      url,
      loadingMessage,
      renderFallbackLink,
      onCancel
    } = this.props;
    const {
      loading,
      height,
      width,
      loggingIn,
      showFallbackLink
    } = this.state;
    const iFrameStyle = {
      height,
      width
    };
    return /*#__PURE__*/jsxs(Dialog, {
      "data-test": "ring-login-dialog",
      className: className,
      contentClassName: styles.dialogContent,
      trapFocus: true,
      autoFocusFirst: false,
      show: show,
      showCloseButton: true,
      onCloseAttempt: onCancel,
      children: [/*#__PURE__*/jsx(ContentWrapper, {
        children: /*#__PURE__*/jsx("iframe", {
          title: "Login dialog",
          style: iFrameStyle,
          src: url,
          className: styles.iFrame,
          scrolling: "no"
        })
      }), loading && /*#__PURE__*/jsx(LoaderScreen, {
        message: loadingMessage,
        containerClassName: styles.nonOpaqueLoader
      }), showFallbackLink && /*#__PURE__*/jsx("div", {
        className: styles.fallbackLinkContainer,
        children: renderFallbackLink(loggingIn)
      })]
    });
  }
}

export { LoginDialog as default };
