// @jetbrains/ring-ui-built/components/login-dialog/service.js
import { createRoot } from 'react-dom/client';
import { ControlsHeightContext, getGlobalControlsHeight } from '../global/controls-height.js';
import LoginDialog from './login-dialog.js';
import { jsx } from 'react/jsx-runtime';
import 'react';
import 'util-deprecate';
import '../global/configuration.js';
import '../dialog/dialog.js';
import 'react-dom';
import 'classnames';
import '@jetbrains/icons/close';
import '../island/island.js';
import '../global/data-tests.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import '../_helpers/island.js';
import '../island/header.js';
import 'react-compiler-runtime';
import '../island/content.js';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../_helpers/button.classes.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '../loader-screen/loader-screen.js';
import '../loader/loader.js';
import '../_helpers/loader-core.js';
import '../auth/background-flow.js';
import '../auth/response-parser.js';
import '../global/url.js';

const containerElement = document.createElement('div');
const reactRoot = createRoot(containerElement);
/**
 * Renders LoginDialog into virtual node to skip maintaining container
 */
function renderLoginDialog(props) {
  reactRoot.render(/*#__PURE__*/jsx(ControlsHeightContext.Provider, {
    value: getGlobalControlsHeight(),
    children: /*#__PURE__*/jsx(LoginDialog, {
      ...props
    })
  }));
}
function noop() {}
function showAuthDialog(props = {
  onCancel: noop
}) {
  renderLoginDialog({
    ...props,
    show: true
  });
  return () => {
    renderLoginDialog({
      ...props,
      show: false
    });
  };
}

export { showAuthDialog as default };
