// @jetbrains/ring-ui-built/components/dialog/dialog.js
import { PureComponent, createRef } from 'react';
import { createPortal } from 'react-dom';
import classNames from 'classnames';
import closeIcon from '@jetbrains/icons/close';
import { AdaptiveIsland } from '../island/island.js';
import getUID from '../global/get-uid.js';
import joinDataTestAttributes from '../global/data-tests.js';
import Shortcuts from '../shortcuts/shortcuts.js';
import TabTrap from '../tab-trap/tab-trap.js';
import { Button } from '../button/button.js';
import { PopupTarget, PopupTargetContext, normalizePopupTarget } from '../popup/popup.target.js';
import { getPopupContainer } from '../popup/popup.js';
import { p as preventerFactory, s as styles } from '../_helpers/dialog-body-scroll-preventer.js';
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import '../_helpers/island.js';
import '../island/header.js';
import 'react-compiler-runtime';
import '../island/content.js';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/dom.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import 'scrollbar-width';

function noop() {}
class Dialog extends PureComponent {
  static defaultProps = {
    label: 'Dialog',
    onOverlayClick: noop,
    onEscPress: noop,
    onCloseClick: noop,
    onCloseAttempt: noop,
    showCloseButton: false,
    closeButtonInside: false,
    shortcutOptions: {
      modal: false
    },
    trapFocus: false,
    autoFocusFirst: true,
    modal: true,
    preventBodyScroll: true
  };
  state = {
    shortcutsScope: getUID('ring-dialog-')
  };
  componentDidMount() {
    const {
      show,
      native
    } = this.props;
    if (native && show) {
      this.toggleNativeDialog();
    }
    this.toggleScrollPreventer();
  }
  componentDidUpdate(prevProps) {
    const {
      show,
      native
    } = this.props;
    if (native && show !== prevProps.show) {
      this.toggleNativeDialog();
    }
    if (prevProps.show !== this.props.show) {
      this.toggleScrollPreventer();
    }
  }
  componentWillUnmount() {
    this.scrollPreventer.reset();
  }
  scrollPreventer = preventerFactory(getUID('preventer-'));
  uid = getUID('dialog-');
  toggleNativeDialog() {
    const {
      show,
      modal
    } = this.props;
    if (this.nativeDialog.current) {
      if (show) {
        this.nativeDialog.current.removeAttribute('open');
        if (modal) {
          this.nativeDialog.current.showModal();
        } else {
          this.nativeDialog.current.show();
        }
      } else {
        this.nativeDialog.current.close();
      }
    }
  }
  toggleScrollPreventer() {
    if (!this.props.preventBodyScroll) {
      return;
    }
    if (this.props.show) {
      this.scrollPreventer.prevent();
    } else {
      this.scrollPreventer.reset();
    }
  }
  handleClick = event => {
    this.props.onOverlayClick(event);
    this.props.onCloseAttempt(event);
  };
  onCloseClick = event => {
    this.props.onCloseClick(event);
    this.props.onCloseAttempt(event);
  };
  getShortcutsMap = () => {
    const onEscape = event => {
      if (this.props.show) {
        this.props.onEscPress(event);
        this.props.onCloseAttempt(event);
      }
    };
    return {
      esc: onEscape
    };
  };
  dialog;
  dialogRef = tabTrap => {
    this.dialog = tabTrap && tabTrap.node;
  };
  nativeDialog = /*#__PURE__*/createRef();
  render() {
    const {
      show,
      showCloseButton,
      onOverlayClick,
      onCloseAttempt,
      onEscPress,
      onCloseClick,
      children,
      className,
      contentClassName,
      trapFocus,
      'data-test': dataTest,
      closeButtonInside,
      portalTarget,
      label,
      closeButtonTitle,
      dense,
      shortcutOptions,
      native,
      modal,
      preventBodyScroll,
      ...restProps
    } = this.props;
    const classes = classNames(styles.container, className);
    const shortcutsMap = this.getShortcutsMap();
    const content = /*#__PURE__*/jsxs(Fragment, {
      children: [/*#__PURE__*/jsx(Shortcuts, {
        map: shortcutsMap,
        scope: this.state.shortcutsScope,
        options: this.props.shortcutOptions
      }), (onOverlayClick !== noop || onCloseAttempt !== noop) && /*#__PURE__*/jsx("div", {
        // click handler is duplicated in close button
        role: "presentation",
        className: styles.clickableOverlay,
        onClick: this.handleClick,
        "data-test": "ring-dialog-overlay"
      }), /*#__PURE__*/jsx("div", {
        className: styles.innerContainer,
        children: /*#__PURE__*/jsxs(AdaptiveIsland, {
          className: classNames(styles.content, contentClassName, {
            [styles.dense]: dense
          }),
          "data-test": "ring-dialog",
          role: "dialog",
          "aria-label": label,
          children: [children, showCloseButton && /*#__PURE__*/jsx(Button, {
            icon: closeIcon,
            "data-test": "ring-dialog-close-button",
            className: classNames(styles.closeButton, {
              [styles.closeButtonOutside]: !closeButtonInside,
              [styles.closeButtonInside]: closeButtonInside
            }),
            iconClassName: classNames(styles.closeIcon, {
              [styles.closeIconOutside]: !closeButtonInside,
              [styles.closeIconInside]: closeButtonInside
            }),
            onClick: this.onCloseClick,
            title: closeButtonTitle,
            "aria-label": closeButtonTitle || 'close dialog'
          })]
        })
      })]
    });
    if (native) {
      return /*#__PURE__*/jsx("dialog", {
        className: classNames(styles.nativeDialog, className),
        ref: this.nativeDialog,
        "data-rg-modal-dialog-container": modal ? '' : undefined,
        children: /*#__PURE__*/jsx(PopupTarget, {
          id: this.uid,
          className: styles.popupTarget,
          children: target => /*#__PURE__*/jsxs(Fragment, {
            children: [content, target]
          })
        })
      });
    }
    return show && /*#__PURE__*/jsx(PopupTargetContext.Consumer, {
      children: contextTarget => {
        const normalizedContextTarget = normalizePopupTarget(contextTarget);
        let targetElement = document.body;
        if (portalTarget instanceof HTMLElement) {
          targetElement = portalTarget;
        } else if (normalizedContextTarget !== undefined) {
          const container = getPopupContainer(normalizedContextTarget);
          if (container) {
            targetElement = container;
          }
        }
        return /*#__PURE__*/createPortal(/*#__PURE__*/jsx(PopupTarget, {
          id: this.uid,
          className: styles.popupTarget,
          children: target => /*#__PURE__*/jsxs(TabTrap, {
            trapDisabled: !trapFocus,
            "data-test": joinDataTestAttributes('ring-dialog-container', dataTest),
            "data-rg-modal-dialog-container": "",
            ref: this.dialogRef,
            className: classes,
            role: "presentation",
            ...restProps,
            children: [content, target]
          })
        }), targetElement);
      }
    });
  }
}

export { Dialog as default };
