// @jetbrains/ring-ui-built/components/user-card/tooltip.js
import { Component } from 'react';
import classNames from 'classnames';
import Dropdown from '../dropdown/dropdown.js';
import Popup from '../popup/popup.js';
import { U as UserCard, s as styles } from '../_helpers/card.js';
import { jsx } from 'react/jsx-runtime';
import '../global/data-tests.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import 'react-compiler-runtime';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '@jetbrains/icons/copy';
import '../avatar/avatar.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../link/link.js';
import '../_helpers/link.js';
import '../clipboard/clipboard.js';
import '../alert-service/alert-service.js';
import 'react-dom/client';
import '../alert/alert.js';
import '@jetbrains/icons/exception';
import '@jetbrains/icons/checkmark';
import '@jetbrains/icons/warning';
import '@jetbrains/icons/close';
import '../loader-inline/loader-inline.js';
import '../alert/container.js';
import '../clipboard/clipboard-fallback.js';
import '../tag/tag.js';
import '@jetbrains/icons/close-12px';
import '../i18n/i18n-context.js';
import '../i18n/i18n.js';
import '../tooltip/tooltip.js';

const DEFAULT_TIMEOUT = 300;
class UserCardTooltip extends Component {
  static defaultProps = {
    renderUserCard: props => {
      const {
        user,
        ...restProps
      } = props;
      return user && /*#__PURE__*/jsx(UserCard, {
        user: user,
        ...restProps
      });
    },
    renderNoUser: () => '',
    dropdownProps: {
      hoverShowTimeOut: DEFAULT_TIMEOUT,
      hoverHideTimeOut: DEFAULT_TIMEOUT
    }
  };
  render() {
    const {
      children,
      renderUserCard,
      renderNoUser,
      dropdownProps,
      user,
      ...restProps
    } = this.props;
    return /*#__PURE__*/jsx(Dropdown, {
      anchor: children,
      hoverMode: true,
      clickMode: false,
      ...dropdownProps,
      children: /*#__PURE__*/jsx(Popup, {
        attached: false,
        children: user ? renderUserCard({
          ...restProps,
          user,
          className: classNames(styles.userCardSpaced, this.props.className)
        }) : renderNoUser()
      })
    });
  }
}

export { UserCardTooltip as default };
