// @jetbrains/ring-ui-built/components/popup-menu/popup-menu.js
import Popup from '../popup/popup.js';
import List from '../list/list.js';
import { jsx } from 'react/jsx-runtime';
import 'react';
import 'react-dom';
import 'classnames';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/data-tests.js';
import '../tab-trap/tab-trap.js';
import 'react-compiler-runtime';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'util-deprecate';
import 'memoize-one';
import '../global/memoize.js';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../avatar/avatar.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';

class PopupMenu extends Popup {
  static isItemType = List.isItemType;
  static ListProps = List.ListProps;
  static defaultProps = {
    ...List.defaultProps,
    ...Popup.defaultProps,
    renderOptimization: false,
    closeOnSelect: false,
    largeBorderRadius: true
  };
  onSelect = (item, event) => {
    if (this.props.closeOnSelect) {
      this._onCloseAttempt(event);
    }
    this.props.onSelect(item, event);
  };
  list;
  listRef = el => {
    this.list = el;
  };
  /** @override */
  getInternalContent() {
    const {
      className,
      ...props
    } = this.props;
    return /*#__PURE__*/jsx("div", {
      children: /*#__PURE__*/jsx(List, {
        ref: this.listRef,
        ...props,
        maxHeight: this.popup && parseFloat(this.popup.style.maxHeight),
        shortcuts: this.shouldUseShortcuts(),
        onSelect: this.onSelect
      })
    });
  }
}
const {
  ListProps
} = List;

export { ListProps, PopupMenu as default };
