// @jetbrains/ring-ui-built/components/island/adaptive-island-hoc.js
import { createContext, Component } from 'react';
import { interpolateLinear } from '../global/linear-function.js';
import { jsx } from 'react/jsx-runtime';

const TITLE_RESIZE_END = 20;
const TITLE_RESIZE_THRESHOLD = 36;
const PhaseContext = /*#__PURE__*/createContext(null);
const ScrollHandlerContext = /*#__PURE__*/createContext(null);
function adaptiveIslandHOC(ComposedComponent) {
  return class AdaptiveIsland extends Component {
    state = {
      phase: null
    };
    onContentScroll = ({
      scrollTop,
      scrollHeight,
      clientHeight
    }) => {
      var _this$state$phase;
      if (scrollHeight - clientHeight >= interpolateLinear(TITLE_RESIZE_THRESHOLD, TITLE_RESIZE_END, (_this$state$phase = this.state.phase) !== null && _this$state$phase !== void 0 ? _this$state$phase : 0)) {
        const phase = Math.min(1, Math.max(0, scrollTop) / TITLE_RESIZE_END);
        this.setState({
          phase
        });
      }
    };
    render() {
      return /*#__PURE__*/jsx(PhaseContext.Provider, {
        value: this.state.phase,
        children: /*#__PURE__*/jsx(ScrollHandlerContext.Provider, {
          value: this.onContentScroll,
          children: /*#__PURE__*/jsx(ComposedComponent, {
            ...this.props
          })
        })
      });
    }
  };
}

export { PhaseContext, ScrollHandlerContext, adaptiveIslandHOC as default };
