// @jetbrains/ring-ui-built/components/island/header.js
import { c } from 'react-compiler-runtime';
import { Component } from 'react';
import classNames from 'classnames';
import { interpolateLinear } from '../global/linear-function.js';
import { PhaseContext } from './adaptive-island-hoc.js';
import { s as styles } from '../_helpers/island.js';
import { jsx, jsxs } from 'react/jsx-runtime';

const Start = {
  FONT_SIZE: 24,
  LINE_HEIGHT: 28,
  PADDING_TOP: 24,
  PADDING_BOTTOM: 0,
  X: 0,
  Y: 0,
  SPACING: 0
};
const End = {
  FONT_SIZE: 14,
  LINE_HEIGHT: 20,
  PADDING_TOP: 16,
  PADDING_BOTTOM: 8,
  // Compensation
  X: 0.4,
  Y: 0.1,
  SPACING: 1.09
};
const BORDER_APPEAR_PHASE = 0.5;
class Header extends Component {
  static defaultProps = {
    wrapWithTitle: true
  };
  style(name) {
    var _this$props$phase;
    return interpolateLinear(Start[name], End[name], (_this$props$phase = this.props.phase) !== null && _this$props$phase !== void 0 ? _this$props$phase : 0);
  }
  getTitleStyle(phase) {
    if (phase === undefined) return;
    if (phase < 1) {
      const scaleFont = this.style('FONT_SIZE') / Start.FONT_SIZE;
      return {
        fontSize: Start.FONT_SIZE,
        transform: `translate(${this.style('X')}px, ${this.style('Y')}px) scale(${scaleFont})`,
        letterSpacing: this.style('SPACING')
      };
    }
    return {
      fontSize: End.FONT_SIZE
    };
  }
  render() {
    const {
      children,
      className,
      wrapWithTitle,
      border,
      phase,
      ...restProps
    } = this.props;
    const classes = classNames(styles.header, className, {
      [styles.withBottomBorder]: border || phase && phase >= BORDER_APPEAR_PHASE
    });
    const headerStyle = phase !== null && phase !== undefined ? {
      lineHeight: `${this.style('LINE_HEIGHT')}px`,
      // need to append px because number is a valid line-height value
      paddingTop: this.style('PADDING_TOP'),
      paddingBottom: this.style('PADDING_BOTTOM')
    } : undefined;
    const titleStyle = this.getTitleStyle(phase);
    return /*#__PURE__*/jsxs("div", {
      ...restProps,
      "data-test": "ring-island-header",
      className: classes,
      style: headerStyle,
      children: [wrapWithTitle && /*#__PURE__*/jsx("h2", {
        className: styles.title,
        style: titleStyle,
        children: children
      }), !wrapWithTitle && children]
    });
  }
}
const HeaderWrapper = props => {
  const $ = c(3);
  if ($[0] !== "dbbfd9db484c81cb5d8abfe2d50de2fd9035c22d61c0a499e8f82dfe385f2c0e") {
    for (let $i = 0; $i < 3; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "dbbfd9db484c81cb5d8abfe2d50de2fd9035c22d61c0a499e8f82dfe385f2c0e";
  }
  let t0;
  if ($[1] !== props) {
    t0 = /*#__PURE__*/jsx(PhaseContext.Consumer, {
      children: phase => {
        const addProps = phase !== null && phase !== undefined ? {
          phase
        } : {};
        return /*#__PURE__*/jsx(Header, {
          ...props,
          ...addProps
        });
      }
    });
    $[1] = props;
    $[2] = t0;
  } else {
    t0 = $[2];
  }
  return t0;
};

export { HeaderWrapper as default };
