// @jetbrains/ring-ui-built/components/island/island.js
import { Component } from 'react';
import classNames from 'classnames';
import joinDataTestAttributes from '../global/data-tests.js';
import adaptiveIslandHOC from './adaptive-island-hoc.js';
import { s as styles } from '../_helpers/island.js';
import { jsx } from 'react/jsx-runtime';
export { default as Header } from './header.js';
export { default as Content } from './content.js';
import '../global/linear-function.js';
import 'react-compiler-runtime';
import 'element-resize-detector';
import '../global/schedule-raf.js';

class Island extends Component {
  render() {
    const {
      children,
      className,
      narrow,
      withoutPaddings,
      'data-test': dataTest,
      ...restProps
    } = this.props;
    const classes = classNames(styles.island, className, {
      [styles.narrowIsland]: narrow,
      [styles.withoutPaddings]: withoutPaddings
    });
    return /*#__PURE__*/jsx("div", {
      ...restProps,
      className: classes,
      "data-test": joinDataTestAttributes('ring-island', dataTest),
      children: children
    });
  }
}
const AdaptiveIsland = adaptiveIslandHOC(Island);

export { AdaptiveIsland, Island as default };
