// @jetbrains/ring-ui-built/components/avatar/avatar.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import deprecate from 'util-deprecate';
import { isDataURI, parseQueryString, encodeURL } from '../global/url.js';
import { getPixelRatio } from '../global/dom.js';
import memoize from '../global/memoize.js';
import FallbackAvatar from './fallback-avatar.js';
import { Size } from './avatar-size.js';
import { s as styles, A as AvatarInfo } from '../_helpers/avatar-info.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react-compiler-runtime';
import '../global/get-uid.js';

const warnSize = memoize(size => deprecate(() => {}, `Avatar: Size${size} is deprecated and will be removed in 8.0. The supported sizes are: Size20, Size24, Size28, Size32, Size40.`));
class Avatar extends PureComponent {
  static defaultProps = {
    dpr: getPixelRatio(),
    size: Size.Size20,
    subavatarSize: Size.Size20 / 2,
    style: {}
  };
  state = {
    errorUrl: ''
  };
  handleError = () => {
    this.setState({
      errorUrl: this.props.url
    });
  };
  handleSuccess = () => {
    this.setState({
      errorUrl: ''
    });
  };
  render() {
    const {
      size,
      url,
      dpr,
      style,
      round,
      subavatar,
      subavatarSize,
      username,
      info,
      skipParams,
      ...restProps
    } = this.props;
    if ([Size.Size18, Size.Size48].includes(size)) {
      warnSize(size)();
    }
    const sizeString = `${size}px`;
    const subavatarSizeString = `${subavatarSize}px`;
    const styleObj = {
      height: sizeString,
      width: sizeString,
      ...style
    };
    const styleObjGroup = {
      borderRadius: '2px',
      height: subavatarSizeString,
      width: subavatarSizeString,
      ...style
    };
    const classes = classNames(styles.avatar, this.props.className, {
      [styles.round]: round
    });
    if (!url || this.state.errorUrl === url) {
      return /*#__PURE__*/jsxs("span", {
        ...restProps,
        "data-test": "avatar",
        className: classNames(classes, {
          [styles.empty]: (username === null || username === undefined) && (info === null || info === undefined)
        }),
        style: styleObj,
        children: [username && /*#__PURE__*/jsx(FallbackAvatar, {
          size: size,
          round: round,
          username: username
        }), info && /*#__PURE__*/jsx(AvatarInfo, {
          size: size,
          children: info
        })]
      });
    }
    let src = url;
    if (!skipParams && !isDataURI(url)) {
      const [urlStart, query] = url.split('?');
      const queryParams = {
        ...parseQueryString(query),
        dpr,
        size
      };
      src = encodeURL(urlStart, queryParams);
    }
    let subavatarSrc = null;
    if (subavatar && !isDataURI(subavatar)) {
      const [urlStart, query] = subavatar.split('?');
      const queryParams = {
        ...parseQueryString(query),
        dpr,
        subavatarSizeString
      };
      subavatarSrc = skipParams ? subavatar : encodeURL(urlStart, queryParams);
      return /*#__PURE__*/jsxs("div", {
        children: [/*#__PURE__*/jsx("img", {
          ...restProps,
          onError: this.handleError,
          onLoad: this.handleSuccess,
          className: classNames(classes, styles.avatarShadow),
          style: styleObj,
          src: src,
          alt: "User avatar"
        }), /*#__PURE__*/jsx("img", {
          ...restProps,
          "data-test": "avatar",
          onError: this.handleError,
          onLoad: this.handleSuccess,
          className: classNames(styles.subavatar),
          style: styleObjGroup,
          src: subavatarSrc,
          alt: "Subavatar"
        })]
      });
    }
    return /*#__PURE__*/jsx("img", {
      ...restProps,
      "data-test": "avatar",
      onError: this.handleError,
      onLoad: this.handleSuccess,
      className: classNames(classes, styles.avatarShadow),
      style: styleObj,
      src: src,
      alt: "User avatar"
    });
  }
}

export { Size, Avatar as default };
