// @jetbrains/ring-ui-built/components/avatar/fallback-avatar.js
import { c } from 'react-compiler-runtime';
import { useState } from 'react';
import getUID from '../global/get-uid.js';
import { Size } from './avatar-size.js';
import { jsx, jsxs } from 'react/jsx-runtime';

const colorPairs = [['#A5E074', '#E8E093'], ['#6AD8A2', '#79E98F'], ['#7FBAF6', '#7BEBB2'], ['#6DDC8C', '#97FFFF'], ['#8EE5C0', '#A2B6FF'], ['#73B4F5', '#78F8DC'], ['#BA8EFF', '#78DFD6'], ['#63AFF5', '#60E2F5'], ['#A192F7', '#93B2F7'], ['#B989FF', '#60E2F5'], ['#CA7EF7', '#9BBAF8'], ['#A192F7', '#CF94F4'], ['#AB7EE1', '#D58AE3'], ['#BE82FF', '#FFA0FF'], ['#C57EF2', '#F2799E'], ['#BE82FF', '#F2CC83'], ['#E16C9D', '#EF8AE8'], ['#F29766', '#EF8AE8'], ['#F27AA9', '#EAF087'], ['#F27AA9', '#FBBB8C'], ['#FFAB66', '#FFEBA0']];
const SizesSquare = {
  [Size.Size16]: {
    radius: 4,
    text: {
      x: 7.5,
      y: 11
    },
    fontSize: '8px',
    textAnchor: 'middle'
  },
  [Size.Size18]: {
    radius: 4,
    text: {
      x: 9,
      y: 13
    },
    fontSize: '11px',
    textAnchor: 'middle'
  },
  [Size.Size20]: {
    radius: 4,
    text: {
      x: 2,
      y: 10
    },
    fontSize: '9px',
    underscore: {
      x: 3,
      y: 16,
      width: 7,
      height: 1
    }
  },
  [Size.Size24]: {
    radius: 4,
    text: {
      x: 2,
      y: 12
    },
    fontSize: '11px',
    underscore: {
      x: 3,
      y: 18.5,
      width: 9,
      height: 1.5
    }
  },
  [Size.Size28]: {
    radius: 4,
    text: {
      x: 2,
      y: 14
    },
    fontSize: '13px',
    underscore: {
      x: 3,
      y: 22,
      width: 11,
      height: 2
    }
  },
  [Size.Size32]: {
    radius: 4,
    text: {
      x: 3,
      y: 17
    },
    fontSize: '15px',
    letterSpacing: 1,
    underscore: {
      x: 4,
      y: 26,
      width: 12,
      height: 2
    }
  },
  [Size.Size40]: {
    radius: 4,
    text: {
      x: 3,
      y: 21
    },
    fontSize: '19px',
    letterSpacing: 1,
    underscore: {
      x: 5,
      y: 32,
      width: 15,
      height: 2.5
    }
  },
  [Size.Size48]: {
    radius: 4,
    text: {
      x: 3,
      y: 21
    },
    fontSize: '19px',
    letterSpacing: 1,
    underscore: {
      x: 5,
      y: 32,
      width: 15,
      height: 2.5
    }
  },
  [Size.Size56]: {
    radius: 4,
    text: {
      x: 4,
      y: 28
    },
    fontSize: '26px',
    letterSpacing: 1,
    underscore: {
      x: 7,
      y: 45,
      width: 21,
      height: 3.5
    }
  }
};
const SizesRound = {
  [Size.Size16]: {
    radius: 4,
    fontSize: '8px',
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  },
  [Size.Size18]: {
    radius: 4,
    fontSize: '11px',
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  },
  [Size.Size20]: {
    radius: 4,
    fontSize: '9px',
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  },
  [Size.Size24]: {
    radius: 4,
    fontSize: '11px',
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  },
  [Size.Size28]: {
    radius: 4,
    fontSize: '13px',
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  },
  [Size.Size32]: {
    radius: 4,
    fontSize: '15px',
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle',
    letterSpacing: 1
  },
  [Size.Size40]: {
    radius: 4,
    fontSize: '19px',
    letterSpacing: 1,
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  },
  [Size.Size48]: {
    radius: 4,
    fontSize: '19px',
    letterSpacing: 1,
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  },
  [Size.Size56]: {
    radius: 4,
    fontSize: '26px',
    letterSpacing: 1,
    text: {
      x: '50%',
      y: '54%'
    },
    textAnchor: 'middle',
    dominantBaseline: 'middle'
  }
};
function extractLetters(name) {
  const names = name.split(/[\s._]+/).filter(Boolean).map(word => Array.from(word));
  if (names.length >= 2) {
    return names[0][0].toUpperCase() + names[1][0].toUpperCase();
  }
  if (names.length === 1) {
    if (names[0].length >= 2) {
      return names[0].slice(0, 2).join('').toUpperCase();
    }
    return `${names[0][0].toUpperCase()}X`;
  }
  return 'XX';
}
// https://gist.github.com/hyamamoto/fd435505d29ebfa3d9716fd2be8d42f0#gistcomment-2775538
const BASE = 32;
function hashCode(s) {
  let h = 0;
  for (let i = 0; i < s.length; i++) {
    h = Math.imul(BASE - 1, h) + s.charCodeAt(i) | 0;
  }
  return h;
}
function FallbackAvatar(t0) {
  const $ = c(38);
  if ($[0] !== "b668230117e1d8da4bff753c3c60cfa65c2cb38b8160a7b90f4e36892987cdc5") {
    for (let $i = 0; $i < 38; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "b668230117e1d8da4bff753c3c60cfa65c2cb38b8160a7b90f4e36892987cdc5";
  }
  const {
    username,
    size,
    round
  } = t0;
  const hash = Math.abs(hashCode(username.toLowerCase()));
  const [fromColor, toColor] = colorPairs[hash % colorPairs.length];
  const sizes = round ? SizesRound[size] : SizesSquare[size];
  const underscore = sizes.underscore;
  const radius = round ? "50%" : sizes.radius;
  const [gradientId] = useState(_temp);
  const t1 = `0 0 ${size} ${size}`;
  let t2;
  if ($[1] !== fromColor) {
    t2 = /*#__PURE__*/jsx("stop", {
      stopColor: fromColor,
      offset: "0"
    });
    $[1] = fromColor;
    $[2] = t2;
  } else {
    t2 = $[2];
  }
  let t3;
  if ($[3] !== toColor) {
    t3 = /*#__PURE__*/jsx("stop", {
      stopColor: toColor,
      offset: "0.8"
    });
    $[3] = toColor;
    $[4] = t3;
  } else {
    t3 = $[4];
  }
  let t4;
  if ($[5] !== gradientId || $[6] !== t2 || $[7] !== t3) {
    t4 = /*#__PURE__*/jsx("defs", {
      children: /*#__PURE__*/jsxs("linearGradient", {
        id: gradientId,
        x1: "1",
        y1: "1",
        x2: "0",
        y2: "0",
        children: [t2, t3]
      })
    });
    $[5] = gradientId;
    $[6] = t2;
    $[7] = t3;
    $[8] = t4;
  } else {
    t4 = $[8];
  }
  const t5 = `url(#${gradientId})`;
  let t6;
  if ($[9] !== radius || $[10] !== size || $[11] !== t5) {
    t6 = /*#__PURE__*/jsx("rect", {
      fill: t5,
      x: "0",
      y: "0",
      width: size,
      height: size,
      rx: radius,
      ry: radius
    });
    $[9] = radius;
    $[10] = size;
    $[11] = t5;
    $[12] = t6;
  } else {
    t6 = $[12];
  }
  const t7 = sizes.text.x;
  const t8 = sizes.text.y;
  const t9 = sizes.fontSize;
  const t10 = sizes.letterSpacing;
  const t11 = sizes.dominantBaseline;
  const t12 = sizes.textAnchor;
  let t13;
  if ($[13] !== username) {
    t13 = extractLetters(username);
    $[13] = username;
    $[14] = t13;
  } else {
    t13 = $[14];
  }
  let t14;
  if ($[15] !== t13) {
    t14 = /*#__PURE__*/jsx("tspan", {
      children: t13
    });
    $[15] = t13;
    $[16] = t14;
  } else {
    t14 = $[16];
  }
  let t15;
  if ($[17] !== sizes.dominantBaseline || $[18] !== sizes.fontSize || $[19] !== sizes.letterSpacing || $[20] !== sizes.text.x || $[21] !== sizes.text.y || $[22] !== sizes.textAnchor || $[23] !== t14) {
    t15 = /*#__PURE__*/jsx("text", {
      x: t7,
      y: t8,
      fontFamily: "var(--ring-font-family, 'Arial, Helvetica, sans-serif')",
      fontSize: t9,
      fontWeight: "600",
      letterSpacing: t10,
      fill: "#FFFFFF",
      dominantBaseline: t11,
      textAnchor: t12,
      cursor: "default",
      children: t14
    });
    $[17] = sizes.dominantBaseline;
    $[18] = sizes.fontSize;
    $[19] = sizes.letterSpacing;
    $[20] = sizes.text.x;
    $[21] = sizes.text.y;
    $[22] = sizes.textAnchor;
    $[23] = t14;
    $[24] = t15;
  } else {
    t15 = $[24];
  }
  let t16;
  if ($[25] !== underscore) {
    t16 = underscore && /*#__PURE__*/jsx("rect", {
      fill: "#FFFFFF",
      x: underscore.x,
      y: underscore.y,
      width: underscore.width,
      height: underscore.height
    });
    $[25] = underscore;
    $[26] = t16;
  } else {
    t16 = $[26];
  }
  let t17;
  if ($[27] !== username) {
    t17 = /*#__PURE__*/jsx("title", {
      children: username
    });
    $[27] = username;
    $[28] = t17;
  } else {
    t17 = $[28];
  }
  let t18;
  if ($[29] !== t15 || $[30] !== t16 || $[31] !== t17 || $[32] !== t6) {
    t18 = /*#__PURE__*/jsxs("g", {
      children: [t6, t15, t16, t17]
    });
    $[29] = t15;
    $[30] = t16;
    $[31] = t17;
    $[32] = t6;
    $[33] = t18;
  } else {
    t18 = $[33];
  }
  let t19;
  if ($[34] !== t1 || $[35] !== t18 || $[36] !== t4) {
    t19 = /*#__PURE__*/jsxs("svg", {
      viewBox: t1,
      xmlns: "http://www.w3.org/2000/svg",
      children: [t4, t18]
    });
    $[34] = t1;
    $[35] = t18;
    $[36] = t4;
    $[37] = t19;
  } else {
    t19 = $[37];
  }
  return t19;
}
function _temp() {
  return getUID("gradient-");
}

export { FallbackAvatar as default };
