// @jetbrains/ring-ui-built/components/error-bubble/error-bubble.js
import { PureComponent, Children, cloneElement } from 'react';
import classNames from 'classnames';
import Popup from '../popup/popup.js';
import { Directions } from '../popup/popup.consts.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/data-tests.js';
import '../tab-trap/tab-trap.js';
import 'react-compiler-runtime';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';

var styles = {"errorBubblePopup":"ring-error-bubble-errorBubblePopup","errorBubbleWrapper":"ring-error-bubble-errorBubbleWrapper","errorBubble":"ring-error-bubble-errorBubble"};

class ErrorBubble extends PureComponent {
  render() {
    const {
      children,
      className,
      ...restProps
    } = this.props;
    const errorBubbleClasses = classNames(styles.errorBubble, className);
    return /*#__PURE__*/jsxs("div", {
      className: styles.errorBubbleWrapper,
      "data-test": "ring-error-bubble-wrapper",
      children: [children && Children.map(children, child => /*#__PURE__*/cloneElement(child, {
        ...child.props,
        ...restProps
      })), restProps.error && /*#__PURE__*/jsx(Popup, {
        trapFocus: false,
        className: styles.errorBubblePopup,
        hidden: false,
        attached: false,
        directions: [Directions.RIGHT_CENTER, Directions.RIGHT_BOTTOM, Directions.RIGHT_TOP],
        children: /*#__PURE__*/jsx("div", {
          className: errorBubbleClasses,
          "data-test": "ring-error-bubble",
          children: restProps.error
        })
      })]
    });
  }
}

export { ErrorBubble as default };
