// @jetbrains/ring-ui-built/components/toggle/toggle.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import joinDataTestAttributes from '../global/data-tests.js';
import ControlHelp from '../control-help/control-help.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react-compiler-runtime';

var styles = {"toggle":"ring-toggle-toggle","switch":"ring-toggle-switch","disabled":"ring-toggle-disabled","label":"ring-toggle-label","help":"ring-toggle-help","leftLabel":"ring-toggle-leftLabel","switchWrapper":"ring-toggle-switchWrapper","input":"ring-toggle-input","size16":"ring-toggle-size16","size14":"ring-toggle-size14","size20":"ring-toggle-size20"};

const Size = {
  Size14: styles.size14,
  Size16: styles.size16,
  Size20: styles.size20
};
class Toggle extends PureComponent {
  static defaultProps = {
    size: Size.Size16
  };
  render() {
    const {
      className,
      children,
      disabled,
      title,
      leftLabel,
      size = Size.Size16,
      'data-test': dataTest,
      help,
      onTransitionEnd,
      ...restProps
    } = this.props;
    const classes = classNames(className, size, styles.toggle, disabled && styles.disabled);
    return /*#__PURE__*/jsxs("label", {
      className: classes,
      title: title,
      "data-test": joinDataTestAttributes('ring-toggle', dataTest),
      children: [leftLabel && /*#__PURE__*/jsxs("span", {
        className: styles.leftLabel,
        children: [leftLabel, help && /*#__PURE__*/jsx(ControlHelp, {
          className: styles.help,
          children: help
        })]
      }), /*#__PURE__*/jsxs("span", {
        className: styles.switchWrapper,
        children: [/*#__PURE__*/jsx("input", {
          "data-test": "ring-toggle-input",
          ...restProps,
          type: "checkbox",
          disabled: disabled,
          className: styles.input
        }), /*#__PURE__*/jsx("span", {
          className: styles.switch,
          onTransitionEnd: onTransitionEnd
        })]
      }), children && /*#__PURE__*/jsxs("div", {
        className: styles.label,
        children: [children, help && /*#__PURE__*/jsx(ControlHelp, {
          className: styles.help,
          children: help
        })]
      })]
    });
  }
}

export { Size, Toggle as default };
