// @jetbrains/ring-ui-built/components/_helpers/anchor.js
import { c } from 'react-compiler-runtime';
import { memo } from 'react';
import classNames from 'classnames';
import { Button } from '../button/button.js';
import { jsx } from 'react/jsx-runtime';

var styles = {"dropdown":"ring-dropdown-dropdown","clickInterceptor":"ring-dropdown-clickInterceptor","anchor":"ring-dropdown-anchor"};

const Anchor = t0 => {
  const $ = c(11);
  if ($[0] !== "3067fd735de1a9a0c23aa0a7b746ef5c5300c54dd93a2de8daf6d53e43ecf99a") {
    for (let $i = 0; $i < 11; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "3067fd735de1a9a0c23aa0a7b746ef5c5300c54dd93a2de8daf6d53e43ecf99a";
  }
  let children;
  let className;
  let restProps;
  if ($[1] !== t0) {
    ({
      children,
      className,
      ...restProps
    } = t0);
    $[1] = t0;
    $[2] = children;
    $[3] = className;
    $[4] = restProps;
  } else {
    children = $[2];
    className = $[3];
    restProps = $[4];
  }
  let t1;
  if ($[5] !== className) {
    t1 = classNames(styles.anchor, className);
    $[5] = className;
    $[6] = t1;
  } else {
    t1 = $[6];
  }
  let t2;
  if ($[7] !== children || $[8] !== restProps || $[9] !== t1) {
    t2 = /*#__PURE__*/jsx(Button, {
      "data-test-ring-dropdown-anchor": true,
      inline: true,
      dropdown: true,
      className: t1,
      ...restProps,
      children: children
    });
    $[7] = children;
    $[8] = restProps;
    $[9] = t1;
    $[10] = t2;
  } else {
    t2 = $[10];
  }
  return t2;
};
var Anchor$1 = /*#__PURE__*/memo(Anchor);

export { Anchor$1 as A, styles as s };
