// @jetbrains/ring-ui-built/components/_helpers/avatar-info.js
import { c } from 'react-compiler-runtime';
import classNames from 'classnames';
import { Size } from '../avatar/avatar-size.js';
import { jsx } from 'react/jsx-runtime';

var styles = {"avatar":"ring-avatar-avatar","round":"ring-avatar-round","avatarShadow":"ring-avatar-avatarShadow","avatarInfo":"ring-avatar-avatarInfo","subavatar":"ring-avatar-subavatar","empty":"ring-avatar-empty"};

const fontSizes = {
  [Size.Size16]: 9,
  [Size.Size18]: 9,
  [Size.Size20]: 9,
  [Size.Size24]: 11,
  [Size.Size28]: 12,
  [Size.Size32]: 14,
  [Size.Size40]: 16,
  [Size.Size48]: 16,
  [Size.Size56]: 22
};
function AvatarInfo(t0) {
  const $ = c(7);
  if ($[0] !== "db386ebce597a72ca92c29748d7d2b70e27ec0dd005eced756c361b3bfc587ee") {
    for (let $i = 0; $i < 7; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "db386ebce597a72ca92c29748d7d2b70e27ec0dd005eced756c361b3bfc587ee";
  }
  const {
    size,
    children
  } = t0;
  const fontSize = fontSizes[size];
  let t1;
  if ($[1] !== fontSize) {
    t1 = {
      fontSize
    };
    $[1] = fontSize;
    $[2] = t1;
  } else {
    t1 = $[2];
  }
  let t2;
  if ($[3] === Symbol.for("react.memo_cache_sentinel")) {
    t2 = classNames(styles.avatarInfo);
    $[3] = t2;
  } else {
    t2 = $[3];
  }
  let t3;
  if ($[4] !== children || $[5] !== t1) {
    t3 = /*#__PURE__*/jsx("span", {
      style: t1,
      className: t2,
      children: children
    });
    $[4] = children;
    $[5] = t1;
    $[6] = t3;
  } else {
    t3 = $[6];
  }
  return t3;
}

export { AvatarInfo as A, fontSizes as f, styles as s };
