// @jetbrains/ring-ui-built/components/_helpers/card.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import copyIcon from '@jetbrains/icons/copy';
import Avatar from '../avatar/avatar.js';
import Link from '../link/link.js';
import clipboard from '../clipboard/clipboard.js';
import Tag, { TagType } from '../tag/tag.js';
import Icon from '../icon/icon.js';
import { I18nContext } from '../i18n/i18n-context.js';
import Tooltip from '../tooltip/tooltip.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { Size } from '../avatar/avatar-size.js';
import { Size as Size$1 } from '../icon/icon.constants.js';

var styles = {"userCardSpaced":"ring-user-card-userCardSpaced","userInformationContainer":"ring-user-card-userInformationContainer","userAvatar":"ring-user-card-userAvatar","userInformation":"ring-user-card-userInformation","userInformationGeneral":"ring-user-card-userInformationGeneral","userNameLine":"ring-user-card-userNameLine","userName":"ring-user-card-userName","userLogin":"ring-user-card-userLogin","userEmail":"ring-user-card-userEmail","userCopyIcon":"ring-user-card-userCopyIcon","userEmailWrapper":"ring-user-card-userEmailWrapper","unverifiedLabel":"ring-user-card-unverifiedLabel","userNameInfo":"ring-user-card-userNameInfo","userActiveStatus":"ring-user-card-userActiveStatus","online":"ring-user-card-online"};

class UserCard extends PureComponent {
  static contextType = I18nContext;
  copyEmail = () => {
    var _translations$copiedT, _translations$copingT;
    const {
      translate
    } = this.context;
    const translations = this.props.translations;
    clipboard.copyText(this.props.user.email || '', (_translations$copiedT = translations?.copiedToClipboard) !== null && _translations$copiedT !== void 0 ? _translations$copiedT : translate('copyToClipboard'), (_translations$copingT = translations?.copingToClipboardError) !== null && _translations$copingT !== void 0 ? _translations$copingT : translate('copingToClipboardError'));
  };
  // eslint-disable-next-line complexity
  render() {
    var _translations$online, _translations$offline, _translations$banned, _translations$unverif, _translations$copyToC;
    const {
      children,
      info,
      className,
      user,
      avatarInfo,
      ...restProps
    } = this.props;
    const {
      translate
    } = this.context;
    const translations = this.props.translations;
    const classes = classNames(className, {});
    const userActiveStatusClasses = classNames(styles.userActiveStatus, user.online ? styles.online : '');
    return /*#__PURE__*/jsx("div", {
      ...restProps,
      className: classes,
      children: /*#__PURE__*/jsxs("div", {
        className: styles.userInformationContainer,
        children: [/*#__PURE__*/jsxs("div", {
          className: styles.userAvatar,
          children: [/*#__PURE__*/jsx(Avatar, {
            size: Size.Size56,
            url: user.avatarUrl,
            username: user.name,
            round: true
          }), !!avatarInfo && avatarInfo]
        }), /*#__PURE__*/jsxs("div", {
          className: styles.userInformation,
          children: [/*#__PURE__*/jsxs("div", {
            className: styles.userInformationGeneral,
            children: [/*#__PURE__*/jsxs("div", {
              className: styles.userNameLine,
              children: [user.href && /*#__PURE__*/jsx(Link, {
                href: user.href,
                className: styles.userName,
                "data-test": "user-card-link",
                children: user.name
              }), !user.href && /*#__PURE__*/jsx("span", {
                className: styles.userName,
                children: user.name
              }), typeof user.online === 'boolean' && /*#__PURE__*/jsx("span", {
                className: userActiveStatusClasses,
                title: user.online ? (_translations$online = translations?.online) !== null && _translations$online !== void 0 ? _translations$online : translate('online') : (_translations$offline = translations?.offline) !== null && _translations$offline !== void 0 ? _translations$offline : translate('offline')
              }), !!info && /*#__PURE__*/jsx("span", {
                className: styles.userNameInfo,
                children: info
              }), user.banned && /*#__PURE__*/jsx(Tooltip, {
                title: user.banReason,
                children: /*#__PURE__*/jsx(Tag, {
                  tagType: TagType.ERROR,
                  children: (_translations$banned = translations?.banned) !== null && _translations$banned !== void 0 ? _translations$banned : translate('banned')
                })
              })]
            }), /*#__PURE__*/jsx("div", {
              className: styles.userLogin,
              children: user.login
            }), user.email && /*#__PURE__*/jsxs("span", {
              className: styles.userEmailWrapper,
              children: [/*#__PURE__*/jsx(Link, {
                href: `mailto:${user.email}`,
                title: `mailto:${user.email}`,
                target: "_blank",
                className: styles.userEmail,
                children: user.email
              }), user.unverifiedEmail && /*#__PURE__*/jsx("span", {
                className: styles.unverifiedLabel,
                children: (_translations$unverif = translations?.unverified) !== null && _translations$unverif !== void 0 ? _translations$unverif : translate('unverified')
              }), /*#__PURE__*/jsx(Icon, {
                title: (_translations$copyToC = translations?.copyToClipboard) !== null && _translations$copyToC !== void 0 ? _translations$copyToC : translate('copyToClipboard'),
                className: styles.userCopyIcon,
                onClick: this.copyEmail,
                glyph: copyIcon,
                size: Size$1.Size14,
                suppressSizeWarning: true
              })]
            })]
          }), children]
        })]
      })
    });
  }
}

export { UserCard as U, styles as s };
