// @jetbrains/ring-ui-built/components/_helpers/input.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import closeIcon from '@jetbrains/icons/close-12px';
import { Button } from '../button/button.js';
import getUID from '../global/get-uid.js';
import Icon from '../icon/icon.js';
import { I18nContext } from '../i18n/i18n-context.js';
import { createComposedRef } from '../global/compose-refs.js';
import { ControlsHeightContext } from '../global/controls-height.js';
import { ControlLabel } from '../control-label/control-label.js';
import ControlHelp from '../control-help/control-help.js';
import { jsx, jsxs } from 'react/jsx-runtime';

var inputStyles = {"outerContainer":"ring-input-outerContainer","borderless":"ring-input-borderless","container":"ring-input-container","input":"ring-input-input","error":"ring-input-error","withIcon":"ring-input-withIcon","clearable":"ring-input-clearable","icon":"ring-input-icon","clear":"ring-input-clear","empty":"ring-input-empty","errorText":"ring-input-errorText","helpText":"ring-input-helpText","sizeS":"ring-input-sizeS","sizeM":"ring-input-sizeM","sizeL":"ring-input-sizeL","sizeFULL":"ring-input-sizeFULL","heightS":"ring-input-heightS","heightM":"ring-input-heightM","heightL":"ring-input-heightL"};

function noop() {}
/**
 * @name Input
 */
var Size;
(function (Size) {
  Size["AUTO"] = "Auto";
  Size["S"] = "S";
  Size["M"] = "M";
  Size["L"] = "L";
  Size["FULL"] = "FULL";
})(Size || (Size = {}));
class Input extends PureComponent {
  static defaultProps = {
    size: Size.M,
    onChange: noop,
    inputRef: noop,
    enableShortcuts: ['esc'],
    autogrow: true
  };
  state = {
    empty: true
  };
  componentDidMount() {
    this.frame = requestAnimationFrame(() => this.adapt());
  }
  componentDidUpdate() {
    this.adapt();
  }
  componentWillUnmount() {
    if (this.frame) {
      cancelAnimationFrame(this.frame);
    }
  }
  static contextType = ControlsHeightContext;
  frame;
  input;
  id = getUID('ring-input-');
  getId() {
    return this.props.id || this.id;
  }
  checkValue() {
    this.setState({
      empty: !this.input?.value
    });
    if (this.props.multiline && this.props.autogrow && this.input && this.input.scrollHeight >= this.input.clientHeight) {
      this.stretch(this.input);
    }
  }
  stretch(el) {
    if (!el || !el.style) {
      return;
    }
    el.style.height = '0'; // To know the real scrollHeight
    el.style.height = `${el.scrollHeight + 2}px`;
  }
  adapt() {
    this.checkValue();
  }
  inputRef = el => {
    this.input = el;
  };
  composedInputRef = createComposedRef();
  clear = e => {
    this.props.onClear?.(e);
  };
  handleInputChange = e => {
    if (!this.props.multiline) {
      this.props.onChange?.(e);
      this.checkValue();
    }
  };
  handleTextareaChange = e => {
    if (this.props.multiline) {
      this.props.onChange?.(e);
      this.checkValue();
    }
  };
  render() {
    const {
      // Modifiers
      size,
      multiline,
      borderless,
      // Props
      label,
      labelType,
      error,
      help,
      className,
      inputClassName,
      children,
      value,
      onClear,
      disabled,
      inputRef,
      onChange,
      enableShortcuts,
      id,
      placeholder,
      icon,
      translations,
      height = typeof this.context === 'function' ? this.context() : this.context,
      beforeInput,
      afterInput,
      autogrow,
      ...restProps
    } = this.props;
    const {
      empty
    } = this.state;
    const clearable = !!onClear;
    const classes = classNames(className, inputStyles.outerContainer, [inputStyles[`size${size}`]], [inputStyles[`height${height}`]], {
      'ring-js-shortcuts': enableShortcuts === true,
      [inputStyles.error]: error !== null && error !== undefined,
      [inputStyles.empty]: empty,
      [inputStyles.withIcon]: icon,
      [inputStyles.clearable]: clearable,
      [inputStyles.borderless]: borderless
    });
    const inputClasses = classNames(inputStyles.input, inputClassName);
    const text = value !== null && value !== void 0 ? value : children;
    const commonProps = {
      ref: this.composedInputRef(this.inputRef, inputRef),
      className: inputClasses,
      value: text,
      disabled,
      id: this.getId(),
      placeholder,
      'aria-label': typeof label === 'string' && label ? label : placeholder,
      'data-enabled-shortcuts': Array.isArray(enableShortcuts) ? enableShortcuts.join(',') : null
    };
    return /*#__PURE__*/jsx(I18nContext.Consumer, {
      children: ({
        translate
      }) => {
        var _translations$clear;
        return /*#__PURE__*/jsxs("div", {
          className: classes,
          "data-test": "ring-input",
          children: [label && /*#__PURE__*/jsx(ControlLabel, {
            htmlFor: this.getId(),
            disabled: disabled,
            type: labelType,
            children: label
          }), /*#__PURE__*/jsxs("div", {
            className: inputStyles.container,
            children: [icon && /*#__PURE__*/jsx(Icon, {
              glyph: icon,
              className: inputStyles.icon
            }), beforeInput, multiline ? /*#__PURE__*/jsx("textarea", {
              onChange: this.handleTextareaChange,
              rows: 1,
              ...commonProps,
              ...restProps
            }) : /*#__PURE__*/jsx("input", {
              onChange: this.handleInputChange,
              ...commonProps,
              ...restProps
            }), clearable && !disabled && /*#__PURE__*/jsx(Button, {
              title: (_translations$clear = translations?.clear) !== null && _translations$clear !== void 0 ? _translations$clear : translate('clear'),
              "data-test": "ring-input-clear",
              className: classNames(inputStyles.clear, this.props.clearButtonClassName),
              icon: closeIcon,
              onClick: this.clear
            }), afterInput]
          }), error ? /*#__PURE__*/jsx("div", {
            className: inputStyles.errorText,
            children: error
          }) : help && /*#__PURE__*/jsx(ControlHelp, {
            className: inputStyles.helpText,
            children: help
          })]
        });
      }
    });
  }
}

export { Input as I, Size as S, inputStyles as i };
