// @jetbrains/ring-ui-built/components/_helpers/select-filter.js
import { Component } from 'react';
import classNames from 'classnames';
import { I as Input } from './input.js';
import sniffr from '../global/sniffer.js';
import { ActiveItemContext } from '../list/list.js';
import { I18nContext } from '../i18n/i18n-context.js';
import { jsx } from 'react/jsx-runtime';

var styles = {"filterWithTagsFocused":"ring-select-popup-filterWithTagsFocused","filterWithTags":"ring-select-popup-filterWithTags","filterWrapper":"ring-select-popup-filterWrapper","filterWithTagsInput":"ring-select-popup-filterWithTagsInput","filter":"ring-select-popup-filter","popup":"ring-select-popup-popup","filterIcon":"ring-select-popup-filterIcon","bottomLine":"ring-select-popup-bottomLine","bottomLineOverItem":"ring-select-popup-bottomLineOverItem","message":"ring-select-popup-message","selectAll":"ring-select-popup-selectAll"};

function noop() {}
class SelectFilter extends Component {
  static defaultProps = {
    inputRef: noop
  };
  focus() {
    const {
      input
    } = this;
    if (input && input !== document.activeElement) {
      if (sniffr.browser.name === 'firefox') {
        input.select();
      } else {
        input.focus();
      }
    }
  }
  blur() {
    if (this.input && this.input === document.activeElement) {
      this.input.blur();
    }
  }
  input;
  inputRef = el => {
    this.input = el;
    this.props.inputRef(el);
  };
  render() {
    const {
      className,
      listId,
      ...restProps
    } = this.props;
    const classes = classNames(styles.filter, className);
    return /*#__PURE__*/jsx(ActiveItemContext.ValueContext.Consumer, {
      children: activeItemId => /*#__PURE__*/jsx(I18nContext.Consumer, {
        children: ({
          translate
        }) => {
          var _restProps$placeholde;
          return /*#__PURE__*/jsx(Input, {
            ...restProps,
            placeholder: (_restProps$placeholde = restProps.placeholder) !== null && _restProps$placeholde !== void 0 ? _restProps$placeholde : translate('filterItems'),
            "aria-owns": listId,
            "aria-activedescendant": activeItemId,
            autoComplete: "off",
            autoFocus: true,
            borderless: true,
            inputRef: this.inputRef,
            className: classes
          });
        }
      })
    });
  }
}

export { SelectFilter as S, styles as s };
