// @jetbrains/ring-ui-built/components/_helpers/services-link.js
import { PureComponent } from 'react';
import Link from '../link/link.js';
import { jsx } from 'react/jsx-runtime';

var styles = {"services":"ring-services-services","active":"ring-services-active","item":"ring-services-item","itemLogo":"ring-services-itemLogo","activeItem":"ring-services-activeItem ring-services-item ring-services-active","line":"ring-services-line","itemStacked":"ring-services-itemStacked","activeItemStacked":"ring-services-activeItemStacked ring-services-itemStacked ring-services-active"};

class ServicesLink extends PureComponent {
  render() {
    const {
      service,
      ...props
    } = this.props;
    return /*#__PURE__*/jsx(Link, {
      target: "_self",
      href: service.homeUrl,
      ...props,
      children: service.iconUrl ? [/*#__PURE__*/jsx("span", {
        className: styles.itemLogo,
        style: {
          backgroundImage: `url(${service.iconUrl})`
        }
      }, 'icon'), /*#__PURE__*/jsx("div", {
        children: service.name
      }, 'text')] : service.name
    });
  }
}

export { ServicesLink as S, styles as s };
