// @jetbrains/ring-ui-built/components/_helpers/sidebar.js
import { Component } from 'react';
import classNames from 'classnames';
import { Waypoint } from 'react-waypoint';
import { jsxs, jsx } from 'react/jsx-runtime';

var styles = {"extra-small-screen-media":"(max-width: 639px)","small-screen-media":"(min-width: 640px) and (max-width: 959px)","contentLayout":"ring-content-layout-contentLayout","contentLayoutContent":"ring-content-layout-contentLayoutContent","sidebarContainer":"ring-content-layout-sidebarContainer","sidebarContainerRight":"ring-content-layout-sidebarContainerRight","sidebar":"ring-content-layout-sidebar","sidebarRight":"ring-content-layout-sidebarRight","sidebarFixedTop":"ring-content-layout-sidebarFixedTop","sidebarFixedBottom":"ring-content-layout-sidebarFixedBottom","bottomMarker":"ring-content-layout-bottomMarker","contentLayoutResponsive":"ring-content-layout-contentLayoutResponsive"};

const ABOVE = 'above';
const INSIDE = 'inside';
/**
 * @name Sidebar
 * @deprecated Will be removed in Ring UI 8.0.
 */
class Sidebar extends Component {
  state = {
    topIsOutside: true,
    bottomIsOutside: true,
    sidebarVisibleHeight: undefined
  };
  sidebarNode;
  handleTopWaypoint = ({
    currentPosition
  }) => {
    this.setState({
      topIsOutside: currentPosition === ABOVE
    });
  };
  handleBottomWaypoint = ({
    currentPosition,
    waypointTop
  }) => {
    this.setState({
      sidebarVisibleHeight: waypointTop,
      bottomIsOutside: currentPosition !== INSIDE
    });
  };
  shouldUseFixation() {
    const {
      contentNode
    } = this.props;
    const {
      sidebarNode
    } = this;
    if (!contentNode || !sidebarNode) {
      return false;
    }
    return contentNode.offsetHeight >= sidebarNode.offsetHeight;
  }
  shouldFixateBottom() {
    const {
      topIsOutside,
      bottomIsOutside
    } = this.state;
    return !bottomIsOutside && topIsOutside && this.shouldUseFixation();
  }
  sidebarRef = node => {
    this.sidebarNode = node;
  };
  render() {
    const {
      right,
      children,
      className,
      containerClassName,
      fixedClassName,
      contentNode,
      ...restProps
    } = this.props;
    const {
      topIsOutside,
      bottomIsOutside,
      sidebarVisibleHeight
    } = this.state;
    const shouldFixateTop = bottomIsOutside && topIsOutside && this.shouldUseFixation();
    const shouldFixateBottom = this.shouldFixateBottom();
    const containerClasses = classNames(styles.sidebarContainer, containerClassName, {
      [styles.sidebarContainerRight]: right
    });
    const classes = classNames(styles.sidebar, className, {
      [styles.sidebarRight]: right,
      [styles.sidebarFixedTop]: shouldFixateTop,
      [styles.sidebarFixedBottom]: shouldFixateBottom,
      [fixedClassName !== null && fixedClassName !== void 0 ? fixedClassName : '']: shouldFixateTop || shouldFixateBottom
    });
    const style = {
      maxHeight: shouldFixateBottom && sidebarVisibleHeight ? `${sidebarVisibleHeight}px` : undefined
    };
    return /*#__PURE__*/jsxs("aside", {
      className: containerClasses,
      ref: this.sidebarRef,
      children: [/*#__PURE__*/jsx(Waypoint, {
        onEnter: this.handleTopWaypoint,
        onLeave: this.handleTopWaypoint
      }), /*#__PURE__*/jsx("div", {
        ...restProps,
        style: style,
        className: classes,
        children: children
      }), /*#__PURE__*/jsx("div", {
        className: styles.bottomMarker,
        children: /*#__PURE__*/jsx(Waypoint, {
          onEnter: this.handleBottomWaypoint,
          onLeave: this.handleBottomWaypoint
        })
      })]
    });
  }
}

export { Sidebar as S, styles as s };
