// @jetbrains/ring-ui-built/components/_helpers/theme.js
import { c } from 'react-compiler-runtime';
import { createContext, forwardRef, useEffect, useState } from 'react';
import classNames from 'classnames';
import { jsx } from 'react/jsx-runtime';

var defaultStyles = {"light":"ring-variables-light"};

var darkStyles = {"dark":"ring-variables_dark-dark"};

var Theme;
(function (Theme) {
  Theme["AUTO"] = "auto";
  Theme["LIGHT"] = "light";
  Theme["DARK"] = "dark";
})(Theme || (Theme = {}));
const ThemeContext = /*#__PURE__*/createContext({
  theme: Theme.LIGHT
});
const GLOBAL_DARK_CLASS_NAME = 'ring-ui-theme-dark';
const darkMatcher = window.matchMedia('(prefers-color-scheme: dark)');
function useTheme() {
  const $ = c(3);
  if ($[0] !== "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682") {
    for (let $i = 0; $i < 3; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682";
  }
  const [dark, setDark] = useState(darkMatcher.matches);
  let t0;
  let t1;
  if ($[1] === Symbol.for("react.memo_cache_sentinel")) {
    t0 = () => {
      const onChange = e => setDark(e.matches);
      darkMatcher.addEventListener("change", onChange);
      return () => darkMatcher.removeEventListener("change", onChange);
    };
    t1 = [];
    $[1] = t0;
    $[2] = t1;
  } else {
    t0 = $[1];
    t1 = $[2];
  }
  useEffect(t0, t1);
  return dark ? Theme.DARK : Theme.LIGHT;
}
function useThemeClasses(theme) {
  const $ = c(5);
  if ($[0] !== "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682") {
    for (let $i = 0; $i < 5; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682";
  }
  const systemTheme = useTheme();
  const resolvedTheme = theme === Theme.AUTO ? systemTheme : theme;
  const t0 = resolvedTheme === Theme.DARK;
  const t1 = resolvedTheme === Theme.DARK;
  const t2 = resolvedTheme === Theme.LIGHT;
  let t3;
  if ($[1] !== t0 || $[2] !== t1 || $[3] !== t2) {
    t3 = classNames({
      [darkStyles.dark]: t0,
      [GLOBAL_DARK_CLASS_NAME]: t1,
      [defaultStyles.light]: t2
    });
    $[1] = t0;
    $[2] = t1;
    $[3] = t2;
    $[4] = t3;
  } else {
    t3 = $[4];
  }
  return t3;
}
function WithThemeClasses(t0) {
  const $ = c(4);
  if ($[0] !== "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682") {
    for (let $i = 0; $i < 4; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682";
  }
  const {
    theme,
    children
  } = t0;
  const themeClasses = useThemeClasses(theme);
  let t1;
  if ($[1] !== children || $[2] !== themeClasses) {
    t1 = children(themeClasses);
    $[1] = children;
    $[2] = themeClasses;
    $[3] = t1;
  } else {
    t1 = $[3];
  }
  return t1;
}
function applyTheme(theme, container) {
  if (theme === Theme.DARK) {
    container.classList.remove(defaultStyles.light);
    container.classList.add(darkStyles.dark);
    container.classList.add(GLOBAL_DARK_CLASS_NAME);
  } else {
    container.classList.remove(darkStyles.dark);
    container.classList.remove(GLOBAL_DARK_CLASS_NAME);
    container.classList.add(defaultStyles.light);
  }
}
const DefaultWrapper = /*#__PURE__*/forwardRef(function Wrapper(props, ref) {
  const $ = c(4);
  if ($[0] !== "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682") {
    for (let $i = 0; $i < 4; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682";
  }
  let t0;
  if ($[1] !== props || $[2] !== ref) {
    t0 = /*#__PURE__*/jsx("div", {
      ...props,
      ref: ref
    });
    $[1] = props;
    $[2] = ref;
    $[3] = t0;
  } else {
    t0 = $[3];
  }
  return t0;
});
function ThemeProviderInner(t0) {
  const $ = c(30);
  if ($[0] !== "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682") {
    for (let $i = 0; $i < 30; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682";
  }
  let children;
  let className;
  let passToPopups;
  let restProps;
  let t1;
  let t2;
  let target;
  let wrapperRef;
  if ($[1] !== t0) {
    ({
      theme: t1,
      className,
      passToPopups,
      children,
      WrapperComponent: t2,
      target,
      wrapperRef,
      ...restProps
    } = t0);
    $[1] = t0;
    $[2] = children;
    $[3] = className;
    $[4] = passToPopups;
    $[5] = restProps;
    $[6] = t1;
    $[7] = t2;
    $[8] = target;
    $[9] = wrapperRef;
  } else {
    children = $[2];
    className = $[3];
    passToPopups = $[4];
    restProps = $[5];
    t1 = $[6];
    t2 = $[7];
    target = $[8];
    wrapperRef = $[9];
  }
  const theme = t1 === undefined ? Theme.AUTO : t1;
  const WrapperComponent = t2 === undefined ? DefaultWrapper : t2;
  const systemTheme = useTheme();
  const resolvedTheme = theme === Theme.AUTO ? systemTheme : theme;
  let t3;
  if ($[10] !== passToPopups || $[11] !== resolvedTheme) {
    t3 = {
      theme: resolvedTheme,
      passToPopups
    };
    $[10] = passToPopups;
    $[11] = resolvedTheme;
    $[12] = t3;
  } else {
    t3 = $[12];
  }
  const themeValue = t3;
  let t4;
  let t5;
  if ($[13] !== resolvedTheme || $[14] !== target) {
    t4 = () => {
      if (target) {
        applyTheme(resolvedTheme, target);
      }
    };
    t5 = [resolvedTheme, target];
    $[13] = resolvedTheme;
    $[14] = target;
    $[15] = t4;
    $[16] = t5;
  } else {
    t4 = $[15];
    t5 = $[16];
  }
  useEffect(t4, t5);
  const themeClasses = useThemeClasses(theme);
  let t6;
  if ($[17] !== className || $[18] !== target || $[19] !== themeClasses) {
    t6 = target ? undefined : classNames(className, themeClasses);
    $[17] = className;
    $[18] = target;
    $[19] = themeClasses;
    $[20] = t6;
  } else {
    t6 = $[20];
  }
  let t7;
  if ($[21] !== WrapperComponent || $[22] !== children || $[23] !== restProps || $[24] !== t6 || $[25] !== wrapperRef) {
    t7 = /*#__PURE__*/jsx(WrapperComponent, {
      ref: wrapperRef,
      className: t6,
      ...restProps,
      children: children
    });
    $[21] = WrapperComponent;
    $[22] = children;
    $[23] = restProps;
    $[24] = t6;
    $[25] = wrapperRef;
    $[26] = t7;
  } else {
    t7 = $[26];
  }
  let t8;
  if ($[27] !== t7 || $[28] !== themeValue) {
    t8 = /*#__PURE__*/jsx(ThemeContext.Provider, {
      value: themeValue,
      children: t7
    });
    $[27] = t7;
    $[28] = themeValue;
    $[29] = t8;
  } else {
    t8 = $[29];
  }
  return t8;
}
const ThemeProvider = /*#__PURE__*/forwardRef(function ThemeProvider(props, ref) {
  const $ = c(4);
  if ($[0] !== "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682") {
    for (let $i = 0; $i < 4; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "eabfde84870498e54e35112b5ff33c796334250351dda2073f3913df7a796682";
  }
  let t0;
  if ($[1] !== props || $[2] !== ref) {
    t0 = /*#__PURE__*/jsx(ThemeProviderInner, {
      ...props,
      wrapperRef: ref
    });
    $[1] = props;
    $[2] = ref;
    $[3] = t0;
  } else {
    t0 = $[3];
  }
  return t0;
});
var Theme$1 = Theme;

export { GLOBAL_DARK_CLASS_NAME as G, Theme$1 as T, WithThemeClasses as W, ThemeProvider as a, ThemeContext as b, useThemeClasses as c, darkStyles as d, applyTheme as e, useTheme as u };
