// @jetbrains/ring-ui-built/components/_helpers/title.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import focusSensorHOC from '../global/focus-sensor-hoc.js';
import Checkbox from '../checkbox/checkbox.js';
import getUID from '../global/get-uid.js';
import { jsxs, jsx } from 'react/jsx-runtime';

var styles = {"dataListWrapper":"ring-data-list-dataListWrapper","dataList":"ring-data-list-dataList","itemContent":"ring-data-list-itemContent","title":"ring-data-list-title","disabledHover":"ring-data-list-disabledHover","titleSelected":"ring-data-list-titleSelected","titleFocused":"ring-data-list-titleFocused","showMore":"ring-data-list-showMore","boxes":"ring-data-list-boxes","checkboxBox":"ring-data-list-checkboxBox","collapseButton":"ring-data-list-collapseButton","collapseIcon":"ring-data-list-collapseIcon","loadingOverlay":"ring-data-list-loadingOverlay","showMoreLoader":"ring-data-list-showMoreLoader"};

class Title extends PureComponent {
  static defaultProps = {
    selectable: false,
    selected: false,
    showFocus: false
  };
  id = getUID('data-list-title');
  onCheckboxFocus = () => {
    this.props.onFocusRestore();
  };
  onCheckboxChange = () => {
    this.toggleSelection();
  };
  toggleSelection() {
    const {
      selectable,
      selected,
      onSelect
    } = this.props;
    if (selectable) {
      onSelect(!selected);
    }
  }
  render() {
    const {
      className,
      title,
      offset,
      showFocus,
      innerRef,
      selectable,
      selected,
      partialSelected,
      collapserExpander
    } = this.props;
    const classes = classNames(className, {
      [styles.title]: true,
      [styles.titleFocused]: showFocus,
      [styles.titleSelected]: selected
    });
    return /*#__PURE__*/jsxs("div", {
      id: this.id,
      className: classes,
      style: {
        paddingLeft: offset
      },
      ref: innerRef,
      children: [/*#__PURE__*/jsxs("div", {
        className: styles.boxes,
        children: [selectable && /*#__PURE__*/jsx("div", {
          className: styles.checkboxBox,
          children: /*#__PURE__*/jsx(Checkbox, {
            "aria-labelledby": this.id,
            className: showFocus ? 'ring-checkbox_focus' : '',
            checked: selected,
            indeterminate: partialSelected,
            onFocus: this.onCheckboxFocus,
            onChange: this.onCheckboxChange,
            tabIndex: -1
          })
        }), collapserExpander]
      }), title]
    });
  }
}
var Title$1 = focusSensorHOC(Title);

export { Title$1 as T, styles as s };
