// @jetbrains/ring-ui-built/components/tag/tag.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import closeIcon from '@jetbrains/icons/close-12px';
import Icon from '../icon/icon.js';
import { Button } from '../button/button.js';
import '../global/controls-height.js';
import joinDataTestAttributes from '../global/data-tests.js';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { ControlsHeight } from '../global/configuration.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../_helpers/button.classes.js';

var styles = {"interactive":"ring-tag-interactive","tag":"ring-tag-tag ring-global-resetButton","withRemove":"ring-tag-withRemove","container":"ring-tag-container","default":"ring-tag-default","main":"ring-tag-main","success":"ring-tag-success","warning":"ring-tag-warning","error":"ring-tag-error","purple":"ring-tag-purple","outline":"ring-tag-outline","focused":"ring-tag-focused","content":"ring-tag-content ring-global-ellipsis","disabled":"ring-tag-disabled","remove":"ring-tag-remove","icon":"ring-tag-icon","avatarContainer":"ring-tag-avatarContainer","customIcon":"ring-tag-customIcon","avatarIcon":"ring-tag-avatarIcon"};

var TagType;
(function (TagType) {
  TagType["DEFAULT"] = "default";
  TagType["MAIN"] = "main";
  TagType["SUCCESS"] = "success";
  TagType["WARNING"] = "warning";
  TagType["ERROR"] = "error";
  TagType["PURPLE"] = "purple";
})(TagType || (TagType = {}));
/**
 * @name Tag
 */
class Tag extends PureComponent {
  static defaultProps = {
    onClick: () => {},
    readOnly: false,
    disabled: false,
    focused: false,
    interactive: true,
    render: (props, interactive) => interactive ? /*#__PURE__*/jsx("button", {
      type: "button",
      ...props
    }) : /*#__PURE__*/jsx("span", {
      ...props
    }),
    tagType: TagType.DEFAULT
  };
  state = {
    focused: false
  };
  componentDidUpdate(prevProps) {
    if (this.props.focused !== prevProps.focused) {
      this.setState({
        focused: this.props.focused
      });
    }
    if (this.state.focused) {
      this.tagNode?.focus();
    }
    this.setDocumentClickListener(this.state.focused);
  }
  componentWillUnmount() {
    this.setDocumentClickListener(false);
    this.setState({
      focused: false
    });
  }
  static Type = TagType;
  onDocumentClick = event => {
    if (this.tagNode) {
      this.setState({
        focused: this.tagNode === event.target
      });
    }
  };
  tagNode;
  tagRef = el => {
    this.tagNode = el;
  };
  setDocumentClickListener(setListener) {
    if (setListener) {
      document.addEventListener('click', this.onDocumentClick);
    } else {
      document.removeEventListener('click', this.onDocumentClick);
    }
  }
  renderCustomIcon() {
    if (this.props.rgTagIcon) {
      return /*#__PURE__*/jsx(Icon, {
        className: styles.icon,
        title: this.props.rgTagTitle,
        glyph: this.props.rgTagIcon
      });
    }
    return null;
  }
  _renderImageElement(avatarSrc) {
    const classes = classNames({
      [styles.customIcon]: this.props.icon,
      [styles.avatarIcon]: avatarSrc
    });
    return /*#__PURE__*/jsx("img", {
      alt: avatarSrc ? 'Avatar' : 'Icon',
      className: classes,
      src: avatarSrc || this.props.icon
    });
  }
  renderImage() {
    if (this.props.icon && !this.props.avatar) {
      return this._renderImageElement();
    }
    return null;
  }
  renderAvatar() {
    if (this.props.avatar) {
      return /*#__PURE__*/jsx("span", {
        className: styles.avatarContainer,
        children: this._renderImageElement(this.props.avatar)
      });
    }
    return null;
  }
  renderRemoveIcon() {
    if (!this.props.readOnly && this.props.onRemove) {
      return /*#__PURE__*/jsx(Button, {
        title: "Remove",
        icon: closeIcon,
        "data-test": "ring-tag-remove",
        className: classNames(styles.remove, this.props.removeButtonClassName),
        onClick: this.props.onRemove,
        style: {
          '--ring-secondary-color': this.props.textColor
        },
        height: ControlsHeight.M
      });
    }
    return null;
  }
  render() {
    const classes = classNames('ring-js-shortcuts', styles.tag, {
      [styles.focused]: this.state.focused,
      [styles.disabled]: this.props.disabled,
      [styles.withRemove]: !this.props.readOnly && this.props.onRemove
    }, this.props.className);
    const {
      backgroundColor,
      textColor,
      render,
      'data-test': dataTest
    } = this.props;
    return /*#__PURE__*/jsxs("span", {
      className: classNames(styles.container, this.props.containerClassName, styles[this.props.tagType], {
        [styles.outline]: this.props.outline,
        [styles.interactive]: this.props.interactive
      }),
      children: [render({
        'data-test': joinDataTestAttributes('ring-tag', dataTest),
        className: classes,
        ref: this.tagRef,
        onClick: this.props.onClick,
        style: {
          backgroundColor,
          color: textColor
        },
        disabled: this.props.disabled,
        children: /*#__PURE__*/jsxs(Fragment, {
          children: [this.renderAvatar(), this.renderCustomIcon(), this.renderImage(), /*#__PURE__*/jsx("span", {
            className: styles.content,
            children: this.props.children
          })]
        })
      }, this.props.interactive), this.renderRemoveIcon()]
    });
  }
}

export { TagType, Tag as default };
