// @jetbrains/ring-ui-built/components/auth/auth.js
import WindowFlow from './window-flow.js';
import onBackendDown from './down-notification.js';
import Auth from './auth-core.js';
export { DEFAULT_BACKGROUND_TIMEOUT, DEFAULT_EXPIRES_TIMEOUT, DOMAIN_USER_CHANGED_EVENT, LOGOUT_EVENT, LOGOUT_POSTPONED_EVENT, USER_CHANGED_EVENT, USER_CHANGE_POSTPONED_EVENT } from './auth-core.js';
import './response-parser.js';
import '../global/url.js';
import 'react-compiler-runtime';
import 'react';
import '../alert-service/alert-service.js';
import 'react-dom/client';
import '../global/get-uid.js';
import '../alert/alert.js';
import 'classnames';
import '@jetbrains/icons/exception';
import '@jetbrains/icons/checkmark';
import '@jetbrains/icons/warning';
import '@jetbrains/icons/close';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import 'react/jsx-runtime';
import '../loader-inline/loader-inline.js';
import '../global/data-tests.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../_helpers/theme.js';
import '../alert/container.js';
import 'react-dom';
import '../link/link.js';
import '../_helpers/link.js';
import '../group/group.js';
import '../global/listeners.js';
import '../http/http.js';
import '../global/promise-with-timeout.js';
import '../i18n/i18n.js';
import './storage.js';
import '../storage/storage-local.js';
import './request-builder.js';
import 'simply-uuid';
import './background-flow.js';
import './token-validator.js';

/**
 * @name Auth
 *
 * @prop {object} config
 * @prop {string} config.serverUri
 * @prop {string} config.redirectUri
 * @prop {string} config.clientId
 * @prop {boolean=false} config.redirect — use redirects instead of loading the token in the background.
 * @prop {string[]} config.scope
 * @prop {string[]} config.optionalScopes
 * @prop {boolean} config.cleanHash - whether or not location.hash will be cleaned after authorization is completed.
 * Should be set to false in angular > 1.2.26 apps to prevent infinite redirect in Firefox
 * @prop {User?} user
 * @prop {string[]} config.userFields List of user data fields to be returned by auth.requestUser (default list is used in Header.HeaderHelper)
 * @prop {string[]} config.fetchCredentials
 *
 * @param {{
 *   serverUri: string,
 *   redirectUri: string?,
 *   requestCredentials: string?,
 *   clientId: string?,
 *   scope: string[]?,
 *   optionalScopes: string[]?,
 *   cleanHash: boolean?,
 *   fetchCredentials: string?,
 *   userFields: string[]?
 * }} config
 *
 */
/**
 * Extend Auth config with non-required and not pure-JS stuff
 */
Auth.DEFAULT_CONFIG = {
  ...Auth.DEFAULT_CONFIG,
  EmbeddedLoginFlow: WindowFlow,
  onBackendDown: onBackendDown
};

export { Auth, Auth as default };
