// @jetbrains/ring-ui-built/components/auth/down-notification.js
import { c } from 'react-compiler-runtime';
import 'react';
import alertService from '../alert-service/alert-service.js';
import Alert from '../alert/alert.js';
import Link from '../link/link.js';
import Group from '../group/group.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'react-dom/client';
import '../global/get-uid.js';
import '../alert/container.js';
import 'react-dom';
import 'classnames';
import '@jetbrains/icons/exception';
import '@jetbrains/icons/checkmark';
import '@jetbrains/icons/warning';
import '@jetbrains/icons/close';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../loader-inline/loader-inline.js';
import '../global/data-tests.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../_helpers/theme.js';
import '../_helpers/link.js';

var styles = {"title":"ring-down-notification-title","error":"ring-down-notification-error"};

let key = null;
function renderAlert(message, type = Alert.Type.WARNING) {
  const existingAlert = alertService.showingAlerts.filter(alert => alert.key === key)[0];
  if (!existingAlert) {
    key = alertService.addAlert(message, type, 0, {
      closeable: false
    });
  } else {
    existingAlert.message = message;
    existingAlert.type = type;
    alertService.renderAlerts();
  }
}
function Message(t0) {
  const $ = c(12);
  if ($[0] !== "f7af74b5afecec4eb25bcdac57e8252929ea2543343e0ea7e29543844f786673") {
    for (let $i = 0; $i < 12; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "f7af74b5afecec4eb25bcdac57e8252929ea2543343e0ea7e29543844f786673";
  }
  const {
    translations,
    onCheckAgain
  } = t0;
  const {
    backendIsNotAvailable,
    checkAgain,
    errorMessage
  } = translations;
  let t1;
  if ($[1] !== backendIsNotAvailable) {
    t1 = /*#__PURE__*/jsx(Group, {
      children: /*#__PURE__*/jsx("div", {
        className: styles.title,
        children: backendIsNotAvailable
      })
    });
    $[1] = backendIsNotAvailable;
    $[2] = t1;
  } else {
    t1 = $[2];
  }
  let t2;
  if ($[3] !== errorMessage) {
    t2 = /*#__PURE__*/jsxs("span", {
      className: styles.error,
      children: [errorMessage, " "]
    });
    $[3] = errorMessage;
    $[4] = t2;
  } else {
    t2 = $[4];
  }
  let t3;
  if ($[5] !== checkAgain || $[6] !== onCheckAgain) {
    t3 = /*#__PURE__*/jsx(Link, {
      onClick: onCheckAgain,
      "data-test": "check-again",
      children: checkAgain
    });
    $[5] = checkAgain;
    $[6] = onCheckAgain;
    $[7] = t3;
  } else {
    t3 = $[7];
  }
  let t4;
  if ($[8] !== t1 || $[9] !== t2 || $[10] !== t3) {
    t4 = /*#__PURE__*/jsxs("div", {
      "data-test": "ring-backend-down-notification",
      children: [t1, t2, t3]
    });
    $[8] = t1;
    $[9] = t2;
    $[10] = t3;
    $[11] = t4;
  } else {
    t4 = $[11];
  }
  return t4;
}
function onBackendDown({
  onCheckAgain,
  translations
}) {
  async function checkAgainWithoutClosing(e) {
    // Alert has weird behaviour of handling clicks by "a" tags
    e.stopPropagation();
    try {
      renderAlert('Connecting...', Alert.Type.LOADING);
      await onCheckAgain();
    } catch (err) {
      renderAlert(/*#__PURE__*/jsx(Message, {
        translations: translations,
        onCheckAgain: checkAgainWithoutClosing
      }));
    }
  }
  renderAlert(/*#__PURE__*/jsx(Message, {
    translations: translations,
    onCheckAgain: checkAgainWithoutClosing
  }));
  return () => alertService.remove(key);
}

export { onBackendDown as default };
