// @jetbrains/ring-ui-built/components/auth/iframe-flow.js
import showAuthDialog from '../login-dialog/service.js';
import Link from '../link/link.js';
import AuthResponseParser from './response-parser.js';
import { jsx } from 'react/jsx-runtime';
import 'react-dom/client';
import '../global/controls-height.js';
import 'react';
import 'util-deprecate';
import '../global/configuration.js';
import '../login-dialog/login-dialog.js';
import '../dialog/dialog.js';
import 'react-dom';
import 'classnames';
import '@jetbrains/icons/close';
import '../island/island.js';
import '../global/data-tests.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import '../_helpers/island.js';
import '../island/header.js';
import 'react-compiler-runtime';
import '../island/content.js';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../_helpers/button.classes.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '../loader-screen/loader-screen.js';
import '../loader/loader.js';
import '../_helpers/loader-core.js';
import './background-flow.js';
import '../global/url.js';
import '../_helpers/link.js';

class IFrameFlow {
  hideDialog = null;
  _requestBuilder;
  _storage;
  _translations;
  reject;
  _promise;
  _loginWindow;
  constructor(requestBuilder, storage, translations) {
    this._requestBuilder = requestBuilder;
    this._storage = storage;
    this._translations = translations;
    this._reset();
  }
  /**
   * Initiates authorization in window
   */
  async _load() {
    const authRequest = await this._requestBuilder.prepareAuthRequest({
      request_credentials: 'required',
      auth_mode: 'bypass_to_login'
    }, {
      nonRedirect: false
    });
    const renderFallbackLink = () => /*#__PURE__*/jsx(Link, {
      href: authRequest.url,
      target: "_self",
      children: this._translations.nothingHappensLink
    });
    return new Promise((resolve, reject) => {
      this.hideDialog = showAuthDialog({
        url: authRequest.url,
        loader: true,
        onCancel: () => {
          // eslint-disable-next-line no-use-before-define
          cleanUp();
          this.stop();
        },
        renderFallbackLink
      });
      this.reject = reject;
      const removeTokenListener = this._storage.onTokenChange(token => {
        if (token) {
          // eslint-disable-next-line no-use-before-define
          cleanUp();
          resolve(token.accessToken);
        }
      });
      const removeStateListener = this._storage.onStateChange(authRequest.stateId, state => {
        if (state && state.error) {
          // eslint-disable-next-line no-use-before-define
          cleanUp();
          reject(new AuthResponseParser.AuthError(state));
        }
      });
      const cleanUp = () => {
        this.hideDialog?.();
        removeStateListener();
        removeTokenListener();
      };
    });
  }
  _reset = () => {
    this._promise = null;
    this.hideDialog = null;
  };
  stop() {
    if (this.hideDialog) {
      this.hideDialog();
    }
    if (this.reject) {
      this.reject('Login form closed');
    }
    this._reset();
  }
  authorize() {
    if (this._promise && this._loginWindow && !this._loginWindow.closed) {
      return this._promise;
    }
    this._promise = this._load();
    this._promise.then(this._reset, this._reset);
    return this._promise;
  }
}

export { IFrameFlow as default };
