// @jetbrains/ring-ui-built/components/auth/request-builder.js
import uuid from 'simply-uuid';
import { encodeURL } from '../global/url.js';

class AuthRequestBuilder {
  config;
  storage;
  /**
   * @param {{
   *   authorization: string,
   *   redirectUri: string?,
   *   requestCredentials: string?,
   *   clientId: string?,
   *   scopes: string[]
   * }} config
   * @param {AuthStorage} storage
   */
  constructor(config, storage) {
    this.config = config;
    this.storage = storage;
  }
  /**
   * @return {string} random string used for state
   */
  static _uuid = uuid.generate;
  /**
   * Save state and build an auth server redirect URL.
   *
   * @param {object=} extraParams additional query parameters for auth request
   * @param {object=} extraState additional state parameters to save
   * @return {Promise.<string>} promise that is resolved to authURL
   */
  async prepareAuthRequest(extraParams, extraState) {
    // eslint-disable-next-line no-underscore-dangle
    const stateId = AuthRequestBuilder._uuid();
    const scopes = this.config.scopes.map(scope => encodeURIComponent(scope));
    const request = Object.assign({
      response_type: 'token',
      state: stateId,
      redirect_uri: this.config.redirectUri,
      request_credentials: this.config.requestCredentials,
      client_id: this.config.clientId,
      scope: scopes.join(' ')
    }, extraParams || {});
    const authURL = encodeURL(this.config.authorization, request);
    const state = Object.assign({
      restoreLocation: window.location.href,
      scopes: this.config.scopes
    }, extraState || {});
    await this._saveState(stateId, state);
    return {
      url: authURL,
      stateId
    };
  }
  /**
   * @param {string} id
   * @param {StoredState} storedState
   * @return {Promise}
   * @private
   */
  _saveState(id, storedState) {
    return this.storage?.saveState(id, storedState);
  }
}

export { AuthRequestBuilder as default };
