// @jetbrains/ring-ui-built/components/text/text.js
import { Component } from 'react';
import classNames from 'classnames';
import { jsx } from 'react/jsx-runtime';

var styles = {"text":"ring-text-text","sizeS":"ring-text-sizeS","sizeM":"ring-text-sizeM","sizeL":"ring-text-sizeL","info":"ring-text-info","bold":"ring-text-bold"};

const TextSizes = {
  S: 's',
  M: 'm',
  L: 'l'
};
/**
 * @name Text
 */
class Text extends Component {
  static defaultProps = {
    size: TextSizes.M
  };
  static Size = TextSizes;
  render() {
    const {
      children,
      className,
      info,
      size,
      bold,
      ...restProps
    } = this.props;
    const classes = classNames(styles.text, className, {
      [styles.info]: info,
      [styles.bold]: bold,
      [styles.sizeS]: size === Text.Size.S,
      [styles.sizeM]: size === Text.Size.M,
      [styles.sizeL]: size === Text.Size.L
    });
    return /*#__PURE__*/jsx("span", {
      className: classes,
      ...restProps,
      children: children
    });
  }
}

export { Text as default };
