// @jetbrains/ring-ui-built/components/user-agreement/service.js
import { createRoot } from 'react-dom/client';
import LocalStorage from '../storage/storage-local.js';
import alertService from '../alert-service/alert-service.js';
import Link from '../link/link.js';
import Alert from '../alert/alert.js';
import Group from '../group/group.js';
import { ControlsHeightContext, getGlobalControlsHeight } from '../global/controls-height.js';
import UserAgreement from './user-agreement.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import '../global/get-uid.js';
import '../alert/container.js';
import 'react';
import 'react-dom';
import 'classnames';
import '../global/data-tests.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '@jetbrains/icons/exception';
import '@jetbrains/icons/checkmark';
import '@jetbrains/icons/warning';
import '@jetbrains/icons/close';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../loader-inline/loader-inline.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../_helpers/button.classes.js';
import '../_helpers/theme.js';
import '../global/configuration.js';
import '../dialog/dialog.js';
import '../island/island.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import '../_helpers/island.js';
import '../island/header.js';
import '../island/content.js';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '../panel/panel.js';
import '../i18n/i18n-context.js';
import '../i18n/i18n.js';

const GUEST_SESSION_KEY = 'end-user-agreement-consent';
const ONE_HOUR = 60 * 60 * 1000; // eslint-disable-line no-magic-numbers
const storageKey = 'userAgreementKey';
const showMessage = 'userAgreementShow';
const hideMessage = 'userAgreementHide';
const DEFAULT_CONSENT = {
  accepted: false,
  majorVersion: 0,
  minorVersion: 0
};
const DEFAULT_AGREEMENT = {
  enabled: false,
  majorVersion: 0,
  minorVersion: 0,
  requiredForREST: false,
  content: ''
};
function ignorePostponedCatchReason(reason) {
  if (reason === 'Postponed') {
    return;
  }
  throw reason;
}
class UserAgreementService {
  config;
  constructor(config) {
    if (!config) {
      throw new Error('Please pass a config to UserAgreementService');
    }
    if (!config.getUserAgreement) {
      throw new Error('Please pass a "getUserAgreement" option to UserAgreementService');
    }
    if (!config.getUserConsent) {
      throw new Error('Please pass a "getUserConsent" option to UserAgreementService');
    }
    if (!config.setUserConsent) {
      throw new Error('Please pass a "setUserConsent" option to UserAgreementService');
    }
    this.config = config;
    this.interval = config.interval || this.interval;
  }
  _dialogPromise = null;
  _alertPromise = null;
  tabId = Math.random();
  interval = ONE_HOUR;
  container = document.createElement('div');
  reactRoot = createRoot(this.container);
  storage = new LocalStorage();
  checkingPromise = null;
  guest = false;
  userAgreement = DEFAULT_AGREEMENT;
  userConsent = DEFAULT_CONSENT;
  intervalId;
  startChecking = () => {
    this.intervalId = window.setInterval(() => this.checkConsentAndShowDialog().catch(ignorePostponedCatchReason), this.interval);
    window.addEventListener('storage', this.onStorageEvent);
    this.checkConsentAndShowDialog().catch(ignorePostponedCatchReason);
  };
  stopChecking = () => {
    clearInterval(this.intervalId);
    window.removeEventListener('storage', this.onStorageEvent);
    this.hideDialog();
  };
  onStorageEvent = event => {
    if (event.key === storageKey && event.newValue !== null) {
      const {
        tabId,
        command
      } = JSON.parse(event.newValue);
      if (tabId !== this.tabId) {
        if (command === showMessage) {
          this.checkConsentAndShowDialog(true);
        } else if (command === hideMessage) {
          this.hideDialogAndAlert(true);
        }
      }
    }
  };
  _notifyAboutShowing = () => {
    localStorage.setItem(storageKey, JSON.stringify({
      command: showMessage,
      tabId: this.tabId
    }));
  };
  _notifyAboutHiding = () => {
    localStorage.setItem(storageKey, JSON.stringify({
      command: hideMessage,
      tabId: this.tabId
    }));
  };
  getUserAgreement = async () => {
    this.userAgreement = (await this.config.getUserAgreement()) || DEFAULT_AGREEMENT;
    return this.userAgreement;
  };
  getUserConsent = async () => {
    const {
      guest,
      endUserAgreementConsent
    } = await this.config.getUserConsent();
    this.guest = guest;
    if (guest) {
      this.userConsent = (await this.storage.get(GUEST_SESSION_KEY)) || this.userConsent;
    } else {
      this.userConsent = endUserAgreementConsent || this.userConsent;
    }
    return this.userConsent;
  };
  checkConsentAndShowDialog = async withoutNotifications => {
    if (await this.checkConsent()) {
      return this.hideDialogAndAlert(withoutNotifications);
    }
    return this.showDialogOrAlert(withoutNotifications);
  };
  checkConsent = async () => {
    if (!this.checkingPromise) {
      this.checkingPromise = Promise.all([this.getUserAgreement(), this.getUserConsent()]);
    }
    const [userAgreement, userConsent] = await this.checkingPromise;
    this.checkingPromise = null;
    const {
      enabled,
      majorVersion: actualVersion
    } = userAgreement;
    const {
      accepted,
      majorVersion: acceptedVersion
    } = userConsent;
    return !enabled || accepted && actualVersion === acceptedVersion;
  };
  alertKey;
  showAlert = withoutNotifications => {
    if (this._alertPromise) {
      return this._alertPromise;
    }
    this._alertPromise = new Promise((resolve, reject) => {
      const {
        userAgreement,
        reviewNow,
        remindLater
      } = this.config.translations || {};
      const onRemind = () => {
        this.hideDialogAndAlert(withoutNotifications);
        reject('Postponed');
      };
      const onReview = async () => {
        await this.showDialog(true, false, {
          onRemindLater: onRemind
        });
        this.hideAlert(withoutNotifications);
        resolve();
      };
      const message = /*#__PURE__*/jsxs(Group, {
        children: [/*#__PURE__*/jsx("span", {
          children: userAgreement || 'User Agreement'
        }), /*#__PURE__*/jsx(Link, {
          onClick: onReview,
          "data-test": "review",
          children: reviewNow || 'Review now'
        }), /*#__PURE__*/jsx(Link, {
          onClick: onRemind,
          "data-test": "later",
          children: remindLater || 'Remind me later'
        })]
      });
      this.alertKey = alertService.addAlert(message, Alert.Type.WARNING, 0, {
        closeable: false
      });
    });
    if (!withoutNotifications) {
      this._notifyAboutShowing();
    }
    return this._alertPromise;
  };
  hideAlert = withoutNotifications => {
    const {
      onRemindLater
    } = this.config;
    alertService.remove(this.alertKey);
    this.alertKey = null;
    this._alertPromise = null;
    if (onRemindLater) {
      onRemindLater();
    }
    if (!withoutNotifications) {
      this._notifyAboutHiding();
    }
  };
  showDialog = (withoutNotifications, preview = false, restOptions) => {
    const {
      translations,
      onDialogShow
    } = this.config;
    const {
      content
    } = this.userAgreement;
    const show = true;
    if (!this._dialogPromise) {
      this._dialogPromise = new Promise((resolve, reject) => {
        const onAccept = async () => {
          await this.onAccept();
          resolve();
        };
        const onDecline = async () => {
          await this.onDecline();
          reject();
        };
        const onClose = this.hideDialogAndAlert;
        const props = {
          children: content,
          show,
          onAccept,
          onDecline,
          onClose,
          translations,
          preview,
          ...restOptions
        };
        this.reactRoot.render(/*#__PURE__*/jsx(ControlsHeightContext.Provider, {
          value: getGlobalControlsHeight(),
          children: /*#__PURE__*/jsx(UserAgreement, {
            ...props
          })
        }));
        if (onDialogShow) {
          onDialogShow();
        }
      });
      if (!withoutNotifications) {
        this._notifyAboutShowing();
      }
    }
    return this._dialogPromise;
  };
  hideDialog = withoutNotifications => {
    const {
      onDialogHide
    } = this.config;
    this.reactRoot.render(null);
    if (onDialogHide) {
      onDialogHide();
      this._dialogPromise = null;
      if (!withoutNotifications) {
        this._notifyAboutHiding();
      }
    }
  };
  showDialogOrAlert = (withoutNotifications, preview, restOptions) => {
    if (this.guest && !this.userAgreement.requiredForREST) {
      return this.showAlert(withoutNotifications);
    }
    return this.showDialog(withoutNotifications, preview, restOptions);
  };
  hideDialogAndAlert = withoutNotifications => {
    this.hideAlert(withoutNotifications);
    this.hideDialog(withoutNotifications);
  };
  onAccept = async () => {
    const {
      setUserConsent,
      onAccept
    } = this.config;
    if (this.guest) {
      const {
        majorVersion,
        minorVersion
      } = this.userAgreement;
      this.userConsent = {
        majorVersion,
        minorVersion,
        accepted: true
      };
      await this.storage.set(GUEST_SESSION_KEY, this.userConsent);
    } else {
      this.userConsent = await setUserConsent();
    }
    this.hideDialog();
    if (onAccept) {
      onAccept();
    }
  };
  onDecline = () => {
    const {
      onDecline
    } = this.config;
    this.hideDialog();
    if (onDecline) {
      onDecline();
    }
  };
}

export { UserAgreementService as default, hideMessage, showMessage };
