// @jetbrains/ring-ui-built/components/user-agreement/user-agreement.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import Dialog from '../dialog/dialog.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import '../_helpers/island.js';
import HeaderWrapper from '../island/header.js';
import ContentWrapper from '../island/content.js';
import Panel from '../panel/panel.js';
import { Button } from '../button/button.js';
import { I18nContext } from '../i18n/i18n-context.js';
import 'react-dom';
import '@jetbrains/icons/close';
import '../island/island.js';
import '../global/data-tests.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import 'react-compiler-runtime';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/dom.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../_helpers/button.classes.js';
import '../i18n/i18n.js';

var style = {"extra-small-screen-media":"(max-width: 639px)","agreementDialog":"ring-user-agreement-agreementDialog","dialogContent":"ring-user-agreement-dialogContent","suggestion":"ring-user-agreement-suggestion","remindLaterButton":"ring-user-agreement-remindLaterButton"};

function noop() {}
/**
 * A component that displays a user agreement dialog.
 */
class UserAgreement extends PureComponent {
  static defaultProps = {
    show: false,
    onAccept: noop,
    onDecline: noop,
    onClose: noop
  };
  state = {
    scrolledDown: false
  };
  onScrollToBottom = () => this.setState({
    scrolledDown: true
  });
  render() {
    const {
      scrolledDown
    } = this.state;
    const {
      translations,
      onAccept,
      onDecline,
      onClose,
      onRemindLater,
      children,
      show,
      preview,
      className
    } = this.props;
    return /*#__PURE__*/jsx(I18nContext.Consumer, {
      children: ({
        translate
      }) => {
        var _translations$userAgr, _translations$userAgr2, _translations$scrollT, _translations$accept, _translations$decline, _translations$scrollT2, _translations$remindL, _translations$close;
        return /*#__PURE__*/jsxs(Dialog, {
          label: (_translations$userAgr = translations?.userAgreement) !== null && _translations$userAgr !== void 0 ? _translations$userAgr : translate('userAgreement'),
          show: show,
          className: classNames(style.agreementDialog, className),
          contentClassName: style.dialogContent,
          trapFocus: true,
          autoFocusFirst: false,
          "data-test": "user-agreement",
          children: [/*#__PURE__*/jsx(HeaderWrapper, {
            children: (_translations$userAgr2 = translations?.userAgreement) !== null && _translations$userAgr2 !== void 0 ? _translations$userAgr2 : translate('userAgreement')
          }), /*#__PURE__*/jsx(ContentWrapper, {
            tabIndex: 0,
            fade: true,
            onScrollToBottom: this.onScrollToBottom,
            children: children
          }), !preview && /*#__PURE__*/jsxs(Panel, {
            children: [onRemindLater && !scrolledDown && /*#__PURE__*/jsx("div", {
              className: style.suggestion,
              children: (_translations$scrollT = translations?.scrollToAccept) !== null && _translations$scrollT !== void 0 ? _translations$scrollT : translate('scrollToAccept')
            }), /*#__PURE__*/jsx(Button, {
              primary: true,
              disabled: !scrolledDown,
              onClick: onAccept,
              "data-test": "accept",
              children: (_translations$accept = translations?.accept) !== null && _translations$accept !== void 0 ? _translations$accept : translate('accept')
            }), /*#__PURE__*/jsx(Button, {
              onClick: onDecline,
              autoFocus: true,
              "data-test": "decline",
              children: (_translations$decline = translations?.decline) !== null && _translations$decline !== void 0 ? _translations$decline : translate('decline')
            }), !onRemindLater && !scrolledDown && /*#__PURE__*/jsx("span", {
              className: style.suggestion,
              children: (_translations$scrollT2 = translations?.scrollToAccept) !== null && _translations$scrollT2 !== void 0 ? _translations$scrollT2 : translate('scrollToAccept')
            }), onRemindLater && /*#__PURE__*/jsx(Button, {
              className: style.remindLaterButton,
              onClick: onRemindLater,
              "data-test": "later",
              children: (_translations$remindL = translations?.remindLater) !== null && _translations$remindL !== void 0 ? _translations$remindL : translate('remindLater')
            })]
          }), preview && /*#__PURE__*/jsx(Panel, {
            children: /*#__PURE__*/jsx(Button, {
              onClick: onClose,
              autoFocus: true,
              "data-test": "close",
              children: (_translations$close = translations?.close) !== null && _translations$close !== void 0 ? _translations$close : translate('close')
            })
          })]
        });
      }
    });
  }
}

export { UserAgreement as default };
