// @jetbrains/ring-ui-built/components/avatar-stack/avatar-stack.js
import { c } from 'react-compiler-runtime';
import { useState, Children } from 'react';
import classNames from 'classnames';
import DropdownMenu from '../dropdown-menu/dropdown-menu.js';
import Avatar from '../avatar/avatar.js';
import { f as fontSizes } from '../_helpers/avatar-info.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { Size } from '../avatar/avatar-size.js';
import '../list/list.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'util-deprecate';
import 'memoize-one';
import '../global/data-tests.js';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/memoize.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../dropdown/dropdown.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../popup-menu/popup-menu.js';
import '../popup/popup.js';
import 'react-dom';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';

var styles = {"avatarStack":"ring-avatar-stack-avatarStack","extraText":"ring-avatar-stack-extraText","item":"ring-avatar-stack-item","size20":"ring-avatar-stack-size20","size24":"ring-avatar-stack-size24","size28":"ring-avatar-stack-size28","size32":"ring-avatar-stack-size32","size40":"ring-avatar-stack-size40","size56":"ring-avatar-stack-size56","hovered":"ring-avatar-stack-hovered","extra":"ring-avatar-stack-extra ring-avatar-stack-item","extraButton":"ring-avatar-stack-extraButton ring-global-resetButton"};

function AvatarStack(t0) {
  const $ = c(28);
  if ($[0] !== "935a2575bcba5af6c406972a71daa0b8850b8f7634076664ddbd547b5e2fc6a6") {
    for (let $i = 0; $i < 28; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "935a2575bcba5af6c406972a71daa0b8850b8f7634076664ddbd547b5e2fc6a6";
  }
  let children;
  let className;
  let dropdownMenuProps;
  let extraItems;
  let restProps;
  let t1;
  if ($[1] !== t0) {
    ({
      children,
      className,
      size: t1,
      extraItems,
      dropdownMenuProps,
      ...restProps
    } = t0);
    $[1] = t0;
    $[2] = children;
    $[3] = className;
    $[4] = dropdownMenuProps;
    $[5] = extraItems;
    $[6] = restProps;
    $[7] = t1;
  } else {
    children = $[2];
    className = $[3];
    dropdownMenuProps = $[4];
    extraItems = $[5];
    restProps = $[6];
    t1 = $[7];
  }
  const size = t1 === undefined ? Size.Size20 : t1;
  const [dropdownOpen, setDropdownOpen] = useState(false);
  const t2 = styles[`size${size}`];
  let t3;
  if ($[8] !== className || $[9] !== dropdownOpen || $[10] !== t2) {
    t3 = classNames(styles.avatarStack, className, t2, {
      [styles.hovered]: dropdownOpen
    });
    $[8] = className;
    $[9] = dropdownOpen;
    $[10] = t2;
    $[11] = t3;
  } else {
    t3 = $[11];
  }
  let t4;
  if ($[12] !== restProps.style || $[13] !== size) {
    t4 = {
      height: size,
      ...restProps.style
    };
    $[12] = restProps.style;
    $[13] = size;
    $[14] = t4;
  } else {
    t4 = $[14];
  }
  let t5;
  if ($[15] !== children) {
    t5 = Children.map(children, _temp);
    $[15] = children;
    $[16] = t5;
  } else {
    t5 = $[16];
  }
  let t6;
  if ($[17] !== children || $[18] !== dropdownMenuProps || $[19] !== extraItems || $[20] !== size) {
    t6 = extraItems?.length ? /*#__PURE__*/jsx(DropdownMenu, {
      hoverMode: true,
      hoverShowTimeOut: 10,
      onShow: () => setDropdownOpen(true),
      onHide: () => setDropdownOpen(false),
      className: styles.extra,
      style: {
        width: size,
        height: size,
        "--ring-avatar-stack-index": Children.count(children),
        fontSize: fontSizes[size]
      },
      anchor: /*#__PURE__*/jsx("button", {
        type: "button",
        className: styles.extraButton,
        children: /*#__PURE__*/jsx(Avatar, {
          size: size,
          info: /*#__PURE__*/jsx("span", {
            className: styles.extraText,
            children: `+${extraItems.length}`
          })
        })
      }),
      data: extraItems,
      menuProps: {
        offset: 4,
        ...dropdownMenuProps?.menuProps
      },
      ...dropdownMenuProps
    }) : null;
    $[17] = children;
    $[18] = dropdownMenuProps;
    $[19] = extraItems;
    $[20] = size;
    $[21] = t6;
  } else {
    t6 = $[21];
  }
  let t7;
  if ($[22] !== restProps || $[23] !== t3 || $[24] !== t4 || $[25] !== t5 || $[26] !== t6) {
    t7 = /*#__PURE__*/jsxs("div", {
      className: t3,
      ...restProps,
      style: t4,
      children: [t5, t6]
    });
    $[22] = restProps;
    $[23] = t3;
    $[24] = t4;
    $[25] = t5;
    $[26] = t6;
    $[27] = t7;
  } else {
    t7 = $[27];
  }
  return t7;
}
function _temp(child, index) {
  return /*#__PURE__*/jsx("div", {
    className: styles.item,
    style: {
      "--ring-avatar-stack-index": index
    },
    children: child
  });
}

export { AvatarStack as default };
