// @jetbrains/ring-ui-built/components/checkbox/checkbox.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import checkmarkIcon from '@jetbrains/icons/checkmark-12px';
import minusIcon from '@jetbrains/icons/remove-12px';
import Icon from '../icon/icon.js';
import { createComposedRef } from '../global/compose-refs.js';
import ControlHelp from '../control-help/control-help.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import 'memoize-one';

var styles = {"checkbox":"ring-checkbox-checkbox","cell":"ring-checkbox-cell","icon":"ring-checkbox-icon","check":"ring-checkbox-check ring-checkbox-icon","minus":"ring-checkbox-minus ring-checkbox-icon","input":"ring-checkbox-input","cellWrapper":"ring-checkbox-cellWrapper","focus":"ring-checkbox-focus","label":"ring-checkbox-label"};

class Checkbox extends PureComponent {
  static defaultProps = {
    indeterminate: false
  };
  componentDidMount() {
    if (this.input) {
      this.input.indeterminate = this.props.indeterminate;
    }
  }
  componentDidUpdate(prevProps) {
    const {
      indeterminate
    } = this.props;
    if (this.input && indeterminate !== prevProps.indeterminate) {
      this.input.indeterminate = this.props.indeterminate;
    }
  }
  input;
  inputRef = el => {
    if (el) {
      el.indeterminate = this.props.indeterminate;
    }
    this.input = el;
  };
  composedInputRef = createComposedRef();
  render() {
    const {
      children,
      label,
      className,
      containerClassName,
      containerStyle,
      cellClassName,
      labelClassName,
      indeterminate,
      inputRef,
      help,
      ...restProps
    } = this.props;
    const classes = classNames(styles.input, className);
    const containerClasses = classNames(styles.checkbox, containerClassName);
    const cellClasses = classNames(styles.cell, cellClassName);
    const labelClasses = classNames(styles.label, labelClassName);
    return /*#__PURE__*/jsxs("label", {
      className: containerClasses,
      style: containerStyle,
      "data-test": "ring-checkbox",
      children: [/*#__PURE__*/jsx("input", {
        ...restProps,
        "data-checked": restProps.checked,
        ref: this.composedInputRef(this.inputRef, inputRef),
        type: "checkbox",
        className: classes
      }), /*#__PURE__*/jsx("div", {
        className: styles.cellWrapper,
        children: /*#__PURE__*/jsxs("span", {
          className: cellClasses,
          children: [/*#__PURE__*/jsx(Icon, {
            glyph: checkmarkIcon,
            className: styles.check
          }), /*#__PURE__*/jsx(Icon, {
            glyph: minusIcon,
            className: styles.minus
          })]
        })
      }), /*#__PURE__*/jsxs("span", {
        className: labelClasses,
        children: [label || children, help && /*#__PURE__*/jsx(ControlHelp, {
          children: help
        })]
      })]
    });
  }
}

export { Checkbox as default };
