// @jetbrains/ring-ui-built/components/pager/pager.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import chevronLeftIcon from '@jetbrains/icons/chevron-left';
import chevronRightIcon from '@jetbrains/icons/chevron-right';
import { Button } from '../button/button.js';
import ButtonGroup from '../button-group/button-group.js';
import ButtonToolbar from '../button-toolbar/button-toolbar.js';
import Select from '../select/select.js';
import memoize from '../global/memoize.js';
import Link from '../link/link.js';
import Icon from '../icon/icon.js';
import { I18nContext } from '../i18n/i18n-context.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/data-tests.js';
import '../control-label/control-label.js';
import '../control-help/control-help.js';
import '../_helpers/caption.js';
import '@jetbrains/icons/close-12px';
import 'dequal';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import '../avatar/avatar.js';
import '../global/url.js';
import '../global/dom.js';
import '../avatar/fallback-avatar.js';
import '../global/get-uid.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../popup/popup.js';
import 'react-dom';
import '../global/schedule-raf.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../list/list.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'memoize-one';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../global/compose-refs.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../_helpers/link.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../_helpers/input.js';
import '../global/rerender-hoc.js';
import '../global/fuzzy-highlight.js';
import '../select/select-popup.js';
import '@jetbrains/icons/search';
import '../loader-inline/loader-inline.js';
import '../shortcuts/shortcuts-hoc.js';
import '../tags-list/tags-list.js';
import '../tag/tag.js';
import '../caret/caret.js';
import '../text/text.js';
import '../_helpers/select-filter.js';
import '../i18n/i18n.js';

var style = {"pager":"ring-pager-pager","links":"ring-pager-links","link":"ring-pager-link","linkDisabled":"ring-pager-linkDisabled","actions":"ring-pager-actions","pageSize":"ring-pager-pageSize"};

class Pager extends PureComponent {
  static defaultProps = {
    currentPage: 1,
    pageSize: 50,
    // eslint-disable-next-line no-magic-numbers
    pageSizes: [20, 50, 100],
    visiblePagesLimit: 7,
    disablePageSizeSelector: false,
    openTotal: false,
    canLoadLastPageWithOpenTotal: false,
    loader: false,
    loaderNavigation: false,
    onPageSizeChange: () => {},
    onLoadPage: () => {}
  };
  static contextType = I18nContext;
  getSelectOptions() {
    const {
      pageSize,
      pageSizes
    } = this.props;
    const {
      translate
    } = this.context;
    const data = pageSizes.map(size => {
      var _this$props$translati;
      return {
        key: size,
        label: `${size} ${(_this$props$translati = this.props.translations?.perPage) !== null && _this$props$translati !== void 0 ? _this$props$translati : translate('perPage')}`
      };
    });
    const selected = data.find(it => it.key === pageSize);
    return {
      selected,
      data
    };
  }
  getTotalPages() {
    const {
      total,
      pageSize
    } = this.props;
    return Math.ceil(total / pageSize);
  }
  handlePageSizeChange = item => {
    if (item) {
      this.props.onPageSizeChange(item.key);
    }
  };
  handlePrevClick = () => {
    const {
      currentPage
    } = this.props;
    if (currentPage !== 1) {
      const prevPage = currentPage - 1;
      this.props.onPageChange?.(prevPage);
    }
  };
  handleNextClick = () => {
    const {
      currentPage,
      onLoadPage
    } = this.props;
    const nextPage = currentPage + 1;
    const total = this.getTotalPages();
    if (currentPage !== total) {
      this.props.onPageChange?.(nextPage);
    } else if (this.props.openTotal) {
      onLoadPage(nextPage);
    }
  };
  handlePageChange = memoize(i => event => {
    this.props.onPageChange?.(i, event);
  });
  handleLoadMore = memoize(i => () => {
    this.props.onLoadPage(i);
  });
  getButton(page, content, key, active) {
    return /*#__PURE__*/jsx(Button, {
      href: this.generateHref(page),
      active: active,
      disabled: this.props.loader && !active && !this.props.loaderNavigation,
      loader: this.props.loader && active,
      ...this.getClickProps(this.handlePageChange(page)),
      children: content
    }, key);
  }
  getClickProps(onClick) {
    const {
      hrefFunc,
      onPageChange
    } = this.props;
    if (!onPageChange) {
      return {};
    }
    if (hrefFunc) {
      return {
        onPlainLeftClick: onClick
      };
    }
    return {
      onClick
    };
  }
  getPageSizeSelector() {
    const selectOptions = this.getSelectOptions();
    return !this.props.disablePageSizeSelector && /*#__PURE__*/jsx("div", {
      "data-test": "ring-pager-page-size-selector",
      className: style.pageSize,
      children: /*#__PURE__*/jsx(Select, {
        data: selectOptions.data,
        selected: selectOptions.selected,
        onSelect: this.handlePageSizeChange,
        type: Select.Type.INLINE,
        disabled: this.props.loader
      })
    });
  }
  getPagerLinks() {
    var _this$props$translati2, _this$props$translati3;
    const {
      translate
    } = this.context;
    const prevLinkAvailable = this.props.currentPage !== 1;
    const nextLinkAvailable = this.props.openTotal || this.props.currentPage !== this.getTotalPages();
    const nextIcon = /*#__PURE__*/jsx(Icon, {
      glyph: chevronRightIcon
    }, 'icon');
    const prevIcon = /*#__PURE__*/jsx(Icon, {
      glyph: chevronLeftIcon
    }, 'icon');
    const prevText = (_this$props$translati2 = this.props.translations?.previousPage) !== null && _this$props$translati2 !== void 0 ? _this$props$translati2 : translate('previousPage');
    const nextText = (_this$props$translati3 = this.props.translations?.nextPage) !== null && _this$props$translati3 !== void 0 ? _this$props$translati3 : translate('nextPage');
    const nextLinkContent = [/*#__PURE__*/jsx("span", {
      children: nextText
    }, 'text'), nextIcon];
    const prevLinkContent = [prevIcon, /*#__PURE__*/jsx("span", {
      children: prevText
    }, 'text')];
    const prevLinkHref = this.generateHref(this.props.currentPage - 1);
    const nextLinkHref = this.generateHref(this.props.currentPage + 1);
    const disabledLinkClasses = classNames({
      [style.link]: true,
      [style.linkDisabled]: true
    });
    return /*#__PURE__*/jsxs("div", {
      className: style.links,
      children: [prevLinkAvailable && (!this.props.loader || this.props.loaderNavigation) ? /*#__PURE__*/jsx(Link, {
        href: prevLinkHref,
        className: style.link,
        ...this.getClickProps(this.handlePrevClick),
        children: prevLinkContent
      }) : /*#__PURE__*/jsxs("span", {
        className: disabledLinkClasses,
        children: [prevIcon, /*#__PURE__*/jsx("span", {
          children: prevText
        }, 'text')]
      }), nextLinkAvailable && (!this.props.loader || this.props.loaderNavigation) ? /*#__PURE__*/jsx(Link, {
        href: nextLinkHref,
        className: style.link,
        ...this.getClickProps(this.handleNextClick),
        children: nextLinkContent
      }) : /*#__PURE__*/jsxs("span", {
        className: disabledLinkClasses,
        children: [/*#__PURE__*/jsx("span", {
          children: nextText
        }, 'text'), nextIcon]
      })]
    });
  }
  generateHref(page) {
    if (this.props.hrefFunc === undefined) {
      return undefined;
    }
    const pageSize = this.props.disablePageSizeSelector ? undefined : this.props.pageSize;
    return this.props.hrefFunc(page, pageSize);
  }
  // eslint-disable-next-line complexity
  getPagerContent() {
    var _this$props$translati4, _this$props$translati5;
    const {
      currentPage,
      visiblePagesLimit
    } = this.props;
    const totalPages = this.getTotalPages();
    const {
      translate
    } = this.context;
    if (totalPages < this.props.currentPage) {
      this.props.onPageChange?.(totalPages);
    }
    let start = 1;
    let end = totalPages;
    if (totalPages >= visiblePagesLimit) {
      const leftHalfFrameSize = Math.ceil(visiblePagesLimit / 2) - 1;
      const rightHalfFrameSize = visiblePagesLimit - leftHalfFrameSize - 1;
      start = currentPage - leftHalfFrameSize;
      end = currentPage + rightHalfFrameSize;
      if (start < 1) {
        const tail = 1 - start;
        start += tail;
        end += tail;
      }
      if (end > totalPages) {
        const tail = end - totalPages;
        start -= tail;
        end -= tail;
      }
      if (start < 1) {
        start += 1 - start;
      }
    }
    const buttons = [];
    for (let i = start; i <= end; i++) {
      buttons.push(this.getButton(i, i, i, i === currentPage));
    }
    const lastPageButtonAvailable = !this.props.disableLastPageButton && end < totalPages && !this.props.openTotal || this.props.openTotal && this.props.canLoadLastPageWithOpenTotal;
    return /*#__PURE__*/jsxs("div", {
      children: [this.getPagerLinks(), /*#__PURE__*/jsxs("div", {
        className: style.actions,
        children: [/*#__PURE__*/jsxs(ButtonToolbar, {
          children: [start > 1 && this.getButton(1, (_this$props$translati4 = this.props.translations?.firstPage) !== null && _this$props$translati4 !== void 0 ? _this$props$translati4 : translate('firstPage')), /*#__PURE__*/jsxs(ButtonGroup, {
            children: [start > 1 && this.getButton(start - 1, '...'), buttons, end < totalPages && this.getButton(end + 1, '...'), end === totalPages && this.props.openTotal && /*#__PURE__*/jsx(Button, {
              href: this.generateHref(end + 1),
              disabled: this.props.loader,
              ...this.getClickProps(this.handleLoadMore(end + 1)),
              children: '...'
            })]
          }), lastPageButtonAvailable && this.getButton(this.props.openTotal ? -1 : totalPages, (_this$props$translati5 = this.props.translations?.lastPage) !== null && _this$props$translati5 !== void 0 ? _this$props$translati5 : translate('lastPage'))]
        }), this.getPageSizeSelector()]
      })]
    });
  }
  render() {
    const classes = classNames(style.pager, this.props.className);
    const shouldRenderPagerContent = this.getTotalPages() > 1 || this.props.openTotal;
    if (!shouldRenderPagerContent && this.props.disablePageSizeSelector) {
      return null;
    }
    return /*#__PURE__*/jsx("div", {
      "data-test": "ring-pager",
      className: classes,
      children: shouldRenderPagerContent ? this.getPagerContent() : this.getPageSizeSelector()
    });
  }
}

export { Pager as default };
