// @jetbrains/ring-ui-built/components/footer/footer.js
import { c } from 'react-compiler-runtime';
import { memo, isValidElement } from 'react';
import classNames from 'classnames';
import Link from '../link/link.js';
import { jsx } from 'react/jsx-runtime';
import '../global/data-tests.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';

var styles = {"footer":"ring-footer-footer ring-global-font","footerFloating":"ring-footer-footerFloating ring-footer-footer ring-global-font","column":"ring-footer-column","columnItem":"ring-footer-columnItem","columnLeft":"ring-footer-columnLeft ring-footer-column","columnCenter":"ring-footer-columnCenter ring-footer-column","line":"ring-footer-line","columnRight":"ring-footer-columnRight ring-footer-column","lineCenter":"ring-footer-lineCenter"};

const FooterColumn = /*#__PURE__*/memo(function FooterColumn(t0) {
  const $ = c(10);
  if ($[0] !== "af9372db3a745385e563b057c762368495083e0451baa8ec82b5600f64499a77") {
    for (let $i = 0; $i < 10; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "af9372db3a745385e563b057c762368495083e0451baa8ec82b5600f64499a77";
  }
  const {
    position,
    children
  } = t0;
  const t1 = position === "left";
  const t2 = position === "center";
  const t3 = position === "right";
  let t4;
  if ($[1] !== t1 || $[2] !== t2 || $[3] !== t3) {
    t4 = classNames({
      [styles.columnLeft]: t1,
      [styles.columnCenter]: t2,
      [styles.columnRight]: t3
    });
    $[1] = t1;
    $[2] = t2;
    $[3] = t3;
    $[4] = t4;
  } else {
    t4 = $[4];
  }
  const classes = t4;
  let t5;
  if ($[5] !== children) {
    t5 = /*#__PURE__*/jsx("ul", {
      className: styles.columnItem,
      children: children
    });
    $[5] = children;
    $[6] = t5;
  } else {
    t5 = $[6];
  }
  let t6;
  if ($[7] !== classes || $[8] !== t5) {
    t6 = /*#__PURE__*/jsx("div", {
      className: classes,
      children: t5
    });
    $[7] = classes;
    $[8] = t5;
    $[9] = t6;
  } else {
    t6 = $[9];
  }
  return t6;
});
/**
 * Return copyright string
 * @param year {int}
 * @mockReturnValue {string}
 */
function copyright(year) {
  const currentYear = new Date().getUTCFullYear();
  const ndash = '–';
  let ret = 'Copyright © ';
  if (year >= currentYear) {
    ret += year;
  } else {
    ret += year + ndash + currentYear;
  }
  return ret;
}
/**
 * @constructor
 * @extends {ReactComponent}
 */
const FooterLine = /*#__PURE__*/memo(function FooterLine(props) {
  const $ = c(8);
  if ($[0] !== "af9372db3a745385e563b057c762368495083e0451baa8ec82b5600f64499a77") {
    for (let $i = 0; $i < 8; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "af9372db3a745385e563b057c762368495083e0451baa8ec82b5600f64499a77";
  }
  let t0;
  if ($[1] !== props.item) {
    t0 = Array.isArray(props.item) ? props.item : [props.item];
    $[1] = props.item;
    $[2] = t0;
  } else {
    t0 = $[2];
  }
  const items = t0;
  let t1;
  if ($[3] === Symbol.for("react.memo_cache_sentinel")) {
    t1 = function renderItem(item) {
      if (/*#__PURE__*/isValidElement(item) || typeof item !== "object") {
        return item;
      }
      const element = (item.copyright ? copyright(item.copyright) : "") + (item.label || item);
      if (item.url) {
        return /*#__PURE__*/jsx(Link, {
          href: item.url,
          target: item.target,
          title: item.title,
          children: element
        }, item.url + item.title);
      }
      return element;
    };
    $[3] = t1;
  } else {
    t1 = $[3];
  }
  const renderItem = t1;
  let t2;
  if ($[4] !== items) {
    t2 = items.map(renderItem);
    $[4] = items;
    $[5] = t2;
  } else {
    t2 = $[5];
  }
  let t3;
  if ($[6] !== t2) {
    t3 = /*#__PURE__*/jsx("li", {
      className: styles.line,
      children: t2
    });
    $[6] = t2;
    $[7] = t3;
  } else {
    t3 = $[7];
  }
  return t3;
});
const Footer = /*#__PURE__*/memo(function Footer(t0) {
  const $ = c(18);
  if ($[0] !== "af9372db3a745385e563b057c762368495083e0451baa8ec82b5600f64499a77") {
    for (let $i = 0; $i < 18; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "af9372db3a745385e563b057c762368495083e0451baa8ec82b5600f64499a77";
  }
  const {
    floating,
    className,
    left,
    center,
    right
  } = t0;
  let t1;
  if ($[1] === Symbol.for("react.memo_cache_sentinel")) {
    t1 = function content(elements, position) {
      if (!elements) {
        return false;
      }
      return /*#__PURE__*/jsx(FooterColumn, {
        position: position,
        children: elements.map(_temp)
      }, position);
    };
    $[1] = t1;
  } else {
    t1 = $[1];
  }
  const content = t1;
  let t2;
  if ($[2] !== className || $[3] !== floating) {
    t2 = classNames(styles.footer, className, {
      [styles.footerFloating]: floating
    });
    $[2] = className;
    $[3] = floating;
    $[4] = t2;
  } else {
    t2 = $[4];
  }
  const classes = t2;
  let t3;
  if ($[5] !== left) {
    t3 = content(left, "left");
    $[5] = left;
    $[6] = t3;
  } else {
    t3 = $[6];
  }
  let t4;
  if ($[7] !== center) {
    t4 = content(center, "center");
    $[7] = center;
    $[8] = t4;
  } else {
    t4 = $[8];
  }
  let t5;
  if ($[9] !== right) {
    t5 = content(right, "right");
    $[9] = right;
    $[10] = t5;
  } else {
    t5 = $[10];
  }
  let t6;
  if ($[11] !== t3 || $[12] !== t4 || $[13] !== t5) {
    t6 = [t3, t4, t5];
    $[11] = t3;
    $[12] = t4;
    $[13] = t5;
    $[14] = t6;
  } else {
    t6 = $[14];
  }
  let t7;
  if ($[15] !== classes || $[16] !== t6) {
    t7 = /*#__PURE__*/jsx("footer", {
      className: classes,
      "data-test": "ring-footer",
      children: t6
    });
    $[15] = classes;
    $[16] = t6;
    $[17] = t7;
  } else {
    t7 = $[17];
  }
  return t7;
});
function _temp(item, idx) {
  return /*#__PURE__*/jsx(FooterLine, {
    item: item
  }, idx);
}

export { copyright, Footer as default };
