// @jetbrains/ring-ui-built/components/radio/radio-item.js
import { c } from 'react-compiler-runtime';
import { forwardRef, createContext, Component } from 'react';
import classNames from 'classnames';
import getUID from '../global/get-uid.js';
import ControlHelp from '../control-help/control-help.js';
import { jsx, jsxs } from 'react/jsx-runtime';

var styles = {"radio":"ring-radio-radio","circle":"ring-radio-circle","input":"ring-radio-input","focus":"ring-radio-focus","label":"ring-radio-label"};

const RadioContext = /*#__PURE__*/createContext({});
class RadioItemInner extends Component {
  uid = getUID('ring-radio-item-');
  input;
  inputRef = el => {
    this.input = el;
  };
  label;
  labelRef = el => {
    this.label = el;
  };
  render() {
    const {
      className,
      children,
      help,
      ...restProps
    } = this.props;
    const classes = classNames(styles.radio, className);
    return /*#__PURE__*/jsxs("label", {
      ref: this.labelRef,
      className: classes,
      htmlFor: this.uid,
      children: [/*#__PURE__*/jsx("input", {
        id: this.uid,
        ...restProps,
        ref: this.inputRef,
        className: styles.input,
        type: "radio"
      }), /*#__PURE__*/jsx("span", {
        className: styles.circle
      }), /*#__PURE__*/jsxs("span", {
        className: styles.label,
        children: [children, help && /*#__PURE__*/jsx(ControlHelp, {
          children: help
        })]
      })]
    });
  }
}
const RadioItem = /*#__PURE__*/forwardRef(function RadioItem(props, ref) {
  const $ = c(4);
  if ($[0] !== "3188350a888d4041d2a02a273f83e84f8d42c8d785cec28bc6b4e5db93df5501") {
    for (let $i = 0; $i < 4; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "3188350a888d4041d2a02a273f83e84f8d42c8d785cec28bc6b4e5db93df5501";
  }
  let t0;
  if ($[1] !== props || $[2] !== ref) {
    t0 = /*#__PURE__*/jsx(RadioContext.Consumer, {
      children: t1 => {
        const {
          value,
          onChange,
          ...restContext
        } = t1;
        return /*#__PURE__*/jsx(RadioItemInner, {
          ref: ref,
          ...restContext,
          checked: value !== undefined ? value === props.value : undefined,
          onChange: onChange ? () => onChange(props.value) : undefined,
          ...props
        });
      }
    });
    $[1] = props;
    $[2] = ref;
    $[3] = t0;
  } else {
    t0 = $[3];
  }
  return t0;
});

export { RadioContext, RadioItemInner, RadioItem as default };
