// @jetbrains/ring-ui-built/components/data-list/data-list.js
import { Component, PureComponent } from 'react';
import classNames from 'classnames';
import focusSensorHOC from '../global/focus-sensor-hoc.js';
import selectionShortcutsHOC from '../table/selection-shortcuts-hoc.js';
import disableHoverHOC from '../table/disable-hover-hoc.js';
import getUID from '../global/get-uid.js';
import Shortcuts from '../shortcuts/shortcuts.js';
import Loader from '../loader/loader.js';
import Item, { moreLessButtonStates } from './item.js';
import { s as styles } from '../_helpers/title.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import '../global/compose-refs.js';
import 'memoize-one';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../_helpers/loader-core.js';
import '../global/dom.js';
import '@jetbrains/icons/chevron-right';
import '@jetbrains/icons/chevron-down';
import '../link/link.js';
import '../global/data-tests.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '../text/text.js';
import '../loader-inline/loader-inline.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../control-help/control-help.js';

class DataList extends PureComponent {
  static defaultProps = {
    loading: false,
    onItemMoreLess: () => {},
    itemMoreLessState: () => moreLessButtonStates.UNUSED,
    remoteSelection: false
  };
  componentDidUpdate(prevProps) {
    const {
      data,
      selection,
      onSelect,
      selectable
    } = this.props;
    if (data !== prevProps.data && !prevProps.remoteSelection) {
      onSelect(selection.cloneWith({
        data
      }));
    }
    if (!selectable && prevProps.selectable) {
      onSelect(selection.resetSelection());
    }
  }
  shortcutsScope = getUID('ring-data-list-');
  onItemFocus = item => {
    const {
      selection,
      onSelect
    } = this.props;
    onSelect(selection.focus(item));
  };
  onItemSelect = (item, selected) => {
    const {
      selection,
      onSelect
    } = this.props;
    if (selected) {
      onSelect(selection.select(item));
    } else {
      onSelect(selection.deselect(item));
    }
  };
  onEqualPress = () => {
    const {
      selection,
      itemFormatter
    } = this.props;
    const focused = selection.getFocused();
    if (focused === null || focused === undefined) {
      throw new Error('No focused item');
    }
    const item = itemFormatter(focused);
    if (item.collapsed) {
      item.onExpand?.();
    } else {
      item.onCollapse?.();
    }
  };
  shortcutsMap = {
    '=': this.onEqualPress
  };
  render() {
    const {
      data,
      className,
      loading,
      selection,
      disabledHover,
      itemFormatter,
      focused,
      innerRef
    } = this.props;
    const shortcutsMap = {
      ...this.shortcutsMap,
      ...this.props.shortcutsMap
    };
    const classes = classNames(className, {
      [styles.dataList]: true,
      [styles.disabledHover]: disabledHover,
      [styles.multiSelection]: selection.getSelected().size > 0
    });
    return /*#__PURE__*/jsxs("div", {
      className: styles.dataListWrapper,
      "data-test": "ring-data-list",
      ref: innerRef,
      children: [focused && /*#__PURE__*/jsx(Shortcuts, {
        map: shortcutsMap,
        scope: this.shortcutsScope
      }), /*#__PURE__*/jsx("ul", {
        className: classes,
        children: data.map(model => {
          const item = itemFormatter(model);
          const {
            id,
            key,
            title,
            items
          } = item;
          const showMoreLessButton = this.props.itemMoreLessState?.(item);
          return /*#__PURE__*/jsx(Item, {
            item: model,
            title: title,
            items: items,
            itemFormatter: itemFormatter,
            collapsible: item.collapsible,
            collapsed: item.collapsed,
            onCollapse: item.onCollapse,
            onExpand: item.onExpand,
            showFocus: selection.isFocused(model),
            onFocus: this.onItemFocus,
            selection: selection,
            selectable: item.selectable,
            selected: selection.isSelected(model),
            partialSelected: selection.isPartialSelected(model),
            onSelect: this.onItemSelect,
            showMoreLessButton: showMoreLessButton,
            onItemMoreLess: this.props.onItemMoreLess
          }, key || id);
        })
      }), loading && /*#__PURE__*/jsx("div", {
        className: data.length > 0 ? styles.loadingOverlay : undefined,
        children: /*#__PURE__*/jsx(Loader, {})
      })]
    });
  }
}
const getContainer = () => disableHoverHOC(selectionShortcutsHOC(focusSensorHOC(DataList)));
// eslint-disable-next-line react/no-multi-comp
class DataListContainer extends Component {
  // https://stackoverflow.com/a/53882322/6304152
  DataList = getContainer();
  render() {
    return /*#__PURE__*/jsx(this.DataList, {
      ...this.props
    });
  }
}

export { DataListContainer as default };
