// @jetbrains/ring-ui-built/components/data-list/data-list.mock.js
import Link from '../link/link.js';
import Tag from '../tag/tag.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react';
import 'classnames';
import '../global/data-tests.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '@jetbrains/icons/close-12px';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';

const items = [{
  id: 1,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx("strong", {
      children: "Assigner"
    }), " in 60 projects"]
  }),
  collapsible: true,
  selectable: true,
  items: [{
    id: 11,
    title: '6 projects: as a member of jetbrains-team',
    collapsible: true,
    selectable: true,
    items: [{
      id: 111,
      selectable: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "TeamCity"
      })
    }, {
      id: 112,
      selectable: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "YouTrack"
      })
    }, {
      id: 113,
      selectable: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "Sandbox"
      })
    }, {
      id: 114,
      selectable: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "Exception Analyzer"
      })
    }, {
      id: 115,
      selectable: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "MPS"
      })
    }]
  }]
}, {
  id: 2,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx("strong", {
      children: "Code Reviewer"
    }), " in 5 projects"]
  }),
  selectable: true,
  collapsible: true,
  items: [{
    id: 21,
    selectable: true,
    title: /*#__PURE__*/jsxs("span", {
      children: [/*#__PURE__*/jsx(Link, {
        href: "#",
        children: "JetProfile"
      }), /*#__PURE__*/jsx("span", {
        children: " "
      }), /*#__PURE__*/jsx(Tag, {
        children: "duplicate"
      })]
    })
  }, {
    id: 22,
    title: '6 projects: as a member of jetbrains-team',
    collapsible: true,
    items: [{
      id: 221,
      collapsible: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "TeamCity"
      })
    }, {
      id: 222,
      collapsible: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "YouTrack"
      })
    }, {
      id: 223,
      collapsible: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "Sandbox"
      })
    }, {
      id: 224,
      collapsible: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "Exception Analyzer"
      })
    }, {
      id: 225,
      collapsible: true,
      title: /*#__PURE__*/jsx(Link, {
        href: "#",
        children: "MPS"
      })
    }]
  }, {
    id: 23,
    collapsible: true,
    title: /*#__PURE__*/jsxs("span", {
      children: [/*#__PURE__*/jsx(Link, {
        href: "#",
        children: "CustomJDK"
      }), /*#__PURE__*/jsx("span", {
        children: ": as a member of ide-developers"
      })]
    })
  }]
}, {
  id: 3,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx("strong", {
      children: "Code Viewer"
    }), " in 5 projects"]
  }),
  selectable: true,
  collapsible: true,
  items: [{
    id: 31,
    collapsible: true,
    title: /*#__PURE__*/jsxs("span", {
      children: [/*#__PURE__*/jsx(Link, {
        href: "#",
        children: "ide-frontend"
      }), /*#__PURE__*/jsx("span", {
        children: ": as a member of ide-developers"
      })]
    })
  }, {
    id: 32,
    collapsible: true,
    title: /*#__PURE__*/jsxs("span", {
      children: [/*#__PURE__*/jsx(Link, {
        href: "#",
        children: "ide-frontend"
      }), /*#__PURE__*/jsx("span", {
        children: ": as a member of ide-frontend Code Viewers"
      })]
    })
  }]
}, {
  id: 4,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx("strong", {
      children: "Commenter"
    }), " in 12 projects"]
  }),
  collapsible: true,
  items: [{
    id: 41,
    collapsible: true,
    title: /*#__PURE__*/jsxs("span", {
      children: [/*#__PURE__*/jsx(Link, {
        href: "#",
        children: "Intellij IDEA"
      }), /*#__PURE__*/jsx("span", {
        children: ": as a member of Registered Users"
      })]
    })
  }, {
    id: 42,
    collapsible: true,
    title: /*#__PURE__*/jsxs("span", {
      children: [/*#__PURE__*/jsx(Link, {
        href: "#",
        children: "YouTrack Backlog"
      }), /*#__PURE__*/jsx("span", {
        children: ": as a member of jetbrains-team"
      })]
    })
  }]
}, {
  id: 5,
  collapsible: true,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx("strong", {
      children: "Developer"
    }), /*#__PURE__*/jsx("span", {
      children: " in 57 projects "
    }), /*#__PURE__*/jsx(Tag, {
      children: "team role"
    })]
  }),
  items: [{
    id: 51,
    selectable: true,
    collapsible: true,
    title: /*#__PURE__*/jsx(Link, {
      href: "#",
      children: "Wasabi"
    })
  }]
}, {
  id: 6,
  collapsible: true,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx("strong", {
      children: "Git Hosting Admin"
    }), /*#__PURE__*/jsx("span", {
      children: " in project "
    }), /*#__PURE__*/jsx(Link, {
      href: "#",
      children: "Hackathon 2016"
    }), /*#__PURE__*/jsx("span", {
      children: ": as a member of jetbrains-team"
    })]
  })
}, {
  id: 7,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx("strong", {
      children: "Git Hosting Checkout"
    }), " in 7 projects"]
  }),
  selectable: true,
  collapsible: true
}];
const moreItems = [{
  id: 91,
  selectable: true,
  collapsible: true,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx(Link, {
      href: "#",
      children: "More item 1"
    }), /*#__PURE__*/jsx("span", {
      children: " "
    }), /*#__PURE__*/jsx(Tag, {
      children: "duplicate"
    })]
  })
}, {
  id: 92,
  selectable: true,
  collapsible: true,
  title: /*#__PURE__*/jsxs("span", {
    children: [/*#__PURE__*/jsx(Link, {
      href: "#",
      children: "More item 2"
    }), /*#__PURE__*/jsx("span", {
      children: " "
    }), /*#__PURE__*/jsx(Tag, {
      children: "duplicate"
    })]
  })
}];

export { items as default, moreItems };
