// @jetbrains/ring-ui-built/components/data-list/item.js
import { PureComponent } from 'react';
import chevronRightIcon from '@jetbrains/icons/chevron-right';
import chevronDownIcon from '@jetbrains/icons/chevron-down';
import Link from '../link/link.js';
import Text from '../text/text.js';
import LoaderInline from '../loader-inline/loader-inline.js';
import { Button } from '../button/button.js';
import { s as styles, T as Title } from '../_helpers/title.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'classnames';
import '../global/data-tests.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../global/focus-sensor-hoc.js';
import '../global/compose-refs.js';
import 'memoize-one';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../control-help/control-help.js';
import '../global/get-uid.js';

var moreLessButtonStates;
(function (moreLessButtonStates) {
  moreLessButtonStates[moreLessButtonStates["UNUSED"] = 0] = "UNUSED";
  moreLessButtonStates[moreLessButtonStates["MORE"] = 1] = "MORE";
  moreLessButtonStates[moreLessButtonStates["MORE_LOADING"] = 2] = "MORE_LOADING";
  moreLessButtonStates[moreLessButtonStates["LESS"] = 3] = "LESS";
})(moreLessButtonStates || (moreLessButtonStates = {}));
const ITEM_LEFT_OFFSET = 32;
const LIST_LEFT_OFFSET = 24;
class Item extends PureComponent {
  static defaultProps = {
    items: [],
    level: 0,
    parentShift: 0,
    showMoreLessButton: moreLessButtonStates.UNUSED,
    onItemMoreLess: () => {}
  };
  onShowMore = () => {
    const {
      onItemMoreLess,
      item
    } = this.props;
    onItemMoreLess(item, true);
  };
  onShowLess = () => {
    const {
      onItemMoreLess,
      item
    } = this.props;
    onItemMoreLess(item, false);
  };
  onFocus = () => {
    const {
      onFocus,
      item
    } = this.props;
    onFocus(item);
  };
  onSelect = selected => {
    const {
      onSelect,
      item
    } = this.props;
    onSelect(item, selected);
  };
  renderItem = (model, parentShift) => {
    const {
      onFocus,
      onSelect,
      selection,
      level,
      itemFormatter
    } = this.props;
    const item = itemFormatter(model);
    return /*#__PURE__*/jsx(Item, {
      item: model,
      title: item.title,
      items: item.items,
      level: level + 1,
      parentShift: parentShift,
      itemFormatter: itemFormatter,
      collapsible: item.collapsible,
      collapsed: item.collapsed,
      onCollapse: item.onCollapse,
      onExpand: item.onExpand,
      showFocus: selection.isFocused(model),
      onFocus: onFocus,
      selection: selection,
      selectable: item.selectable,
      selected: selection.isSelected(model),
      partialSelected: selection.isPartialSelected(model),
      onSelect: onSelect
    }, item.key || item.id);
  };
  render() {
    const {
      title,
      items,
      showMoreLessButton,
      level,
      parentShift,
      showFocus,
      selectable,
      selected,
      partialSelected,
      collapsible,
      collapsed,
      onCollapse,
      onExpand
    } = this.props;
    let moreLessButton;
    if (showMoreLessButton === moreLessButtonStates.MORE || showMoreLessButton === moreLessButtonStates.MORE_LOADING) {
      moreLessButton = /*#__PURE__*/jsxs(Text, {
        info: true,
        size: Text.Size.S,
        children: [/*#__PURE__*/jsx(Link, {
          inherit: true,
          pseudo: true,
          onClick: this.onShowMore,
          children: 'Show more'
        }), showMoreLessButton === moreLessButtonStates.MORE_LOADING && /*#__PURE__*/jsx(LoaderInline, {
          className: styles.showMoreLoader
        })]
      });
    } else if (showMoreLessButton === moreLessButtonStates.LESS) {
      moreLessButton = /*#__PURE__*/jsx(Text, {
        info: true,
        size: Text.Size.S,
        children: /*#__PURE__*/jsx(Link, {
          inherit: true,
          pseudo: true,
          onClick: this.onShowLess,
          children: 'Show less'
        })
      });
    }
    let collapserExpander = null;
    if (collapsible) {
      if (collapsed) {
        collapserExpander = /*#__PURE__*/jsx(Button, {
          title: "Expand",
          onClick: onExpand,
          icon: chevronRightIcon,
          className: styles.collapseButton,
          iconClassName: styles.collapseIcon,
          "data-test": "ring-data-list-expand"
        });
      } else {
        collapserExpander = /*#__PURE__*/jsx(Button, {
          title: "Collapse",
          onClick: onCollapse,
          icon: chevronDownIcon,
          className: styles.collapseButton,
          iconClassName: styles.collapseIcon,
          "data-test": "ring-data-list-collapse"
        });
      }
    }
    const itemIsEmpty = !items?.length || collapsible && collapsed;
    const offset = level * LIST_LEFT_OFFSET + ITEM_LEFT_OFFSET + parentShift;
    return /*#__PURE__*/jsxs("li", {
      children: [/*#__PURE__*/jsx(Title, {
        title: title,
        focused: showFocus,
        showFocus: showFocus,
        selectable: selectable,
        selected: selected,
        partialSelected: partialSelected,
        collapserExpander: collapserExpander,
        onFocus: this.onFocus,
        onSelect: this.onSelect,
        offset: offset
      }), !itemIsEmpty ? /*#__PURE__*/jsxs("ul", {
        className: styles.itemContent,
        children: [items.map(model => this.renderItem(model, parentShift)), showMoreLessButton !== moreLessButtonStates.UNUSED ? /*#__PURE__*/jsx("li", {
          className: styles.showMore,
          children: moreLessButton
        }) : null]
      }) : null]
    });
  }
}

export { Item as default, moreLessButtonStates };
