// @jetbrains/ring-ui-built/components/data-list/selection.js
import Selection$1 from '../table/selection.js';

class Selection extends Selection$1 {
  _partialSelected;
  constructor({
    partialSelected = new Set(),
    ...props
  }) {
    super(props);
    this._partialSelected = partialSelected;
  }
  _buildData(data) {
    return new Set(this._getDescendants(data));
  }
  _buildSelected(data, selected) {
    // eslint-disable-next-line no-underscore-dangle
    const _selected = new Set(selected);
    [...data].forEach(item => {
      if (_selected.has(item)) {
        this._selectDescendants(item, _selected, this._partialSelected);
      }
    });
    return _selected;
  }
  _getDescendants(items) {
    let result = [];
    items.forEach(item => {
      result.push(item);
      result = [...result, ...this._getDescendants(this._getChildren(item))];
    });
    return result;
  }
  _getAncestors(item) {
    let result = [];
    const parent = [...this._data].find(it => this._getChildren(it).includes(item));
    if (parent) {
      result = [parent, ...this._getAncestors(parent)];
    }
    return result;
  }
  _selectDescendants(item, selected, partialSelected) {
    this._getDescendants(this._getChildren(item)).forEach(it => {
      selected.add(it);
      partialSelected.delete(it);
    });
  }
  _deselectDescendants(item, selected, partialSelected) {
    this._getDescendants(this._getChildren(item)).forEach(it => {
      selected.delete(it);
      partialSelected.delete(it);
    });
  }
  _selectAncestors(item, selected, partialSelected) {
    this._getAncestors(item).forEach(ancestor => {
      const groupIsSelected = this._getChildren(ancestor).filter(it => this._isItemSelectable(it)).every(it => selected.has(it));
      const groupIsPartialSelected = this._getChildren(ancestor).filter(it => this._isItemSelectable(it)).some(it => selected.has(it) || partialSelected.has(it));
      if (groupIsSelected) {
        selected.add(ancestor);
        partialSelected.delete(ancestor);
      } else if (groupIsPartialSelected) {
        partialSelected.add(ancestor);
      }
    });
  }
  _deselectAncestors(item, selected, partialSelected) {
    this._getAncestors(item).forEach(ancestor => {
      const groupIsPartialSelected = this._getChildren(ancestor).filter(it => this._isItemSelectable(it)).filter(it => it !== item).some(it => selected.has(it) || partialSelected.has(it));
      if (groupIsPartialSelected) {
        partialSelected.add(ancestor);
      } else {
        partialSelected.delete(ancestor);
      }
      selected.delete(ancestor);
    });
  }
  select(value = this._focused) {
    if (!value || !this._isItemSelectable(value)) {
      return this;
    }
    const selected = new Set(this._selected);
    const partialSelected = new Set(this._partialSelected);
    selected.add(value);
    partialSelected.delete(value);
    this._selectDescendants(value, selected, partialSelected);
    this._selectAncestors(value, selected, partialSelected);
    return this.cloneWith({
      selected,
      partialSelected
    });
  }
  isPartialSelected(value) {
    return value !== null && value !== undefined && this._partialSelected.has(value);
  }
  focus(value) {
    return super.focus(value);
  }
  resetSelection() {
    return super.resetSelection();
  }
  cloneWith({
    partialSelected = this._partialSelected,
    ...rest
  }) {
    const parentClone = super.cloneWith(rest);
    return new this.constructor({
      // eslint-disable-next-line no-underscore-dangle
      data: parentClone._rawData,
      // eslint-disable-next-line no-underscore-dangle
      selected: parentClone._selected,
      // eslint-disable-next-line no-underscore-dangle
      focused: parentClone._focused,
      // eslint-disable-next-line no-underscore-dangle
      getKey: parentClone._getKey,
      // eslint-disable-next-line no-underscore-dangle
      getChildren: parentClone._getChildren,
      // eslint-disable-next-line no-underscore-dangle
      isItemSelectable: parentClone._isItemSelectable,
      partialSelected
    });
  }
  deselect(value = this._focused) {
    if (!value || !this._isItemSelectable(value)) {
      return this;
    }
    const selected = new Set(this._selected);
    const partialSelected = new Set(this._partialSelected);
    selected.delete(value);
    partialSelected.delete(value);
    this._deselectDescendants(value, selected, partialSelected);
    this._deselectAncestors(value, selected, partialSelected);
    return this.cloneWith({
      selected,
      partialSelected
    });
  }
}

export { Selection as default };
