// @jetbrains/ring-ui-built/components/list/list-custom.js
import { PureComponent } from 'react';
import joinDataTestAttributes from '../global/data-tests.js';
import getEventKey from '../global/get-event-key.js';
import { getListClasses } from './list.classes.js';
import { jsx } from 'react/jsx-runtime';
import 'classnames';
import './consts.js';
import '../_helpers/list.js';

class ListCustom extends PureComponent {
  handleKeyPress = event => {
    const key = getEventKey(event);
    if (key === 'Enter' || key === ' ') {
      this.props.onClick(event);
    }
  };
  render() {
    const {
      disabled,
      template,
      tabIndex,
      onClick,
      onMouseOver,
      onMouseUp,
      role,
      tagName
    } = this.props;
    const classes = getListClasses(this.props);
    const dataTest = joinDataTestAttributes('ring-list-item-custom', {
      'ring-list-item-action': !disabled
    }, this.props['data-test']);
    const content = typeof template === 'function' ? template(this.props) : template;
    const TagName = tagName || 'span';
    return /*#__PURE__*/jsx(TagName, {
      role: role || 'button',
      "aria-disabled": disabled,
      tabIndex: tabIndex,
      onClick: onClick,
      onKeyPress: this.handleKeyPress,
      onMouseOver: onMouseOver,
      onFocus: onMouseOver,
      onMouseUp: onMouseUp,
      className: classes,
      "data-test": dataTest,
      children: content
    });
  }
}

export { ListCustom as default };
