// @jetbrains/ring-ui-built/components/list/list-item.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import joinDataTestAttributes from '../global/data-tests.js';
import Avatar from '../avatar/avatar.js';
import Checkbox from '../checkbox/checkbox.js';
import Icon from '../icon/icon.js';
import getUID from '../global/get-uid.js';
import Link, { linkHOC } from '../link/link.js';
import { Type } from './consts.js';
import { getListClasses } from './list.classes.js';
import { s as styles } from '../_helpers/list.js';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { Size } from '../avatar/avatar-size.js';
import 'util-deprecate';
import '../global/url.js';
import '../global/dom.js';
import '../global/memoize.js';
import '../avatar/fallback-avatar.js';
import 'react-compiler-runtime';
import '../_helpers/avatar-info.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../global/compose-refs.js';
import 'memoize-one';
import '../control-help/control-help.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';

const RING_UNIT = 8;
const DEFAULT_PADDING = 8;
const CHECKBOX_WIDTH = 28;
class ListItem extends PureComponent {
  id = getUID('list-item-');
  stopBubbling = e => e.stopPropagation();
  _isString = val => typeof val === 'string' || val instanceof String;
  // eslint-disable-next-line complexity
  render() {
    const {
      disabled,
      checkbox,
      avatar,
      subavatar,
      glyph,
      icon,
      rightGlyph,
      description,
      label,
      title,
      details,
      hover,
      level,
      tabIndex,
      onClick,
      onCheckboxChange,
      onMouseOver,
      onMouseDown,
      onMouseUp,
      rightNodes,
      leftNodes,
      showGeneratedAvatar,
      username,
      labelWrapper,
      rgItemType,
      scrolling,
      'data-test': dataTest,
      className,
      url,
      LinkComponent,
      compact,
      hoverClassName,
      children,
      itemClassName,
      labelClassName,
      descriptionClassName,
      detailsClassName,
      ...restLinkProps // TODO use an allow list in 8.0
    } = this.props;
    const checkable = checkbox !== undefined;
    const shouldShowGeneratedAvatar = showGeneratedAvatar && !!username;
    const hasLeftNodes = leftNodes || glyph || avatar || shouldShowGeneratedAvatar;
    const showCheckbox = checkable && (checkbox || !hasLeftNodes || hover && !disabled);
    const classes = getListClasses(this.props);
    const detailsClasses = classNames({
      [styles.details]: details,
      [styles.padded]: icon !== undefined || checkbox !== undefined || glyph !== undefined
    }, detailsClassName);
    const style = {
      paddingLeft: `${(Number(level) || 0) * RING_UNIT + DEFAULT_PADDING + (showCheckbox ? CHECKBOX_WIDTH : 0)}px`
    };
    let computedTitle = null;
    if (this._isString(title)) {
      // if title is specified and is a string then use it
      computedTitle = title;
    } else {
      // otherwise use label if it is a string;
      // label can also be an element, use empty string in this case
      computedTitle = this._isString(label) ? label : '';
    }
    const isLink = rgItemType === Type.LINK;
    const combinedDataTest = joinDataTestAttributes({
      'ring-list-item': (dataTest || '').indexOf('ring-list-item') === -1,
      'ring-list-item-action': !disabled,
      'ring-list-item-selected': checkbox,
      'ring-list-link': isLink
    }, dataTest);
    const labelElement = /*#__PURE__*/jsx("span", {
      className: classNames(styles.label, labelClassName),
      title: computedTitle,
      "data-test": "ring-list-item-label",
      children: label !== null && label !== void 0 ? label : children
    });
    const commonProps = {
      id: this.id,
      tabIndex,
      onClick,
      onMouseOver,
      onMouseDown,
      onFocus: onMouseOver,
      onMouseUp,
      className: classes,
      style,
      disabled,
      children: /*#__PURE__*/jsxs(Fragment, {
        children: [/*#__PURE__*/jsxs("div", {
          className: styles.top,
          onMouseOut: this.stopBubbling,
          onBlur: this.stopBubbling,
          children: [!showCheckbox && /*#__PURE__*/jsxs("div", {
            className: styles.left,
            children: [leftNodes, glyph && /*#__PURE__*/jsx(Icon, {
              className: styles.glyph,
              glyph: glyph,
              size: this.props.iconSize,
              suppressSizeWarning: this.props.suppressSizeWarning
            }), (avatar || shouldShowGeneratedAvatar) && /*#__PURE__*/jsx(Avatar, {
              className: styles.avatar,
              url: avatar,
              size: Size.Size20,
              subavatar: subavatar,
              username: username
            })]
          }), labelWrapper ? labelWrapper(labelElement) : labelElement, description && /*#__PURE__*/jsx("span", {
            className: classNames(styles.description, descriptionClassName),
            "data-test": "ring-list-item-description",
            children: description
          }), /*#__PURE__*/jsxs("div", {
            className: styles.right,
            children: [rightGlyph && /*#__PURE__*/jsx(Icon, {
              className: styles.rightGlyph,
              glyph: rightGlyph,
              suppressSizeWarning: this.props.suppressSizeWarning,
              size: this.props.iconSize
            }), icon && /*#__PURE__*/jsx("div", {
              className: styles.icon,
              style: {
                backgroundImage: `url("${icon}")`
              }
            }), rightNodes]
          })]
        }), details && /*#__PURE__*/jsx("div", {
          className: detailsClasses,
          children: details
        })]
      })
    };
    const LinkComponentToUse = LinkComponent ? linkHOC(LinkComponent) : Link;
    return /*#__PURE__*/jsxs("div", {
      className: classNames(styles.itemContainer, itemClassName),
      "data-test": combinedDataTest,
      children: [showCheckbox && /*#__PURE__*/jsx("div", {
        className: styles.checkboxContainer,
        children: /*#__PURE__*/jsx(Checkbox, {
          "aria-labelledby": this.id,
          checked: checkbox,
          disabled: disabled,
          onChange: onCheckboxChange,
          onClick: this.stopBubbling
        })
      }), isLink ? /*#__PURE__*/jsx(LinkComponentToUse, {
        pseudo: !restLinkProps.href,
        ...commonProps,
        ...restLinkProps
      }) : /*#__PURE__*/jsx("button", {
        type: "button",
        ...commonProps
      })]
    });
  }
}

export { ListItem as default };
