// @jetbrains/ring-ui-built/components/list/list-users-groups-source.js
import GroupIcon from '@jetbrains/icons/group';
import HubSourceUsersGroups from '../hub-source/hub-source-users-groups.js';
import List from './list.js';
import '../hub-source/hub-source.js';
import 'react';
import 'classnames';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'util-deprecate';
import 'memoize-one';
import '../global/data-tests.js';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/memoize.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/create-stateful-context.js';
import 'react-compiler-runtime';
import 'react/jsx-runtime';
import './list-item.js';
import '../avatar/avatar.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import './consts.js';
import './list.classes.js';
import '../_helpers/list.js';
import './list-custom.js';
import '../global/get-event-key.js';
import './list-title.js';
import './list-separator.js';
import './list-hint.js';

const defaultOptions = {
  GroupsTitle: 'Groups',
  NoGroupsTitle: 'No groups',
  UsersTitle: 'Users',
  NoUsersTitle: 'No users',
  getPluralForUserCount: count => {
    // eslint-disable-next-line no-magic-numbers
    const plural = count % 10 !== 1 || count % 100 === 11;
    return `${count} member${plural ? 's' : ''}`;
  }
};
var Filter;
(function (Filter) {
  Filter[Filter["ALL"] = 0] = "ALL";
  Filter[Filter["USERS"] = 1] = "USERS";
  Filter[Filter["GROUPS"] = 2] = "GROUPS";
})(Filter || (Filter = {}));
class ListUsersGroupsSource extends HubSourceUsersGroups {
  static Filter = Filter;
  listSourceOptions;
  constructor(auth, options) {
    super(auth, options);
    this.listSourceOptions = Object.assign({}, defaultOptions, options);
  }
  getGroupsSectionTitle(groups) {
    return groups.length ? this.listSourceOptions.GroupsTitle : this.listSourceOptions.NoGroupsTitle;
  }
  getUsersSectionTitle(users) {
    return users.length ? this.listSourceOptions.UsersTitle : this.listSourceOptions.NoUsersTitle;
  }
  async getForList(query, filter = Filter.ALL) {
    const [users, groups] = await this.getUserAndGroups(query);
    const items = [];
    if (filter === Filter.ALL) {
      items.push({
        rgItemType: List.ListProps.Type.SEPARATOR,
        key: 2,
        description: this.getUsersSectionTitle(users)
      });
    }
    if (filter !== Filter.GROUPS) {
      users.forEach(user => items.push({
        ...user,
        key: user.id,
        label: user.name,
        avatar: user.profile ? user.profile.avatar?.url : null,
        description: user.login
      }));
    }
    if (filter === Filter.ALL) {
      items.push({
        rgItemType: List.ListProps.Type.SEPARATOR,
        key: 1,
        description: this.getGroupsSectionTitle(groups)
      });
    }
    if (filter !== Filter.USERS) {
      groups.forEach(group => items.push({
        ...group,
        key: group.id,
        label: group.name,
        avatar: group.iconUrl,
        glyph: group.iconUrl ? null : GroupIcon,
        description: this.listSourceOptions.getPluralForUserCount(group.userCount)
      }));
    }
    return items;
  }
}

export { ListUsersGroupsSource as default };
