// @jetbrains/ring-ui-built/components/confirm-service/confirm-service.js
import { createRoot } from 'react-dom/client';
import Confirm from '../confirm/confirm.js';
import { getGlobalControlsHeight, ControlsHeightContext } from '../global/controls-height.js';
import { jsx } from 'react/jsx-runtime';
import 'react';
import '../dialog/dialog.js';
import 'react-dom';
import 'classnames';
import '@jetbrains/icons/close';
import '../island/island.js';
import '../global/data-tests.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import '../_helpers/island.js';
import '../island/header.js';
import 'react-compiler-runtime';
import '../island/content.js';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../_helpers/button.classes.js';
import '../global/configuration.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '../panel/panel.js';

const containerElement = document.createElement('div');
const reactRoot = createRoot(containerElement);
/**
 * Renders Confirm into virtual node to skip maintaining container
 */
function renderConfirm(props) {
  const {
    buttonsHeight = getGlobalControlsHeight(),
    ...restProps
  } = props;
  reactRoot.render(/*#__PURE__*/jsx(ControlsHeightContext.Provider, {
    value: buttonsHeight,
    children: /*#__PURE__*/jsx(Confirm, {
      ...restProps
    })
  }));
}
function confirm({
  text,
  description,
  confirmLabel = 'OK',
  rejectLabel = 'Cancel',
  cancelIsDefault,
  onBeforeConfirm,
  buttonsHeight
}) {
  return new Promise((resolve, reject) => {
    const props = {
      text,
      description,
      confirmLabel,
      rejectLabel,
      cancelIsDefault,
      buttonsHeight,
      show: true,
      onConfirm: () => {
        if (onBeforeConfirm) {
          renderConfirm({
            ...props,
            inProgress: true
          });
          return Promise.resolve(onBeforeConfirm()).then(() => {
            renderConfirm({
              ...props,
              show: false
            });
            resolve();
          }).catch(err => {
            renderConfirm({
              ...props,
              show: false
            });
            reject(err);
          });
        }
        renderConfirm({
          ...props,
          show: false
        });
        return resolve();
      },
      onReject: () => {
        renderConfirm({
          ...props,
          show: false
        });
        reject(new Error('Confirm(@jetbrains/ring-ui): null exception'));
      }
    };
    renderConfirm(props);
  });
}
function hideConfirm() {
  renderConfirm({
    text: '',
    show: false
  });
}

export { confirm as default, hideConfirm, reactRoot };
