// @jetbrains/ring-ui-built/components/date-picker/consts.js
import { add } from 'date-fns/add';

const unit = 8; // px;
const units = {
  unit,
  /* eslint-disable no-magic-numbers */
  cellSize: unit * 3,
  calHeight: unit * 36,
  yearHeight: unit * 4
  /* eslint-enable */
};
const YEAR = 12;
const WEEK = 7;
const FIFTH_DAY = 4;
const weekdays = {
  MO: 1,
  TU: 2,
  WE: 3,
  TH: 4,
  FR: 5,
  SA: 6,
  SU: 0
};
const MIDDLE_DAY = 15;
const durationToMillis = duration => +add(0, duration);
const yearDuration = durationToMillis({
  years: 1
});
const yearScrollSpeed = yearDuration / (YEAR * units.cellSize);
const DOUBLE = 2;
const HALF = 0.5;
function parseTime(time) {
  let result = null;
  if (/^([01][0-9]|2[0-3]):[0-5][0-9]$/.test(time)) {
    result = time;
  } else if (/^([0-9]|2[0-3]):[0-5][0-9]$/.test(time)) {
    result = `0${time}`;
  }
  return result;
}
function shiftWeekArray(arr, startOfWeek) {
  const shiftTimes = startOfWeek - 1;
  return arr.slice(shiftTimes).concat(arr.slice(0, shiftTimes));
}
function getWeekStartsOn(locale) {
  var _locale$options$weekS;
  return (_locale$options$weekS = locale?.options?.weekStartsOn) !== null && _locale$options$weekS !== void 0 ? _locale$options$weekS : weekdays.MO;
}
function getDayNumInWeek(locale, day) {
  const weekDays = shiftWeekArray(Object.values(weekdays), getWeekStartsOn(locale));
  return weekDays.indexOf(day);
}

export { DOUBLE, FIFTH_DAY, HALF, MIDDLE_DAY, WEEK, YEAR, units as default, getDayNumInWeek, getWeekStartsOn, parseTime, shiftWeekArray, weekdays, yearDuration, yearScrollSpeed };
