// @jetbrains/ring-ui-built/components/date-picker/date-input.js
import * as React from 'react';
import classNames from 'classnames';
import { I as Input } from '../_helpers/input.js';
import '../global/controls-height.js';
import { I18nContext } from '../i18n/i18n-context.js';
import { s as styles } from '../_helpers/date-picker.js';
import { jsx } from 'react/jsx-runtime';
import { ControlsHeight } from '../global/configuration.js';
import '@jetbrains/icons/close-12px';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../_helpers/button.classes.js';
import '../global/get-uid.js';
import '../global/compose-refs.js';
import 'memoize-one';
import '../control-label/control-label.js';
import '../control-help/control-help.js';
import '../i18n/i18n.js';

class DateInput extends React.PureComponent {
  componentDidMount() {
    this.updateInput(this.props);
  }
  componentDidUpdate(prevProps) {
    const {
      text,
      active
    } = this.props;
    if (text !== prevProps.text || active !== prevProps.active) {
      this.updateInput({
        text,
        active
      });
    }
  }
  static contextType = I18nContext;
  input;
  inputRef = el => {
    this.input = el;
  };
  updateInput({
    text,
    active
  }) {
    const el = this.input;
    if (!el) {
      return;
    }
    if (active) {
      el.focus();
      if (!text) {
        el.select();
      }
    } else {
      el.blur();
    }
  }
  handleChange = e => this.props.onInput(e.currentTarget.value, e.currentTarget.dataset.name);
  handleKeyDown = e => {
    if (e.key === 'Enter') {
      // We have to prevent the default behavior, because restoring focus by TabTrap caused by
      // pressing Enter will trigger the onClick event on the DatePicker Dropdown anchor,
      // so DatePicker will be open again.
      // https://youtrack.jetbrains.com/issue/RG-2450/Anchor-should-be-focused-after-closing-datepicker#focus=Comments-27-10044234.0-0.
      e.preventDefault();
      this.props.onConfirm();
    }
  };
  render() {
    const {
      active,
      divider,
      text,
      time,
      name,
      hoverDate,
      date,
      displayFormat,
      translations,
      onActivate,
      onClear,
      fromPlaceholder,
      toPlaceholder,
      timePlaceholder,
      locale
    } = this.props;
    const {
      translate
    } = this.context;
    let displayText = '';
    if (active && hoverDate) {
      displayText = displayFormat(hoverDate, locale);
    } else if (active && text !== null && text !== undefined) {
      displayText = text;
    } else if (date) {
      displayText = displayFormat(date, locale);
    } else if (name === 'time') {
      displayText = time || '';
    }
    const placeholder = ((_translations$addFirs, _translations$addSeco, _translations$addTime, _translations$selectN) => {
      switch (name) {
        case 'from':
          return fromPlaceholder || ((_translations$addFirs = translations?.addFirstDate) !== null && _translations$addFirs !== void 0 ? _translations$addFirs : translate('addFirstDate'));
        case 'to':
          return toPlaceholder || ((_translations$addSeco = translations?.addSecondDate) !== null && _translations$addSeco !== void 0 ? _translations$addSeco : translate('addSecondDate'));
        case 'time':
          return timePlaceholder || ((_translations$addTime = translations?.addTime) !== null && _translations$addTime !== void 0 ? _translations$addTime : translate('addTime'));
        default:
          return ((_translations$selectN = translations?.selectName) !== null && _translations$selectN !== void 0 ? _translations$selectN : translate('selectName')).replace('%name%', name).replace('{{name}}', name);
      }
    })();
    const classes = classNames(styles.filter, styles[`${name}Input`], divider && styles[`${name}InputWithDivider`], 'ring-js-shortcuts');
    return /*#__PURE__*/jsx(Input, {
      autoComplete: "off",
      borderless: true,
      height: ControlsHeight.L,
      "data-name": name,
      inputRef: this.inputRef,
      className: classes,
      value: displayText,
      onChange: this.handleChange,
      onFocus: onActivate,
      onKeyDown: this.handleKeyDown,
      onClear: onClear,
      placeholder: placeholder
    });
  }
}

export { DateInput as default };
