// @jetbrains/ring-ui-built/components/date-picker/date-picker.js
import { c } from 'react-compiler-runtime';
import { PureComponent } from 'react';
import classNames from 'classnames';
import { format } from 'date-fns/format';
import { isSameDay } from 'date-fns/isSameDay';
import { isSameMonth } from 'date-fns/isSameMonth';
import { isSameYear } from 'date-fns/isSameYear';
import { isValid } from 'date-fns/isValid';
import { parse } from 'date-fns/parse';
import { set } from 'date-fns/set';
import calendarIcon from '@jetbrains/icons/calendar';
import chevronDownIcon from '@jetbrains/icons/chevron-down';
import memoize from '../global/memoize.js';
import Popup from '../popup/popup.js';
import Dropdown from '../dropdown/dropdown.js';
import Icon from '../icon/icon.js';
import { Button } from '../button/button.js';
import Link from '../link/link.js';
import { S as Size } from '../_helpers/input.js';
import { I18nContext } from '../i18n/i18n-context.js';
import DatePopup from './date-popup.js';
import formats from './formats.js';
import { s as styles } from '../_helpers/date-picker.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/data-tests.js';
import '../tab-trap/tab-trap.js';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../_helpers/button.classes.js';
import '../_helpers/link.js';
import '@jetbrains/icons/close-12px';
import '../global/compose-refs.js';
import 'memoize-one';
import '../control-label/control-label.js';
import '../control-help/control-help.js';
import '../i18n/i18n.js';
import 'date-fns/isAfter';
import 'date-fns/isBefore';
import 'date-fns/startOfDay';
import 'date-fns';
import './date-input.js';
import './months.js';
import 'date-fns/addMonths';
import 'date-fns/getDay';
import 'date-fns/getDaysInMonth';
import 'date-fns/startOfMonth';
import 'date-fns/subMonths';
import 'date-fns/endOfMonth';
import '../global/linear-function.js';
import '../global/use-event-callback.js';
import './month.js';
import 'date-fns/addDays';
import 'date-fns/setDay';
import './day.js';
import 'date-fns/getDate';
import 'date-fns/isToday';
import './consts.js';
import 'date-fns/add';
import './month-names.js';
import 'date-fns/isThisMonth';
import 'date-fns/startOfYear';
import './month-slider.js';
import 'date-fns/addYears';
import 'date-fns/subYears';
import './years.js';
import 'date-fns/getYear';
import 'date-fns/isThisYear';
import 'date-fns/setYear';
import './weekdays.js';

const PopupComponent = t0 => {
  const $ = c(20);
  if ($[0] !== "33db58a6327d40ad4b7bf91a7fec9f9c197da5c37e0c1c6775581fe962cb7d9d") {
    for (let $i = 0; $i < 20; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "33db58a6327d40ad4b7bf91a7fec9f9c197da5c37e0c1c6775581fe962cb7d9d";
  }
  let className;
  let datePopupProps;
  let onClear;
  let onComplete;
  let popupRef;
  let restProps;
  let t1;
  if ($[1] !== t0) {
    ({
      hidden: t1,
      className,
      popupRef,
      onClear,
      datePopupProps,
      onComplete,
      ...restProps
    } = t0);
    $[1] = t0;
    $[2] = className;
    $[3] = datePopupProps;
    $[4] = onClear;
    $[5] = onComplete;
    $[6] = popupRef;
    $[7] = restProps;
    $[8] = t1;
  } else {
    className = $[2];
    datePopupProps = $[3];
    onClear = $[4];
    onComplete = $[5];
    popupRef = $[6];
    restProps = $[7];
    t1 = $[8];
  }
  const hidden = t1 === undefined ? false : t1;
  let t2;
  if ($[9] === Symbol.for("react.memo_cache_sentinel")) {
    t2 = [Popup.PopupProps.Directions.BOTTOM_RIGHT, Popup.PopupProps.Directions.BOTTOM_LEFT, Popup.PopupProps.Directions.TOP_LEFT, Popup.PopupProps.Directions.TOP_RIGHT];
    $[9] = t2;
  } else {
    t2 = $[9];
  }
  let t3;
  if ($[10] !== datePopupProps || $[11] !== onClear || $[12] !== onComplete) {
    t3 = /*#__PURE__*/jsx(DatePopup, {
      onClear: onClear,
      ...datePopupProps,
      onComplete: onComplete
    });
    $[10] = datePopupProps;
    $[11] = onClear;
    $[12] = onComplete;
    $[13] = t3;
  } else {
    t3 = $[13];
  }
  let t4;
  if ($[14] !== className || $[15] !== hidden || $[16] !== popupRef || $[17] !== restProps || $[18] !== t3) {
    t4 = /*#__PURE__*/jsx(Popup, {
      hidden: hidden,
      className: className,
      ref: popupRef,
      directions: t2,
      ...restProps,
      trapFocus: true,
      children: t3
    });
    $[14] = className;
    $[15] = hidden;
    $[16] = popupRef;
    $[17] = restProps;
    $[18] = t3;
    $[19] = t4;
  } else {
    t4 = $[19];
  }
  return t4;
};
/**
 * @name Date Picker
 */
class DatePicker extends PureComponent {
  static defaultProps = {
    className: '',
    date: null,
    withTime: false,
    range: false,
    from: null,
    to: null,
    clear: false,
    inline: false,
    size: Size.M,
    displayFormat: (date, locale) => date ? format(date, 'd MMM yyyy', {
      locale
    }) : '',
    displayMonthFormat: (date, locale) => date ? format(date, 'd MMM', {
      locale
    }) : '',
    displayDayFormat: (date, locale) => date ? format(date, 'd', {
      locale
    }) : '',
    displayTimeFormat: (date, locale) => date ? format(date, 'HH:mm', {
      locale
    }) : '',
    minDate: null,
    maxDate: null,
    onChange() {},
    applyTimeInput(date, timeString) {
      var _timeString$split$map;
      const [hours, minutes] = (_timeString$split$map = timeString?.split(':').map(Number)) !== null && _timeString$split$map !== void 0 ? _timeString$split$map : [];
      return minutes !== null && minutes !== undefined ? set(date, {
        hours,
        minutes
      }) : date;
    },
    parseDateInput(string) {
      if (!string) {
        return null;
      }
      const today = new Date();
      for (const format of formats) {
        const date = parse(string, format, today);
        if (isValid(date)) {
          return date;
        }
      }
      return null;
    }
  };
  static contextType = I18nContext;
  handleChange = change => {
    const {
      onChange,
      withTime,
      applyTimeInput
    } = this.props;
    const adjustedChange = withTime && !(change instanceof Date) && change?.date ? applyTimeInput(change.date, change.time) : change;
    onChange(adjustedChange);
  };
  clear = () => {
    let change = null;
    if (this.props.range) {
      change = {
        from: null,
        to: null
      };
    }
    this.handleChange(change);
  };
  popup;
  popupRef = el => {
    this.popup = el;
  };
  closePopup = () => {
    // eslint-disable-next-line no-underscore-dangle
    this.popup?._onCloseAttempt();
  };
  parse = memoize(date => {
    const {
      parseDateInput
    } = this.props;
    if (date instanceof Date) {
      return date;
    }
    if (typeof date === 'number') {
      return new Date(date);
    }
    return parseDateInput(date);
  });
  formatTime() {
    const {
      displayTimeFormat,
      locale
    } = this.props;
    const date = this.parse(this.props.date);
    if (date) {
      return displayTimeFormat(date, locale);
    }
    return null;
  }
  // eslint-disable-next-line complexity
  getAnchorText = () => {
    var _ref3;
    const {
      range,
      datePlaceholder,
      dateTimePlaceholder,
      rangePlaceholder,
      withTime,
      displayFormat,
      displayMonthFormat,
      displayDayFormat,
      translations,
      locale
    } = this.props;
    const {
      translate
    } = this.context;
    const date = this.parse(this.props.date);
    const from = this.parse(this.props.from);
    const to = this.parse(this.props.to);
    const time = this.formatTime();
    if (!range && !withTime) {
      var _ref;
      return date ? displayFormat(date, locale) : (_ref = datePlaceholder !== null && datePlaceholder !== void 0 ? datePlaceholder : translations?.setDate) !== null && _ref !== void 0 ? _ref : translate('setDate');
    }
    if (!range && withTime) {
      if (!date && !time) {
        var _ref2;
        return (_ref2 = dateTimePlaceholder !== null && dateTimePlaceholder !== void 0 ? dateTimePlaceholder : translations?.setDateTime) !== null && _ref2 !== void 0 ? _ref2 : translate('setDateTime');
      }
      return `${date && displayFormat(date, locale) || '—'}, ${time || '—'}`;
    }
    if (from && to) {
      if (!isSameYear(from, to)) {
        return `${displayFormat(from, locale)} — ${displayFormat(to, locale)}`;
      }
      if (!isSameMonth(from, to)) {
        return `${displayMonthFormat(from, locale)} — ${displayFormat(to, locale)}`;
      }
      if (!isSameDay(from, to)) {
        return `${displayDayFormat(from, locale)} — ${displayFormat(to, locale)}`;
      }
      return `${displayFormat(to, locale)}`;
    }
    if (from) {
      return `${displayFormat(from, locale)} —`;
    }
    if (to) {
      return `— ${displayFormat(to, locale)}`;
    }
    return (_ref3 = rangePlaceholder !== null && rangePlaceholder !== void 0 ? rangePlaceholder : translations?.setPeriod) !== null && _ref3 !== void 0 ? _ref3 : translate('setPeriod');
  };
  render() {
    var _this$props$disabled, _this$props$disabled2;
    const anchorContent = /*#__PURE__*/jsxs("div", {
      className: styles.anchorContent,
      children: [/*#__PURE__*/jsx(Icon, {
        glyph: calendarIcon,
        className: styles.calendarIcon
      }), this.getAnchorText(), /*#__PURE__*/jsx(Icon, {
        glyph: chevronDownIcon,
        className: styles.chevronDownIcon
      })]
    });
    const {
      className,
      popupClassName,
      clear,
      inline,
      dropdownProps,
      translations,
      ...datePopupProps
    } = this.props;
    const classes = classNames(styles.datePicker, className, styles[`size${this.props.size}`], {
      [styles.inline]: inline
    });
    return /*#__PURE__*/jsx(Dropdown, {
      className: classes,
      disabled: this.props.disabled,
      "data-test": "ring-date-picker",
      anchor: inline ? /*#__PURE__*/jsx(Link, {
        "data-test-ring-dropdown-anchor": true,
        className: styles.anchor,
        disabled: (_this$props$disabled = this.props.disabled) !== null && _this$props$disabled !== void 0 ? _this$props$disabled : false,
        pseudo: true,
        children: this.getAnchorText()
      }) : /*#__PURE__*/jsx(Button, {
        "data-test-ring-dropdown-anchor": true,
        className: styles.anchor,
        inline: false,
        disabled: (_this$props$disabled2 = this.props.disabled) !== null && _this$props$disabled2 !== void 0 ? _this$props$disabled2 : false,
        ...this.props.buttonAttributes,
        children: anchorContent
      }),
      ...dropdownProps,
      children: /*#__PURE__*/jsx(PopupComponent, {
        className: popupClassName,
        popupRef: this.popupRef,
        onClear: clear ? this.clear : null,
        datePopupProps: {
          ...datePopupProps,
          translations,
          onChange: this.handleChange,
          parseDateInput: this.parse,
          time: this.formatTime()
        },
        onComplete: this.closePopup
      })
    });
  }
}

export { DatePicker as default };
