// @jetbrains/ring-ui-built/components/date-picker/months.js
import { c } from 'react-compiler-runtime';
import { useRef, useEffect, createElement } from 'react';
import { addMonths } from 'date-fns/addMonths';
import { getDay } from 'date-fns/getDay';
import { getDaysInMonth } from 'date-fns/getDaysInMonth';
import { startOfMonth } from 'date-fns/startOfMonth';
import { subMonths } from 'date-fns/subMonths';
import { endOfMonth } from 'date-fns/endOfMonth';
import scheduleRAF from '../global/schedule-raf.js';
import linearFunction from '../global/linear-function.js';
import useEventCallback from '../global/use-event-callback.js';
import Month from './month.js';
import MonthNames from './month-names.js';
import units, { DOUBLE, HALF, WEEK, weekdays } from './consts.js';
import { s as styles } from '../_helpers/date-picker.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'date-fns/addDays';
import 'date-fns/format';
import 'date-fns/setDay';
import './day.js';
import 'classnames';
import 'date-fns/getDate';
import 'date-fns/isAfter';
import 'date-fns/isBefore';
import 'date-fns/isSameDay';
import 'date-fns/isToday';
import 'date-fns/startOfDay';
import 'date-fns/isThisMonth';
import 'date-fns/set';
import 'date-fns/startOfYear';
import './month-slider.js';
import 'date-fns/addYears';
import 'date-fns/subYears';
import 'date-fns/add';

const {
  unit,
  cellSize,
  calHeight
} = units;
const FridayToSunday = WEEK + weekdays.SU - weekdays.FR;
const FIVELINES = 31;
const TALLMONTH = 6;
const SHORTMONTH = 5;
const PADDING = 2;
const MONTHSBACK = 2;
function monthHeight(date) {
  const monthStart = startOfMonth(date);
  const daysSinceLastFriday = (getDay(monthStart) + FridayToSunday) % WEEK;
  const monthLines = daysSinceLastFriday + getDaysInMonth(monthStart) > FIVELINES ? TALLMONTH : SHORTMONTH;
  return monthLines * cellSize + unit * PADDING;
}
// in milliseconds per pixel
function scrollSpeed(date) {
  const monthStart = startOfMonth(date);
  const monthEnd = endOfMonth(date);
  return (Number(monthEnd) - Number(monthStart)) / monthHeight(monthStart);
}
const scrollSchedule = scheduleRAF();
let dy = 0;
function Months(props) {
  const $ = c(30);
  if ($[0] !== "5d5fe78f7bda393a523d9472cfc721d7ae4d117fe8278fd6d3d9589b20c1969a") {
    for (let $i = 0; $i < 30; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "5d5fe78f7bda393a523d9472cfc721d7ae4d117fe8278fd6d3d9589b20c1969a";
  }
  const {
    scrollDate
  } = props;
  let monthStart;
  let months;
  if ($[1] !== scrollDate) {
    const monthDate = scrollDate instanceof Date ? scrollDate : new Date(scrollDate);
    monthStart = startOfMonth(monthDate);
    let month = subMonths(monthStart, MONTHSBACK);
    months = [month];
    for (let i = 0; i < MONTHSBACK * DOUBLE; i++) {
      month = addMonths(month, 1);
      months.push(month);
    }
    $[1] = scrollDate;
    $[2] = monthStart;
    $[3] = months;
  } else {
    monthStart = $[2];
    months = $[3];
  }
  let pxToDate;
  let t0;
  if ($[4] !== monthStart || $[5] !== scrollDate) {
    const currentSpeed = scrollSpeed(scrollDate);
    pxToDate = linearFunction(0, Number(scrollDate), currentSpeed);
    t0 = pxToDate.x(Number(monthStart));
    $[4] = monthStart;
    $[5] = scrollDate;
    $[6] = pxToDate;
    $[7] = t0;
  } else {
    pxToDate = $[6];
    t0 = $[7];
  }
  const offset = t0;
  const bottomOffset = monthHeight(scrollDate) + offset;
  const componentRef = useRef(null);
  let t1;
  if ($[8] !== bottomOffset || $[9] !== months || $[10] !== offset || $[11] !== props || $[12] !== pxToDate) {
    t1 = e => {
      e.preventDefault();
      dy = dy + e.deltaY;
      scrollSchedule(() => {
        let date;
        if (dy < offset) {
          date = pxToDate.y(offset) + (dy - offset) * scrollSpeed(months[1]);
        } else {
          if (dy > bottomOffset) {
            date = pxToDate.y(bottomOffset) + (dy - bottomOffset) * scrollSpeed(months[MONTHSBACK + 1]);
          } else {
            date = pxToDate.y(dy);
          }
        }
        props.onScroll(date);
        dy = 0;
      });
    };
    $[8] = bottomOffset;
    $[9] = months;
    $[10] = offset;
    $[11] = props;
    $[12] = pxToDate;
    $[13] = t1;
  } else {
    t1 = $[13];
  }
  const handleWheel = useEventCallback(t1);
  let t2;
  let t3;
  if ($[14] !== handleWheel) {
    t2 = () => {
      const current = componentRef.current;
      if (current) {
        current.addEventListener("wheel", handleWheel, {
          passive: false
        });
      }
      return () => {
        if (current) {
          current.removeEventListener("wheel", handleWheel);
        }
      };
    };
    t3 = [handleWheel];
    $[14] = handleWheel;
    $[15] = t2;
    $[16] = t3;
  } else {
    t2 = $[15];
    t3 = $[16];
  }
  useEffect(t2, t3);
  const t4 = Math.floor(calHeight * HALF - monthHeight(months[0]) - monthHeight(months[1]) + offset);
  let t5;
  if ($[17] !== t4) {
    t5 = {
      top: t4
    };
    $[17] = t4;
    $[18] = t5;
  } else {
    t5 = $[18];
  }
  let t6;
  if ($[19] !== months || $[20] !== props) {
    t6 = months.map(date_0 => /*#__PURE__*/createElement(Month, {
      ...props,
      month: date_0,
      key: +date_0
    }));
    $[19] = months;
    $[20] = props;
    $[21] = t6;
  } else {
    t6 = $[21];
  }
  let t7;
  if ($[22] !== t5 || $[23] !== t6) {
    t7 = /*#__PURE__*/jsx("div", {
      style: t5,
      className: styles.days,
      children: t6
    });
    $[22] = t5;
    $[23] = t6;
    $[24] = t7;
  } else {
    t7 = $[24];
  }
  let t8;
  if ($[25] !== props) {
    t8 = /*#__PURE__*/jsx(MonthNames, {
      ...props
    });
    $[25] = props;
    $[26] = t8;
  } else {
    t8 = $[26];
  }
  let t9;
  if ($[27] !== t7 || $[28] !== t8) {
    t9 = /*#__PURE__*/jsxs("div", {
      className: styles.months,
      ref: componentRef,
      children: [t7, t8]
    });
    $[27] = t7;
    $[28] = t8;
    $[29] = t9;
  } else {
    t9 = $[29];
  }
  return t9;
}

export { Months as default };
