// @jetbrains/ring-ui-built/components/date-picker/month-slider.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import { addYears } from 'date-fns/addYears';
import { startOfDay } from 'date-fns/startOfDay';
import { subYears } from 'date-fns/subYears';
import linearFunction from '../global/linear-function.js';
import units, { yearScrollSpeed } from './consts.js';
import { s as styles } from '../_helpers/date-picker.js';
import { jsx } from 'react/jsx-runtime';
import 'date-fns/add';

const COVERYEARS = 3;
class MonthSlider extends PureComponent {
  state = {
    dragging: false
  };
  componentDidUpdate(prevProps, prevState) {
    if (this.state.dragging && !prevState.dragging) {
      window.addEventListener('mousemove', this.onMouseMove);
      window.addEventListener('mouseup', this.onMouseUp);
    } else if (!this.state.dragging && prevState.dragging) {
      window.removeEventListener('mousemove', this.onMouseMove);
      window.removeEventListener('mouseup', this.onMouseUp);
    }
  }
  onMouseDown = () => {
    this.setState({
      dragging: true
    });
  };
  onMouseUp = () => {
    this.setState({
      dragging: false
    });
  };
  onMouseMove = e => {
    this.props.onScroll(linearFunction(0, Number(this.props.scrollDate), yearScrollSpeed).y(e.movementY));
  };
  render() {
    let year = subYears(startOfDay(this.props.scrollDate), 1);
    const years = [year];
    for (let i = 0; i <= COVERYEARS; i++) {
      year = addYears(year, 1);
      years.push(year);
    }
    const classes = classNames(styles.monthSlider, {
      [styles.dragging]: this.state.dragging
    });
    return /*#__PURE__*/jsx("div", {
      children: years.map(date => /*#__PURE__*/jsx("button", {
        type: "button",
        className: classes,
        style: {
          top: Math.floor(this.props.pxToDate.x(Number(date)) - units.cellSize)
        },
        onMouseDown: this.onMouseDown
      }, +date))
    });
  }
}

export { MonthSlider as default };
