// @jetbrains/ring-ui-built/components/date-picker/month-names.js
import { c } from 'react-compiler-runtime';
import { PureComponent } from 'react';
import classNames from 'classnames';
import { endOfMonth } from 'date-fns/endOfMonth';
import { format } from 'date-fns/format';
import { isThisMonth } from 'date-fns/isThisMonth';
import { set } from 'date-fns/set';
import { startOfDay } from 'date-fns/startOfDay';
import { startOfYear } from 'date-fns/startOfYear';
import linearFunction from '../global/linear-function.js';
import MonthSlider from './month-slider.js';
import { YEAR, MIDDLE_DAY, yearScrollSpeed } from './consts.js';
import { s as styles } from '../_helpers/date-picker.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'date-fns/addYears';
import 'date-fns/subYears';
import 'date-fns/add';

class MonthName extends PureComponent {
  handleClick = () => {
    const end = endOfMonth(this.props.month);
    this.props.onScrollChange(end.getTime());
  };
  render() {
    const {
      month,
      locale
    } = this.props;
    return /*#__PURE__*/jsx("button", {
      type: "button",
      className: classNames(styles.monthName, {
        [styles.today]: isThisMonth(month)
      }),
      onClick: this.handleClick,
      children: format(month, 'LLL', {
        locale
      })
    });
  }
}
function MonthNames(props) {
  const $ = c(25);
  if ($[0] !== "bc2d18bdf292acf1ee5e9f58d8bdc0707131612de3d6a30b9b3c175a40cf30a9") {
    for (let $i = 0; $i < 25; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "bc2d18bdf292acf1ee5e9f58d8bdc0707131612de3d6a30b9b3c175a40cf30a9";
  }
  const {
    scrollDate,
    locale
  } = props;
  let months;
  if ($[1] !== scrollDate) {
    months = [];
    for (let i = 0; i < YEAR; i++) {
      const middleDay = set(scrollDate, {
        month: i,
        date: MIDDLE_DAY
      });
      months.push(startOfDay(middleDay));
    }
    $[1] = scrollDate;
    $[2] = months;
  } else {
    months = $[2];
  }
  let bottom;
  let pxToDate;
  let top;
  if ($[3] !== props.currentRange || $[4] !== scrollDate) {
    let t0;
    if ($[8] !== scrollDate) {
      t0 = startOfYear(scrollDate);
      $[8] = scrollDate;
      $[9] = t0;
    } else {
      t0 = $[9];
    }
    pxToDate = linearFunction(0, Number(t0), yearScrollSpeed);
    top = 0;
    bottom = 0;
    if (props.currentRange) {
      [top, bottom] = props.currentRange.map(date => Math.floor(pxToDate.x(Number(date))));
    }
    $[3] = props.currentRange;
    $[4] = scrollDate;
    $[5] = bottom;
    $[6] = pxToDate;
    $[7] = top;
  } else {
    bottom = $[5];
    pxToDate = $[6];
    top = $[7];
  }
  let t0;
  if ($[10] !== locale || $[11] !== months || $[12] !== props.onScrollChange) {
    t0 = months.map(month => /*#__PURE__*/jsx(MonthName, {
      month: month,
      onScrollChange: props.onScrollChange,
      locale: locale
    }, +month));
    $[10] = locale;
    $[11] = months;
    $[12] = props.onScrollChange;
    $[13] = t0;
  } else {
    t0 = $[13];
  }
  let t1;
  if ($[14] !== bottom || $[15] !== props.currentRange || $[16] !== top) {
    t1 = props.currentRange && /*#__PURE__*/jsx("div", {
      className: styles.range,
      style: {
        top: top - 1,
        height: bottom + 1 - (top - 1)
      }
    });
    $[14] = bottom;
    $[15] = props.currentRange;
    $[16] = top;
    $[17] = t1;
  } else {
    t1 = $[17];
  }
  let t2;
  if ($[18] !== props || $[19] !== pxToDate) {
    t2 = /*#__PURE__*/jsx(MonthSlider, {
      ...props,
      pxToDate: pxToDate
    });
    $[18] = props;
    $[19] = pxToDate;
    $[20] = t2;
  } else {
    t2 = $[20];
  }
  let t3;
  if ($[21] !== t0 || $[22] !== t1 || $[23] !== t2) {
    t3 = /*#__PURE__*/jsxs("div", {
      className: styles.monthNames,
      children: [t0, t1, t2]
    });
    $[21] = t0;
    $[22] = t1;
    $[23] = t2;
    $[24] = t3;
  } else {
    t3 = $[24];
  }
  return t3;
}

export { MonthNames as default };
