// @jetbrains/ring-ui-built/components/error-message/error-message.js
import { Component } from 'react';
import classNames from 'classnames';
import Icon from '../icon/icon.js';
import { Size } from '../icon/icon.constants.js';
import joinDataTestAttributes from '../global/data-tests.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'util-deprecate';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';

var styles = {"errorMessage":"ring-error-message-errorMessage","title":"ring-error-message-title","description":"ring-error-message-description","content":"ring-error-message-content","icon":"ring-error-message-icon"};

class ErrorMessage extends Component {
  render() {
    const {
      className,
      icon,
      code,
      message,
      description,
      children,
      'data-test': dataTest
    } = this.props;
    const classes = classNames(styles.errorMessage, className);
    return /*#__PURE__*/jsxs("div", {
      className: classes,
      "data-test": joinDataTestAttributes('ring-error-message', dataTest),
      children: [icon && /*#__PURE__*/jsx(Icon, {
        className: styles.icon,
        glyph: icon,
        size: Size.Size64,
        suppressSizeWarning: true
      }), /*#__PURE__*/jsxs("div", {
        className: styles.content,
        children: [/*#__PURE__*/jsxs("div", {
          className: styles.title,
          "data-test": "ring-error-message-title",
          children: [code && `${code}:`, " ", message]
        }), description && /*#__PURE__*/jsx("div", {
          className: styles.description,
          children: description
        }), children]
      })]
    });
  }
}

export { ErrorMessage as default };
