// @jetbrains/ring-ui-built/components/table/disable-hover-hoc.js
import { PureComponent } from 'react';
import { jsx } from 'react/jsx-runtime';

function disableHoverHOC(ComposedComponent) {
  return class DisableHover extends PureComponent {
    static defaultProps = ComposedComponent.defaultProps;
    state = {
      disabledHover: false
    };
    componentDidMount() {
      document.addEventListener('mousemove', this.onMouseMove);
      document.addEventListener('keydown', this.onKeyDown, true);
    }
    componentWillUnmount() {
      document.removeEventListener('mousemove', this.onMouseMove);
      document.removeEventListener('keydown', this.onKeyDown, true);
    }
    onMouseMove = () => {
      if (this.state.disabledHover) {
        this.setState({
          disabledHover: false
        });
      }
    };
    onKeyDown = e => {
      const metaKeys = [16, 17, 18, 19, 20, 91]; // eslint-disable-line no-magic-numbers
      if (!this.state.disabledHover && !metaKeys.includes(e.keyCode)) {
        this.setState({
          disabledHover: true
        });
      }
    };
    render() {
      return /*#__PURE__*/jsx(ComposedComponent, {
        ...this.props,
        disabledHover: this.state.disabledHover
      });
    }
  };
}

export { disableHoverHOC as default };
