// @jetbrains/ring-ui-built/components/table/header.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import Checkbox from '../checkbox/checkbox.js';
import getUID from '../global/get-uid.js';
import HeaderCell from './header-cell.js';
import { s as style } from '../_helpers/table.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../global/compose-refs.js';
import 'memoize-one';
import '../control-help/control-help.js';
import '@jetbrains/icons/unsorted-12px';
import '@jetbrains/icons/chevron-12px-down';
import '../global/data-tests.js';

class Header extends PureComponent {
  static defaultProps = {
    selectable: true,
    draggable: false,
    checked: true,
    sticky: true,
    topStickOffset: '0px',
    onSort: () => {},
    onCheckboxChange: () => {},
    sortKey: 'id',
    sortOrder: true
  };
  id = getUID('table-header-');
  onCheckboxFocus = event => {
    event.currentTarget.blur();
  };
  createCells(widths = []) {
    const {
      selectable,
      draggable,
      columns,
      checked,
      checkboxDisabled,
      onCheckboxChange,
      onSort,
      sortKey,
      sortOrder
    } = this.props;
    const metaColumnClasses = classNames(style.metaColumn, style.headerMetaColumn);
    const metaColumn = /*#__PURE__*/jsx("div", {
      className: metaColumnClasses,
      children: selectable && /*#__PURE__*/jsx(Checkbox, {
        "aria-labelledby": this.id,
        disabled: checkboxDisabled,
        checked: checked,
        onChange: onCheckboxChange,
        onFocus: this.onCheckboxFocus
      })
    });
    return columns.map((column, index) => {
      const columnStyle = widths[index] ? {
        width: widths[index]
      } : undefined;
      const props = {
        column,
        onSort,
        sortKey,
        sortOrder,
        style: columnStyle
      };
      return /*#__PURE__*/jsx(HeaderCell, {
        "data-test": column.id,
        ...props,
        children: index === 0 && (draggable || selectable) && metaColumn
      }, column.id);
    });
  }
  render() {
    const {
      caption,
      sticky,
      topStickOffset
    } = this.props;
    const regularCells = this.createCells();
    return /*#__PURE__*/jsxs("thead", {
      id: this.id,
      "data-test": "ring-table-header",
      style: {
        top: topStickOffset
      },
      className: classNames({
        [style.tableHead]: true,
        [style.subHeaderSticky]: sticky
      }, this.props.className),
      children: [caption && /*#__PURE__*/jsx("tr", {
        "data-test": "ring-table-header-row",
        children: /*#__PURE__*/jsx("th", {
          className: classNames(style.headerCell, style.caption),
          colSpan: regularCells.length + 1,
          "data-test": "ring-table-header-cell",
          children: caption
        })
      }), /*#__PURE__*/jsx("tr", {
        className: style.subHeader,
        "data-test": "ring-table-header-row",
        children: regularCells
      })]
    });
  }
}

export { Header as default };
