// @jetbrains/ring-ui-built/components/table/header-cell.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import sortableIcon from '@jetbrains/icons/unsorted-12px';
import chevron12pxDown from '@jetbrains/icons/chevron-12px-down';
import Icon from '../icon/icon.js';
import joinDataTestAttributes from '../global/data-tests.js';
import { s as style } from '../_helpers/table.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';

class HeaderCell extends PureComponent {
  static defaultProps = {
    onSort: () => {}
  };
  sortable;
  sorted;
  onClick = () => {
    if (this.sortable) {
      const {
        column,
        onSort,
        sortOrder
      } = this.props;
      onSort({
        column,
        order: !(this.sorted && sortOrder)
      });
    }
  };
  onChildrenClick(e) {
    e.stopPropagation();
  }
  render() {
    const {
      className,
      column,
      onSort,
      sortKey,
      sortOrder,
      'data-test': dataTest,
      ...restProps
    } = this.props;
    this.sortable = column.sortable === true;
    this.sorted = sortKey === column.id;
    const glyph = this.sorted ? chevron12pxDown : sortableIcon;
    const classes = classNames(className, column.headerClassName, {
      [style.headerCell]: true,
      [style.headerCellSortable]: this.sortable,
      [style.headerCellSorted]: this.sorted,
      [style.sortedUp]: sortOrder && this.sorted,
      [style.cellRight]: column.rightAlign
    });
    return /*#__PURE__*/jsxs("th", {
      ...restProps,
      className: classes,
      onClick: this.onClick,
      "data-test": joinDataTestAttributes('ring-table-header-cell', dataTest),
      children: [/*#__PURE__*/jsx("span", {
        onClick: this.onChildrenClick,
        role: "presentation",
        children: this.props.children
      }), column.getHeaderValue ? column.getHeaderValue() : column.title, this.sortable && /*#__PURE__*/jsx("span", {
        className: style.sorter,
        children: /*#__PURE__*/jsx(Icon, {
          glyph: glyph,
          className: style.icon
        })
      })]
    });
  }
}

export { HeaderCell as default };
