// @jetbrains/ring-ui-built/components/table/row-with-focus-sensor.js
import { PureComponent } from 'react';
import focusSensorHOC from '../global/focus-sensor-hoc.js';
import Row from './row.js';
import { jsx } from 'react/jsx-runtime';
import '../global/compose-refs.js';
import 'memoize-one';
import 'react-compiler-runtime';
import 'classnames';
import '@jetbrains/icons/chevron-right';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/drag';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../control-help/control-help.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../tooltip/tooltip.js';
import '../popup/popup.js';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/data-tests.js';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import './cell.js';
import '../_helpers/table.js';

const getContainer = () => focusSensorHOC(Row);
class RowWithFocusSensorCallbacks extends PureComponent {
  // https://stackoverflow.com/a/53882322/6304152
  RowWithFocusSensor = getContainer();
  onFocus = () => {
    this.props.onFocus?.(this.props.item);
  };
  onSelect = (item, selected) => {
    this.props.onSelect?.(item, selected);
  };
  onCollapse = () => {
    this.props.onCollapse?.(this.props.item);
  };
  onExpand = () => {
    this.props.onExpand?.(this.props.item);
  };
  render() {
    return /*#__PURE__*/jsx(this.RowWithFocusSensor, {
      ...this.props,
      onFocus: this.onFocus,
      onSelect: this.onSelect,
      onCollapse: this.onCollapse,
      onExpand: this.onExpand
    });
  }
}

export { RowWithFocusSensorCallbacks as default };
