// @jetbrains/ring-ui-built/components/table/row.js
import { c } from 'react-compiler-runtime';
import { PureComponent } from 'react';
import classNames from 'classnames';
import chevronRightIcon from '@jetbrains/icons/chevron-right';
import chevronDownIcon from '@jetbrains/icons/chevron-down';
import dragIcon from '@jetbrains/icons/drag';
import Checkbox from '../checkbox/checkbox.js';
import { Button } from '../button/button.js';
import Tooltip from '../tooltip/tooltip.js';
import joinDataTestAttributes from '../global/data-tests.js';
import getUID from '../global/get-uid.js';
import { createComposedRef } from '../global/compose-refs.js';
import Cell from './cell.js';
import { s as style } from '../_helpers/table.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../control-help/control-help.js';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../popup/popup.js';
import 'react-dom';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import 'memoize-one';

const DragHandle = t0 => {
  const $ = c(6);
  if ($[0] !== "55dd4b764fc7ce682e91a4f9ed76700dd0c03ecdbbc71fc25d11ed30dc931312") {
    for (let $i = 0; $i < 6; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "55dd4b764fc7ce682e91a4f9ed76700dd0c03ecdbbc71fc25d11ed30dc931312";
  }
  const {
    alwaysShowDragHandle,
    dragHandleTitle: t1
  } = t0;
  const dragHandleTitle = t1 === undefined ? "Drag to reorder" : t1;
  let t2;
  if ($[1] !== alwaysShowDragHandle) {
    t2 = classNames(style.dragHandle, {
      [style.visibleDragHandle]: alwaysShowDragHandle
    });
    $[1] = alwaysShowDragHandle;
    $[2] = t2;
  } else {
    t2 = $[2];
  }
  const classes = t2;
  let t3;
  if ($[3] !== classes || $[4] !== dragHandleTitle) {
    t3 = /*#__PURE__*/jsx(Button, {
      "data-movable-handle": true,
      title: dragHandleTitle,
      className: classes,
      icon: dragIcon
    });
    $[3] = classes;
    $[4] = dragHandleTitle;
    $[5] = t3;
  } else {
    t3 = $[5];
  }
  return t3;
};
class Row extends PureComponent {
  static defaultProps = {
    selectable: true,
    showFocus: false,
    draggable: false,
    alwaysShowDragHandle: false,
    selected: false,
    onHover: () => {},
    onSelect: () => {},
    onDoubleClick: () => {},
    onClick: () => {},
    level: 0,
    collapsible: false,
    parentCollapsible: false,
    collapsed: false,
    onCollapse: () => {},
    onExpand: () => {}
  };
  id = getUID('table-row-');
  onMouseEnter = e => {
    const {
      item,
      onHover
    } = this.props;
    onHover(item, e);
  };
  onClick = e => {
    const {
      item
    } = this.props;
    this.props.onClick(item, e);
    if (e.shiftKey) {
      this.toggleSelection();
    }
  };
  onCheckboxFocus = () => {
    this.props.onFocusRestore();
  };
  onCheckboxChange = () => {
    this.toggleSelection();
  };
  toggleSelection() {
    const {
      selectable,
      selected,
      onSelect,
      item
    } = this.props;
    if (selectable) {
      onSelect(item, !selected);
    }
  }
  onDoubleClick = () => {
    const {
      item
    } = this.props;
    this.props.onDoubleClick(item);
  };
  row;
  rowRef = el => {
    this.row = el;
  };
  composedRowRef = createComposedRef();
  render() {
    const {
      item,
      columns: columnProps,
      selectable,
      selected,
      showFocus,
      draggable,
      alwaysShowDragHandle,
      dragHandleTitle,
      level,
      collapsible,
      parentCollapsible,
      collapsed,
      onCollapse,
      onExpand,
      showDisabledSelection,
      onSelect,
      checkboxTooltip,
      innerRef,
      focused,
      autofocus,
      onFocusReset,
      onFocusRestore,
      onHover,
      className,
      metaColumnClassName,
      cellClassName,
      'data-test': dataTest,
      ...restProps
    } = this.props;
    const classes = classNames(className, {
      [style.row]: true,
      [style.rowFocused]: showFocus,
      [style.rowSelected]: selected
    });
    const testAttrs = {
      'data-test-focused': showFocus || undefined,
      'data-test-selected': selected || undefined
    };
    const metaColumnClasses = classNames(metaColumnClassName, style.metaColumn);
    const SUBITEM_OFFSET = 30;
    const COLLAPSIBLE_PARENT_OFFSET = 20;
    const gap = level * SUBITEM_OFFSET + (parentCollapsible ? COLLAPSIBLE_PARENT_OFFSET : 0);
    const metaColumnStyle = {
      paddingLeft: `${gap}px`
    };
    const metaColumn = /*#__PURE__*/jsxs("div", {
      className: metaColumnClasses,
      style: metaColumnStyle,
      children: [draggable && /*#__PURE__*/jsx(DragHandle, {
        alwaysShowDragHandle: alwaysShowDragHandle,
        dragHandleTitle: dragHandleTitle
      }), selectable && /*#__PURE__*/jsx(Tooltip, {
        title: checkboxTooltip,
        children: /*#__PURE__*/jsx(Checkbox, {
          "aria-labelledby": this.id,
          className: showFocus ? 'ring-checkbox_focus' : '',
          checked: selected,
          onFocus: this.onCheckboxFocus,
          onChange: this.onCheckboxChange,
          tabIndex: -1
        })
      }), !selectable && showDisabledSelection && /*#__PURE__*/jsx(Tooltip, {
        title: checkboxTooltip,
        children: /*#__PURE__*/jsx(Checkbox, {
          "aria-labelledby": this.id,
          checked: selected,
          disabled: true
        })
      }), collapsible && collapsed && /*#__PURE__*/jsx(Button, {
        className: style.rowCollapseExpandButton,
        icon: chevronRightIcon,
        onClick: () => onExpand(item)
      }), collapsible && !collapsed && /*#__PURE__*/jsx(Button, {
        className: style.rowCollapseExpandButton,
        icon: chevronDownIcon,
        onClick: () => onCollapse(item)
      })]
    });
    const columns = typeof columnProps === 'function' ? columnProps(item) : columnProps;
    const cells = columns.map((column, index) => {
      const getValue = column.getValue || (() => item[column.id]);
      const getDataTest = column.getDataTest || (() => column.id);
      const value = getValue(item, column);
      const cellClasses = classNames({
        [style.cellRight]: column.rightAlign
      }, column.className, cellClassName);
      const showMetaColumn = draggable || selectable || collapsible || showDisabledSelection || !!level;
      return /*#__PURE__*/jsxs(Cell, {
        colSpan: column.colSpan,
        className: cellClasses,
        "data-test": getDataTest(item, column),
        children: [index === 0 && showMetaColumn && metaColumn, value]
      }, column.id);
    });
    return /*#__PURE__*/jsx("tr", {
      id: this.id,
      ref: this.composedRowRef(this.rowRef, innerRef),
      className: classes,
      tabIndex: 0,
      "data-test": joinDataTestAttributes('ring-table-row', dataTest),
      ...testAttrs,
      ...restProps,
      onMouseMove: this.onMouseEnter,
      onClick: this.onClick,
      onDoubleClick: this.onDoubleClick,
      children: cells
    });
  }
}

export { Row as default };
