// @jetbrains/ring-ui-built/components/table/smart-table.js
import { PureComponent } from 'react';
import TableContainer from './table.js';
import Selection$1 from './selection.js';
import { jsx } from 'react/jsx-runtime';
import 'classnames';
import 'react-movable';
import '../global/focus-sensor-hoc.js';
import '../global/compose-refs.js';
import 'memoize-one';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../loader/loader.js';
import '../_helpers/loader-core.js';
import '../global/dom.js';
import './header.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../control-help/control-help.js';
import './header-cell.js';
import '@jetbrains/icons/unsorted-12px';
import '@jetbrains/icons/chevron-12px-down';
import '../global/data-tests.js';
import '../_helpers/table.js';
import './selection-shortcuts-hoc.js';
import './disable-hover-hoc.js';
import './row-with-focus-sensor.js';
import './row.js';
import '@jetbrains/icons/chevron-right';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/drag';
import '../button/button.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../tooltip/tooltip.js';
import '../popup/popup.js';
import 'react-dom';
import '../global/schedule-raf.js';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import './cell.js';

class SmartTable extends PureComponent {
  static defaultProps = {
    onSelectionChange: () => {}
  };
  state = {
    selection: new Selection$1({
      data: this.props.data,
      isItemSelectable: this.props.isItemSelectable
    })
  };
  UNSAFE_componentWillReceiveProps(nextProps) {
    const {
      data,
      isItemSelectable,
      selection
    } = nextProps;
    if (this.props.remoteSelection && this.props.selection !== selection) {
      this.setState({
        selection
      });
    } else if (this.props.data !== data || this.props.isItemSelectable !== isItemSelectable) {
      this.setState({
        selection: new Selection$1({
          data,
          isItemSelectable
        })
      });
    }
  }
  onSelect = selection => {
    this.setState({
      selection
    });
    this.props.onSelectionChange(selection);
  };
  render() {
    return /*#__PURE__*/jsx(TableContainer, {
      ...this.props,
      selection: this.state.selection,
      onSelect: this.onSelect
    });
  }
}

export { SmartTable as default };
