// @jetbrains/ring-ui-built/components/heading/heading.js
import { c } from 'react-compiler-runtime';
import { memo, useEffect } from 'react';
import classNames from 'classnames';
import deprecate from 'util-deprecate';
import { s as styles } from '../_helpers/heading.js';
import { jsx } from 'react/jsx-runtime';

var Levels;
(function (Levels) {
  Levels[Levels["H1"] = 1] = "H1";
  Levels[Levels["H2"] = 2] = "H2";
  Levels[Levels["H3"] = 3] = "H3";
  Levels[Levels["H4"] = 4] = "H4";
})(Levels || (Levels = {}));
const warnFallbackHeading = deprecate(() => {}, 'Headings of level 5 and higher are replaced with h3');
const HeadingMemo = /*#__PURE__*/memo(function Heading(t0) {
  const $ = c(16);
  if ($[0] !== "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8") {
    for (let $i = 0; $i < 16; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8";
  }
  let children;
  let className;
  let restProps;
  let t1;
  if ($[1] !== t0) {
    ({
      children,
      className,
      level: t1,
      ...restProps
    } = t0);
    $[1] = t0;
    $[2] = children;
    $[3] = className;
    $[4] = restProps;
    $[5] = t1;
  } else {
    children = $[2];
    className = $[3];
    restProps = $[4];
    t1 = $[5];
  }
  const level = t1 === undefined ? Levels.H1 : t1;
  let t2;
  if ($[6] !== className) {
    t2 = classNames(styles.heading, className);
    $[6] = className;
    $[7] = t2;
  } else {
    t2 = $[7];
  }
  const classes = t2;
  const isSupportedLevel = level <= Levels.H4;
  let t3;
  let t4;
  if ($[8] !== isSupportedLevel) {
    t3 = () => {
      if (!isSupportedLevel) {
        warnFallbackHeading();
      }
    };
    t4 = [isSupportedLevel];
    $[8] = isSupportedLevel;
    $[9] = t3;
    $[10] = t4;
  } else {
    t3 = $[9];
    t4 = $[10];
  }
  useEffect(t3, t4);
  const Tag = isSupportedLevel ? `h${level}` : "h3";
  let t5;
  if ($[11] !== Tag || $[12] !== children || $[13] !== classes || $[14] !== restProps) {
    t5 = /*#__PURE__*/jsx(Tag, {
      ...restProps,
      className: classes,
      children: children
    });
    $[11] = Tag;
    $[12] = children;
    $[13] = classes;
    $[14] = restProps;
    $[15] = t5;
  } else {
    t5 = $[15];
  }
  return t5;
});
const Heading = HeadingMemo;
Heading.Levels = Levels;
const H1 = /*#__PURE__*/memo(function H1(props) {
  const $ = c(3);
  if ($[0] !== "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8") {
    for (let $i = 0; $i < 3; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8";
  }
  let t0;
  if ($[1] !== props) {
    t0 = /*#__PURE__*/jsx(Heading, {
      ...props,
      level: Levels.H1
    });
    $[1] = props;
    $[2] = t0;
  } else {
    t0 = $[2];
  }
  return t0;
});
const H2 = /*#__PURE__*/memo(function H2(props) {
  const $ = c(3);
  if ($[0] !== "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8") {
    for (let $i = 0; $i < 3; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8";
  }
  let t0;
  if ($[1] !== props) {
    t0 = /*#__PURE__*/jsx(Heading, {
      ...props,
      level: Levels.H2
    });
    $[1] = props;
    $[2] = t0;
  } else {
    t0 = $[2];
  }
  return t0;
});
const H3 = /*#__PURE__*/memo(function H3(props) {
  const $ = c(3);
  if ($[0] !== "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8") {
    for (let $i = 0; $i < 3; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8";
  }
  let t0;
  if ($[1] !== props) {
    t0 = /*#__PURE__*/jsx(Heading, {
      ...props,
      level: Levels.H3
    });
    $[1] = props;
    $[2] = t0;
  } else {
    t0 = $[2];
  }
  return t0;
});
const H4 = /*#__PURE__*/memo(function H4(t0) {
  const $ = c(11);
  if ($[0] !== "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8") {
    for (let $i = 0; $i < 11; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "9f4538e180b30a0a8f1026b4057246b9b8f6ecaa39875bb67aa3caa7d9a451e8";
  }
  let bold;
  let className;
  let restProps;
  if ($[1] !== t0) {
    ({
      className,
      bold,
      ...restProps
    } = t0);
    $[1] = t0;
    $[2] = bold;
    $[3] = className;
    $[4] = restProps;
  } else {
    bold = $[2];
    className = $[3];
    restProps = $[4];
  }
  let t1;
  if ($[5] !== bold || $[6] !== className) {
    t1 = classNames(className, {
      [styles.bold]: bold
    });
    $[5] = bold;
    $[6] = className;
    $[7] = t1;
  } else {
    t1 = $[7];
  }
  const classes = t1;
  let t2;
  if ($[8] !== classes || $[9] !== restProps) {
    t2 = /*#__PURE__*/jsx(Heading, {
      ...restProps,
      className: classes,
      level: Levels.H4
    });
    $[8] = classes;
    $[9] = restProps;
    $[10] = t2;
  } else {
    t2 = $[10];
  }
  return t2;
});

export { H1, H2, H3, H4, Levels, Heading as default };
