// @jetbrains/ring-ui-built/components/collapse/collapse-content.js
import { c } from 'react-compiler-runtime';
import { useContext, useRef, useState, useEffect } from 'react';
import classNames from 'classnames';
import joinDataTestAttributes from '../global/data-tests.js';
import { getRect } from '../global/dom.js';
import { toPx } from './utils.js';
import { CollapseContext } from './collapse-context.js';
import { COLLAPSE_CONTENT_CONTAINER_TEST_ID, COLLAPSE_CONTENT_TEST_ID } from './consts.js';
import { jsx, jsxs } from 'react/jsx-runtime';

var styles = {"container":"ring-collapse-container","transition":"ring-collapse-transition","summary":"ring-collapse-summary","trigger":"ring-collapse-trigger","fade":"ring-collapse-fade"};

const DURATION_FACTOR = 0.5;
const DEFAULT_HEIGHT = 0;
const VISIBLE = 1;
const HIDDEN = 0;
/**
 * @name CollapseContent
 */
const CollapseContent = t0 => {
  const $ = c(29);
  if ($[0] !== "f05635de80c1ecb2484d9a8273c8da5ce7d933172be03e8d761443c32d96d937") {
    for (let $i = 0; $i < 29; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "f05635de80c1ecb2484d9a8273c8da5ce7d933172be03e8d761443c32d96d937";
  }
  const {
    children,
    minHeight: t1,
    "data-test": dataTest
  } = t0;
  const minHeight = t1 === undefined ? DEFAULT_HEIGHT : t1;
  const {
    collapsed,
    duration,
    id,
    disableAnimation
  } = useContext(CollapseContext);
  const containerRef = useRef(null);
  const contentRef = useRef(null);
  const [initialContentHeight] = useState(minHeight);
  const [contentHeight, setContentHeight] = useState(DEFAULT_HEIGHT);
  const nextHeight = collapsed ? initialContentHeight : contentHeight;
  let t2;
  if ($[1] !== nextHeight) {
    t2 = toPx(nextHeight);
    $[1] = nextHeight;
    $[2] = t2;
  } else {
    t2 = $[2];
  }
  const height = t2;
  const [shouldHideContent, setShouldHideContent] = useState(collapsed && minHeight <= DEFAULT_HEIGHT);
  let t3;
  let t4;
  if ($[3] !== collapsed || $[4] !== initialContentHeight) {
    t3 = () => {
      const onTransitionEnd = function onTransitionEnd() {
        if (initialContentHeight <= DEFAULT_HEIGHT) {
          setShouldHideContent(collapsed);
        }
      };
      const container = containerRef.current;
      container?.addEventListener("transitionend", onTransitionEnd);
      return () => {
        container?.removeEventListener("transitionend", onTransitionEnd);
      };
    };
    t4 = [collapsed, initialContentHeight];
    $[3] = collapsed;
    $[4] = initialContentHeight;
    $[5] = t3;
    $[6] = t4;
  } else {
    t3 = $[5];
    t4 = $[6];
  }
  useEffect(t3, t4);
  if (!collapsed && shouldHideContent) {
    setShouldHideContent(false);
  }
  let t5;
  let t6;
  if ($[7] === Symbol.for("react.memo_cache_sentinel")) {
    t5 = () => {
      if (contentRef.current) {
        const observer = new ResizeObserver(() => {
          setContentHeight(getRect(contentRef.current).height);
        });
        observer.observe(contentRef.current);
      }
    };
    t6 = [];
    $[7] = t5;
    $[8] = t6;
  } else {
    t5 = $[7];
    t6 = $[8];
  }
  useEffect(t5, t6);
  const calculatedDuration = duration + contentHeight * DURATION_FACTOR;
  const t7 = `${calculatedDuration}ms`;
  const t8 = collapsed && !minHeight ? HIDDEN : VISIBLE;
  let t9;
  if ($[9] !== height || $[10] !== t7 || $[11] !== t8) {
    t9 = {
      "--duration": t7,
      height,
      opacity: t8
    };
    $[9] = height;
    $[10] = t7;
    $[11] = t8;
    $[12] = t9;
  } else {
    t9 = $[12];
  }
  const style = t9;
  const fadeShouldBeVisible = Boolean(minHeight && collapsed);
  const shouldRenderContent = disableAnimation ? !collapsed : !shouldHideContent;
  const t10 = `collapse-content-${id}`;
  let t11;
  if ($[13] === Symbol.for("react.memo_cache_sentinel")) {
    t11 = joinDataTestAttributes(COLLAPSE_CONTENT_CONTAINER_TEST_ID);
    $[13] = t11;
  } else {
    t11 = $[13];
  }
  const t12 = !disableAnimation;
  let t13;
  if ($[14] !== t12) {
    t13 = classNames(styles.container, {
      [styles.transition]: t12
    });
    $[14] = t12;
    $[15] = t13;
  } else {
    t13 = $[15];
  }
  let t14;
  if ($[16] !== dataTest) {
    t14 = joinDataTestAttributes(COLLAPSE_CONTENT_TEST_ID, dataTest);
    $[16] = dataTest;
    $[17] = t14;
  } else {
    t14 = $[17];
  }
  const t15 = shouldRenderContent ? children : null;
  let t16;
  if ($[18] !== t14 || $[19] !== t15) {
    t16 = /*#__PURE__*/jsx("div", {
      ref: contentRef,
      "data-test": t14,
      children: t15
    });
    $[18] = t14;
    $[19] = t15;
    $[20] = t16;
  } else {
    t16 = $[20];
  }
  let t17;
  if ($[21] !== fadeShouldBeVisible) {
    t17 = fadeShouldBeVisible && /*#__PURE__*/jsx("div", {
      className: styles.fade
    });
    $[21] = fadeShouldBeVisible;
    $[22] = t17;
  } else {
    t17 = $[22];
  }
  let t18;
  if ($[23] !== style || $[24] !== t10 || $[25] !== t13 || $[26] !== t16 || $[27] !== t17) {
    t18 = /*#__PURE__*/jsxs("div", {
      ref: containerRef,
      id: t10,
      "data-test": t11,
      className: t13,
      style: style,
      children: [t16, t17]
    });
    $[23] = style;
    $[24] = t10;
    $[25] = t13;
    $[26] = t16;
    $[27] = t17;
    $[28] = t18;
  } else {
    t18 = $[28];
  }
  return t18;
};

export { CollapseContent, CollapseContent as default };
