// @jetbrains/ring-ui-built/components/collapse/collapse.js
import { c } from 'react-compiler-runtime';
import { useState, useId } from 'react';
import { CollapseContext } from './collapse-context.js';
import { BASE_ANIMATION_DURATION } from './consts.js';
import { jsx } from 'react/jsx-runtime';

const Collapse = t0 => {
  const $ = c(16);
  if ($[0] !== "4b1bc7149bb2f70cda665f353438e1d305c0392543f5779c6ac7ceb49ca39f10") {
    for (let $i = 0; $i < 16; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "4b1bc7149bb2f70cda665f353438e1d305c0392543f5779c6ac7ceb49ca39f10";
  }
  const {
    children,
    duration: t1,
    disableAnimation: t2,
    className: t3,
    onChange: t4,
    defaultCollapsed: t5,
    collapsed: t6
  } = t0;
  const duration = t1 === undefined ? BASE_ANIMATION_DURATION : t1;
  const disableAnimation = t2 === undefined ? false : t2;
  const className = t3 === undefined ? "" : t3;
  const onChange = t4 === undefined ? _temp : t4;
  const defaultCollapsed = t5 === undefined ? true : t5;
  const collapsed = t6 === undefined ? null : t6;
  const [innerCollapsed, setInnerCollapsed] = useState(defaultCollapsed);
  const id = useId();
  const finalCollapsedValue = collapsed !== null && collapsed !== void 0 ? collapsed : innerCollapsed;
  let t7;
  if ($[1] !== finalCollapsedValue || $[2] !== onChange) {
    t7 = () => {
      setInnerCollapsed(!finalCollapsedValue);
      onChange(!finalCollapsedValue);
    };
    $[1] = finalCollapsedValue;
    $[2] = onChange;
    $[3] = t7;
  } else {
    t7 = $[3];
  }
  const setCollapsed = t7;
  let t8;
  if ($[4] !== disableAnimation || $[5] !== duration || $[6] !== finalCollapsedValue || $[7] !== id || $[8] !== setCollapsed) {
    t8 = {
      collapsed: finalCollapsedValue,
      setCollapsed,
      duration,
      disableAnimation,
      id
    };
    $[4] = disableAnimation;
    $[5] = duration;
    $[6] = finalCollapsedValue;
    $[7] = id;
    $[8] = setCollapsed;
    $[9] = t8;
  } else {
    t8 = $[9];
  }
  let t9;
  if ($[10] !== children || $[11] !== t8) {
    t9 = /*#__PURE__*/jsx(CollapseContext.Provider, {
      value: t8,
      children: children
    });
    $[10] = children;
    $[11] = t8;
    $[12] = t9;
  } else {
    t9 = $[12];
  }
  let t10;
  if ($[13] !== className || $[14] !== t9) {
    t10 = /*#__PURE__*/jsx("div", {
      className: className,
      children: t9
    });
    $[13] = className;
    $[14] = t9;
    $[15] = t10;
  } else {
    t10 = $[15];
  }
  return t10;
};
function _temp() {}

export { Collapse, Collapse as default };
