// @jetbrains/ring-ui-built/components/dropdown/dropdown.js
import { Component, cloneElement } from 'react';
import classNames from 'classnames';
import joinDataTestAttributes from '../global/data-tests.js';
import { isArray } from '../global/typescript-utils.js';
import { s as styles, A as Anchor } from '../_helpers/anchor.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react-compiler-runtime';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';

class Dropdown extends Component {
  static defaultProps = {
    initShown: false,
    clickMode: true,
    hoverMode: false,
    hoverShowTimeOut: 300,
    hoverHideTimeOut: 600,
    disabled: false,
    onShow: () => {},
    onHide: () => {},
    onMouseEnter: () => {},
    onMouseLeave: () => {}
  };
  state = {
    show: this.props.initShown,
    pinned: false
  };
  onClick = () => {
    if (this.props.disabled) {
      return;
    }
    const {
      show,
      pinned
    } = this.state;
    let nextPinned = pinned;
    if (this.props.hoverMode) {
      if (!pinned) {
        nextPinned = true;
        if (show) {
          this.setState({
            pinned: true
          });
          return;
        }
      } else {
        nextPinned = false;
      }
    }
    this._toggle(!show, nextPinned);
  };
  onChildCloseAttempt = () => {
    let nextPinned = this.state.pinned;
    if (this.props.hoverMode) {
      nextPinned = false;
    }
    this._toggle(false, nextPinned);
  };
  hoverTimer;
  onMouseEnter = event => {
    if (this.props.disabled) {
      return;
    }
    this._clearTimer();
    this.props.onMouseEnter?.(event);
    this.hoverTimer = window.setTimeout(() => {
      if (!this.state.show) {
        this._toggle(true);
      }
    }, this.props.hoverShowTimeOut);
  };
  onMouseLeave = event => {
    if (this.props.disabled) {
      return;
    }
    this.props.onMouseLeave?.(event);
    if (this.state.pinned) {
      return;
    }
    this._clearTimer();
    this.hoverTimer = window.setTimeout(() => {
      if (this.state.show) {
        this._toggle(false);
      }
    }, this.props.hoverHideTimeOut);
  };
  handlePopupInteraction = () => {
    this.setState(({
      pinned
    }) => pinned ? null : {
      pinned: true
    });
  };
  toggle(show = !this.state.show) {
    this._toggle(show);
  }
  _toggle(show, pinned = this.state.pinned) {
    this.setState({
      show,
      pinned
    }, () => show ? this.props.onShow() : this.props.onHide());
  }
  _clearTimer() {
    if (this.hoverTimer) {
      clearTimeout(this.hoverTimer);
      this.hoverTimer = null;
    }
  }
  render() {
    const {
      show,
      pinned
    } = this.state;
    const {
      initShown,
      onShow,
      onHide,
      hoverShowTimeOut,
      hoverHideTimeOut,
      children,
      anchor,
      className,
      activeClassName,
      hoverMode,
      clickMode,
      'data-test': dataTest,
      disabled,
      ...restProps
    } = this.props;
    const classes = classNames(styles.dropdown, className, {
      [activeClassName !== null && activeClassName !== void 0 ? activeClassName : '']: activeClassName && show
    });
    let anchorElement;
    const active = hoverMode ? pinned : show;
    switch (typeof anchor) {
      case 'string':
        anchorElement = /*#__PURE__*/jsx(Anchor, {
          active: active,
          children: anchor
        });
        break;
      case 'function':
        anchorElement = anchor({
          active: show,
          pinned
        });
        break;
      default:
        if (isArray(anchor) || typeof anchor.type === 'string') {
          anchorElement = anchor;
        } else {
          anchorElement = /*#__PURE__*/cloneElement(anchor, {
            active
          });
        }
    }
    const childProps = {
      hidden: !show,
      onCloseAttempt: this.onChildCloseAttempt,
      onMouseDown: hoverMode ? this.handlePopupInteraction : undefined,
      onContextMenu: hoverMode ? this.handlePopupInteraction : undefined,
      dontCloseOnAnchorClick: true
    };
    return /*#__PURE__*/jsxs("div", {
      "data-test": joinDataTestAttributes('ring-dropdown', dataTest),
      ...restProps,
      onMouseEnter: hoverMode ? this.onMouseEnter : undefined,
      onMouseLeave: hoverMode ? this.onMouseLeave : undefined,
      className: classes,
      children: [clickMode ? /*#__PURE__*/jsx("div", {
        "data-test": "ring-dropdown-anchor-click-interceptor",
        className: styles.clickInterceptor,
        onClick: this.onClick
        // anchorElement should be a `button` or an `a`
        ,
        role: "presentation",
        children: anchorElement
      }) : anchorElement, typeof children === 'function' ? children(childProps) : /*#__PURE__*/cloneElement(children, childProps)]
    });
  }
}

export { Anchor, Dropdown as default };
