// @jetbrains/ring-ui-built/components/slider/slider.utils.js
import { isArray } from '../global/typescript-utils.js';

const HUNDRED = 100;
function toPercent(value, min, max) {
  return (value - min) * HUNDRED / (max - min);
}
function toValue(percent, min, max) {
  return (max - min) * percent + min;
}
function roundToStep(value, step, min) {
  return Math.round((value - min) / step) * step + min;
}
function calculateValue(ref, x, min, max, step) {
  if (!ref.current) {
    return null;
  }
  const {
    width,
    left
  } = ref.current.getBoundingClientRect();
  const value = toValue((x - left) / width, min, max);
  const precision = step.toString().split('.')?.[1]?.length;
  return Number(roundToStep(value, step, min).toFixed(precision));
}
function validateValue(value, min, max) {
  if (value <= min) {
    return min;
  }
  if (value >= max) {
    return max;
  }
  return value;
}
function toRange(value, min, max) {
  if (isArray(value)) {
    const nextValues = value.slice(0, 2).map(val => validateValue(val, min, max));
    if (nextValues[0] > nextValues[1]) {
      nextValues.reverse();
    }
    return nextValues;
  }
  return [validateValue(value, min, max)];
}
function adjustValues(values, ref, index, x, max, min, step) {
  const nextValue = calculateValue(ref, x, min, max, step);
  const nextValues = [...values];
  if (nextValue && !isNaN(nextValue)) {
    nextValues[index] = validateValue(nextValue, min, max);
  }
  return nextValues;
}
function calculateMarks(min, max, step) {
  const numMarks = Math.floor((max - min) / step) + 1;
  return Array.from({
    length: numMarks
  }, (_, index) => ({
    value: validateValue(min + step * index, min, max)
  }));
}

export { HUNDRED, adjustValues, calculateMarks, calculateValue, toPercent, toRange, validateValue };
