// @jetbrains/ring-ui-built/components/loader/loader.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import { L as LoaderCore, s as styles } from '../_helpers/loader-core.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import '../global/dom.js';

class Loader extends PureComponent {
  componentDidUpdate(prevProps) {
    if (this.loader) {
      if (!prevProps.stop && this.props.stop) {
        this.loader.stopAnimation();
      } else if (prevProps.stop && !this.props.stop) {
        this.loader.startAnimation();
      }
    }
  }
  componentWillUnmount() {
    this.loader?.destroy();
  }
  loader;
  initLoader = el => {
    if (el) {
      this.loader = new LoaderCore(el, this.props);
    }
  };
  render() {
    const {
      message,
      size,
      colors,
      'data-test': dataTest,
      stop,
      deterministic,
      squares,
      ...restProps
    } = this.props;
    return squares ? /*#__PURE__*/jsxs("div", {
      ...restProps,
      children: [/*#__PURE__*/jsxs("div", {
        className: classNames(styles.canvas, styles.squares),
        style: {
          '--ring-loader-color-1': colors?.[0] ? `rgb(${colors[0].r}, ${colors[0].g}, ${colors[0].b})` : undefined,
          '--ring-loader-color-2': colors?.[1] ? `rgb(${colors[1].r}, ${colors[1].g}, ${colors[1].b})` : undefined,
          '--ring-loader-color-3': colors?.[2] ? `rgb(${colors[2].r}, ${colors[2].g}, ${colors[2].b})` : undefined
        },
        children: [/*#__PURE__*/jsx("div", {
          className: classNames(styles.square, styles.outer)
        }), /*#__PURE__*/jsx("div", {
          className: classNames(styles.square, styles.middle)
        }), /*#__PURE__*/jsx("div", {
          className: classNames(styles.square, styles.inner)
        })]
      }), message && /*#__PURE__*/jsx("div", {
        className: styles.text,
        children: message
      })]
    }) : /*#__PURE__*/jsx("div", {
      ...restProps,
      ref: this.initLoader
    });
  }
}

export { Loader as default };
