// @jetbrains/ring-ui-built/components/tags-list/tags-list.js
import { Component, createElement } from 'react';
import classNames from 'classnames';
import Tag from '../tag/tag.js';
import { jsxs } from 'react/jsx-runtime';
import '@jetbrains/icons/close-12px';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../global/data-tests.js';

function noop() {}
/**
 * @name Tags List
 */
class TagsList extends Component {
  static defaultProps = {
    customTagComponent: null,
    canNotBeEmpty: false,
    disabled: false,
    handleClick: () => noop,
    handleRemove: () => noop
  };
  renderTag(tag, focusTag) {
    const TagComponent = this.props.customTagComponent || Tag;
    const readOnly = this.props.disabled || tag.readOnly || this.props.canNotBeEmpty && this.props.tags.length === 1;
    const {
      tagClassName
    } = this.props;
    return /*#__PURE__*/createElement(TagComponent, {
      ...tag,
      key: tag.key,
      readOnly: readOnly,
      disabled: this.props.disabled || tag.disabled,
      focused: focusTag,
      onClick: this.props.handleClick(tag),
      onRemove: this.props.handleRemove(tag),
      className: tagClassName
    }, tag.label);
  }
  render() {
    const {
      children,
      className,
      customTagComponent,
      canNotBeEmpty,
      handleClick,
      tagClassName,
      handleRemove,
      tags,
      activeIndex,
      ...props
    } = this.props;
    const classes = classNames('ring-js-shortcuts', className);
    const tagsList = (this.props.tags || []).map((tag, index) => this.renderTag(tag, this.props.activeIndex === index));
    return /*#__PURE__*/jsxs("div", {
      "data-test": "ring-tags-list",
      className: classes,
      ...props,
      children: [tagsList, children]
    });
  }
}

export { TagsList as default };
