// @jetbrains/ring-ui-built/components/auth-dialog/auth-dialog.js
import { Component } from 'react';
import classNames from 'classnames';
import { jsx, jsxs } from 'react/jsx-runtime';
import '../_helpers/island.js';
import 'react-compiler-runtime';
import ContentWrapper from '../island/content.js';
import Dialog from '../dialog/dialog.js';
import { Button } from '../button/button.js';
import { H2 } from '../heading/heading.js';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import 'react-dom';
import '@jetbrains/icons/close';
import '../island/island.js';
import '../global/data-tests.js';
import '../island/header.js';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/dom.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../_helpers/button.classes.js';
import '../_helpers/heading.js';

var styles = {"dialog":"ring-auth-dialog-dialog","content":"ring-auth-dialog-content ring-global-font","button":"ring-auth-dialog-button","firstButton":"ring-auth-dialog-firstButton ring-auth-dialog-button","title":"ring-auth-dialog-title","logo":"ring-auth-dialog-logo","error":"ring-auth-dialog-error"};

class AuthDialog extends Component {
  static defaultProps = {
    loginCaption: 'Log in',
    loginToCaption: 'Log in to %serviceName%',
    tryAgainLabel: 'Try again',
    show: false,
    cancelOnEsc: true,
    confirmLabel: 'Log in',
    cancelLabel: 'Remain a guest',
    onConfirm: () => {},
    onCancel: () => {}
  };
  state = {
    retrying: false
  };
  componentDidMount() {
    window.addEventListener('online', this.onRetryPress);
  }
  componentWillUnmount() {
    window.removeEventListener('online', this.onRetryPress);
  }
  onEscPress = () => {
    if (this.props.cancelOnEsc) {
      this.props.onCancel();
    }
  };
  onRetryPress = async () => {
    if (!this.props.onTryAgain || this.state.retrying) {
      return;
    }
    this.setState({
      retrying: true
    });
    try {
      await this.props.onTryAgain();
    } catch (e) {
      // do nothing, error is handled in onTryAgain
    } finally {
      this.setState({
        retrying: false
      });
    }
  };
  render() {
    const {
      show,
      className,
      errorMessage,
      serviceImage,
      serviceName,
      loginCaption,
      loginToCaption,
      confirmLabel,
      cancelLabel,
      tryAgainLabel,
      onConfirm,
      onCancel,
      onTryAgain
    } = this.props;
    const {
      retrying
    } = this.state;
    const defaultTitle = serviceName ? loginToCaption : loginCaption;
    const title = (this.props.title || defaultTitle).replace('%serviceName%', serviceName !== null && serviceName !== void 0 ? serviceName : '').replace('{{serviceName}}', serviceName !== null && serviceName !== void 0 ? serviceName : '');
    return /*#__PURE__*/jsx(Dialog, {
      label: title,
      "data-test": "ring-auth-dialog",
      className: className,
      contentClassName: classNames(className, styles.dialog),
      onEscPress: this.onEscPress,
      show: show,
      trapFocus: true,
      children: /*#__PURE__*/jsx(ContentWrapper, {
        children: /*#__PURE__*/jsxs("div", {
          className: styles.content,
          children: [serviceImage && /*#__PURE__*/jsx("img", {
            alt: `${serviceName} logo`,
            className: styles.logo,
            src: serviceImage
          }), /*#__PURE__*/jsx(H2, {
            className: styles.title,
            children: title
          }), errorMessage && /*#__PURE__*/jsx("div", {
            className: styles.error,
            children: errorMessage
          }), /*#__PURE__*/jsx(Button, {
            primary: true,
            className: styles.firstButton,
            "data-test": "auth-dialog-confirm-button",
            onClick: onConfirm,
            children: confirmLabel
          }), onTryAgain && /*#__PURE__*/jsx(Button, {
            className: styles.button,
            "data-test": "auth-dialog-retry-button",
            onClick: () => this.onRetryPress(),
            loader: retrying,
            disabled: retrying,
            children: tryAgainLabel
          }), /*#__PURE__*/jsx(Button, {
            className: styles.button,
            "data-test": "auth-dialog-cancel-button",
            onClick: onCancel,
            children: cancelLabel
          })]
        })
      })
    });
  }
}

export { AuthDialog as default };
