// @jetbrains/ring-ui-built/components/dropdown-menu/dropdown-menu.js
import { c } from 'react-compiler-runtime';
import { forwardRef, useState, cloneElement } from 'react';
import List, { ActiveItemContext } from '../list/list.js';
import Dropdown from '../dropdown/dropdown.js';
import PopupMenu from '../popup-menu/popup-menu.js';
import getUID from '../global/get-uid.js';
import { A as Anchor } from '../_helpers/anchor.js';
import { isArray } from '../global/typescript-utils.js';
import { jsx } from 'react/jsx-runtime';
import 'classnames';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'util-deprecate';
import 'memoize-one';
import '../global/data-tests.js';
import '../global/schedule-raf.js';
import '../global/memoize.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../avatar/avatar.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../popup/popup.js';
import 'react-dom';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';

const defaultAriaLabel = 'Dropdown menu';
function DropdownAnchorWrapper(t0) {
  const $ = c(4);
  if ($[0] !== "0926e12cd4fd28a8347702e5d2e46598e80e29bc205f01f12156ff6e80fd64e8") {
    for (let $i = 0; $i < 4; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "0926e12cd4fd28a8347702e5d2e46598e80e29bc205f01f12156ff6e80fd64e8";
  }
  let t1;
  let t2;
  if ($[1] !== t0) {
    t2 = Symbol.for("react.early_return_sentinel");
    bb0: {
      const {
        anchor,
        pinned,
        active,
        activeListItemId,
        listId,
        ...restProps
      } = t0;
      const anchorAriaProps = {
        ...(listId ? {
          "aria-haspopup": true
        } : {}),
        ...(activeListItemId ? {
          "aria-activedescendant": activeListItemId,
          "aria-owns": listId
        } : {}),
        ...(active ? {
          "aria-expanded": true
        } : {})
      };
      const anchorProps = {
        active,
        pinned,
        ...restProps,
        ...anchorAriaProps
      };
      const anchorComponentProps = {
        ...anchorProps,
        pinned: `${anchorProps.pinned}`
      };
      if (typeof anchor === "string") {
        t2 = /*#__PURE__*/jsx(Anchor, {
          ...anchorComponentProps,
          children: anchor
        });
        break bb0;
      }
      if (typeof anchor === "function") {
        t2 = anchor({
          active,
          pinned,
          ...restProps
        }, anchorAriaProps);
        break bb0;
      }
      if (!isArray(anchor)) {
        t2 = /*#__PURE__*/cloneElement(anchor, typeof anchor.type === "string" ? anchorAriaProps : anchorComponentProps);
        break bb0;
      }
      t1 = /*#__PURE__*/jsx("div", {
        ...anchorAriaProps,
        children: anchor
      });
    }
    $[1] = t0;
    $[2] = t1;
    $[3] = t2;
  } else {
    t1 = $[2];
    t2 = $[3];
  }
  if (t2 !== Symbol.for("react.early_return_sentinel")) {
    return t2;
  }
  return t1;
}
function renderDropdownMenuChildren({
  children,
  popupMenuProps
}) {
  if (!children) {
    return /*#__PURE__*/jsx(PopupMenu, {
      ...popupMenuProps
    });
  }
  return popupProps => children({
    ...popupProps,
    ...popupMenuProps
  });
}
const DropdownMenu = /*#__PURE__*/forwardRef(function DropdownMenu(t0, forwardedRef) {
  const $ = c(35);
  if ($[0] !== "0926e12cd4fd28a8347702e5d2e46598e80e29bc205f01f12156ff6e80fd64e8") {
    for (let $i = 0; $i < 35; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "0926e12cd4fd28a8347702e5d2e46598e80e29bc205f01f12156ff6e80fd64e8";
  }
  let anchor;
  let ariaLabel;
  let children;
  let data;
  let id;
  let menuProps;
  let onSelect;
  let restDropdownProps;
  if ($[1] !== t0) {
    ({
      id,
      anchor,
      ariaLabel,
      data,
      onSelect,
      menuProps,
      children,
      ...restDropdownProps
    } = t0);
    $[1] = t0;
    $[2] = anchor;
    $[3] = ariaLabel;
    $[4] = children;
    $[5] = data;
    $[6] = id;
    $[7] = menuProps;
    $[8] = onSelect;
    $[9] = restDropdownProps;
  } else {
    anchor = $[2];
    ariaLabel = $[3];
    children = $[4];
    data = $[5];
    id = $[6];
    menuProps = $[7];
    onSelect = $[8];
    restDropdownProps = $[9];
  }
  const [uid] = useState(_temp);
  const listId = id || uid;
  let T0;
  let T1;
  let t1;
  let t2;
  let t3;
  if ($[10] !== anchor || $[11] !== ariaLabel || $[12] !== children || $[13] !== data || $[14] !== forwardedRef || $[15] !== listId || $[16] !== menuProps || $[17] !== onSelect || $[18] !== restDropdownProps) {
    const popupMenuProps = {
      ref: forwardedRef,
      id: listId,
      ariaLabel: ariaLabel || defaultAriaLabel,
      closeOnSelect: true,
      activateFirstItem: true,
      data,
      onSelect,
      ...menuProps
    };
    T1 = ActiveItemContext.Provider;
    T0 = Dropdown;
    if ($[24] !== anchor || $[25] !== listId) {
      t1 = t4 => {
        const {
          pinned,
          active,
          ...restAnchorProps
        } = t4;
        return /*#__PURE__*/jsx(ActiveItemContext.ValueContext.Consumer, {
          children: activeItemId => /*#__PURE__*/jsx(DropdownAnchorWrapper, {
            anchor: anchor,
            pinned: pinned,
            active: active,
            activeListItemId: activeItemId,
            listId: listId,
            ...restAnchorProps
          })
        });
      };
      $[24] = anchor;
      $[25] = listId;
      $[26] = t1;
    } else {
      t1 = $[26];
    }
    t2 = restDropdownProps;
    t3 = renderDropdownMenuChildren({
      children,
      popupMenuProps
    });
    $[10] = anchor;
    $[11] = ariaLabel;
    $[12] = children;
    $[13] = data;
    $[14] = forwardedRef;
    $[15] = listId;
    $[16] = menuProps;
    $[17] = onSelect;
    $[18] = restDropdownProps;
    $[19] = T0;
    $[20] = T1;
    $[21] = t1;
    $[22] = t2;
    $[23] = t3;
  } else {
    T0 = $[19];
    T1 = $[20];
    t1 = $[21];
    t2 = $[22];
    t3 = $[23];
  }
  let t4;
  if ($[27] !== T0 || $[28] !== t1 || $[29] !== t2 || $[30] !== t3) {
    t4 = /*#__PURE__*/jsx(T0, {
      anchor: t1,
      ...t2,
      children: t3
    });
    $[27] = T0;
    $[28] = t1;
    $[29] = t2;
    $[30] = t3;
    $[31] = t4;
  } else {
    t4 = $[31];
  }
  let t5;
  if ($[32] !== T1 || $[33] !== t4) {
    t5 = /*#__PURE__*/jsx(T1, {
      children: t4
    });
    $[32] = T1;
    $[33] = t4;
    $[34] = t5;
  } else {
    t5 = $[34];
  }
  return t5;
});
var DropdownMenu$1 = Object.assign(DropdownMenu, {
  ListProps: List.ListProps
});
function _temp() {
  return getUID("dropdown-menu-list");
}

export { DropdownMenu$1 as default };
