// @jetbrains/ring-ui-built/components/storage/storage-local.js
import alertService from '../alert-service/alert-service.js';
import 'react-dom/client';
import '../global/get-uid.js';
import '../alert/alert.js';
import 'react';
import 'classnames';
import '@jetbrains/icons/exception';
import '@jetbrains/icons/checkmark';
import '@jetbrains/icons/warning';
import '@jetbrains/icons/close';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import 'react/jsx-runtime';
import '../loader-inline/loader-inline.js';
import '../global/data-tests.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../_helpers/theme.js';
import '../alert/container.js';
import 'react-dom';

/**
 * @return {LocalStorage}
 * @param {{type: string}} config Set to "session" to use sessionStorage
 * @constructor
 */
class LocalStorage {
  static async safePromise(resolver) {
    try {
      return await new Promise(resolver);
    } catch (e) {
      if (e instanceof Error && e.name === 'NS_ERROR_FILE_CORRUPTED') {
        alertService.error('Sorry, it looks like your browser storage is corrupted. ' + 'Please clear your storage by going to Tools -> Clear Recent History -> Cookies' + ' and setting time range to "Everything". This will remove the corrupted browser storage across all sites.');
      }
      throw e;
    }
  }
  storageType;
  constructor(config = {}) {
    this.storageType = config.type === 'session' ? 'sessionStorage' : 'localStorage';
  }
  /**
   * @param {string} name
   * @return {Promise}
   */
  get(name) {
    return LocalStorage.safePromise(resolve => {
      const value = window[this.storageType].getItem(name);
      if (value) {
        try {
          resolve(JSON.parse(value));
        } catch (e) {
          resolve(value);
        }
      } else {
        resolve(null);
      }
    });
  }
  /**
   * @param {string} name
   * @param {object} value
   * @return {Promise}
   */
  set(name, value) {
    return LocalStorage.safePromise(resolve => {
      window[this.storageType].setItem(name, JSON.stringify(value));
      resolve(value);
    });
  }
  /**
   * @param {string} name
   * @return {Promise}
   */
  remove(name) {
    const storageType = this.storageType;
    return LocalStorage.safePromise(resolve => {
      if (Object.prototype.hasOwnProperty.call(window[storageType], name)) {
        window[storageType].removeItem(name);
      }
      resolve();
    });
  }
  /**
   * @param callback
   * @return {Promise}s
   */
  each(callback) {
    const storageType = this.storageType;
    return LocalStorage.safePromise(resolve => {
      const promises = [];
      for (const item in window[storageType]) {
        if (Object.prototype.hasOwnProperty.call(window[storageType], item)) {
          const value = window[storageType].getItem(item);
          let resolvedValue = null;
          if (value) {
            try {
              resolvedValue = JSON.parse(value);
            } catch (e) {
              resolvedValue = value;
            }
          }
          promises.push(Promise.resolve(callback(item, resolvedValue)));
        }
      }
      resolve(Promise.all(promises));
    });
  }
  /**
   * @param {string} name
   * @param {Function} callback
   * @return {Function}
   */
  on(name, callback) {
    function handleStorage(e) {
      if (e.key === name) {
        if (e.newValue) {
          try {
            callback(JSON.parse(e.newValue));
          } catch (err) {
            callback(e.newValue);
          }
        } else {
          callback(null);
        }
      }
    }
    window.addEventListener('storage', handleStorage, false);
    return () => window.removeEventListener('storage', handleStorage, false);
  }
}

export { LocalStorage as default };
