// @jetbrains/ring-ui-built/components/upload/upload.js
import { c } from 'react-compiler-runtime';
import { forwardRef, useRef, useState, useImperativeHandle } from 'react';
import classNames from 'classnames';
import attachmentIcon from '@jetbrains/icons/attachment';
import Icon from '../icon/icon.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';

var styles = {"upload":"ring-upload-upload","success":"ring-upload-success","error":"ring-upload-error","dragOver":"ring-upload-dragOver","disabled":"ring-upload-disabled","invisibleFileInput":"ring-upload-invisibleFileInput","attachmentIcon":"ring-upload-attachmentIcon"};

const defaultRenderIcon = () => /*#__PURE__*/jsx(Icon, {
  className: styles.attachmentIcon,
  glyph: attachmentIcon
});
function UploadInner(t0) {
  const $ = c(29);
  if ($[0] !== "7d5a4ecc65f01d46fa3e08e16dbed39a1cd6f740a025c62cc404b96900b00084") {
    for (let $i = 0; $i < 29; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "7d5a4ecc65f01d46fa3e08e16dbed39a1cd6f740a025c62cc404b96900b00084";
  }
  const {
    children,
    className,
    onFilesSelected,
    onFilesRejected,
    validate: t1,
    variant: t2,
    multiple,
    renderIcon: t3,
    accept,
    disabled,
    forwardedRef
  } = t0;
  const validate = t1 === undefined ? _temp : t1;
  const variant = t2 === undefined ? "empty" : t2;
  const renderIcon = t3 === undefined ? defaultRenderIcon : t3;
  const fileInputRef = useRef(null);
  const [dragOver, setDragOver] = useState(false);
  let t4;
  let t5;
  if ($[1] === Symbol.for("react.memo_cache_sentinel")) {
    t4 = () => ({
      openFilePicker: () => fileInputRef.current?.click()
    });
    t5 = [];
    $[1] = t4;
    $[2] = t5;
  } else {
    t4 = $[1];
    t5 = $[2];
  }
  useImperativeHandle(forwardedRef, t4, t5);
  let t6;
  if ($[3] !== onFilesRejected || $[4] !== onFilesSelected || $[5] !== validate) {
    t6 = files => {
      if (!files.length) {
        return;
      }
      const rejected = files.filter(file => !validate(file));
      if (rejected.length > 0) {
        onFilesRejected?.(files);
        return;
      }
      onFilesSelected(files);
    };
    $[3] = onFilesRejected;
    $[4] = onFilesSelected;
    $[5] = validate;
    $[6] = t6;
  } else {
    t6 = $[6];
  }
  const handleSelectedFiles = t6;
  let t7;
  if ($[7] === Symbol.for("react.memo_cache_sentinel")) {
    t7 = () => setDragOver(true);
    $[7] = t7;
  } else {
    t7 = $[7];
  }
  const onDragEnter = t7;
  const onDragOver = _temp2;
  let t8;
  if ($[8] === Symbol.for("react.memo_cache_sentinel")) {
    t8 = () => setDragOver(false);
    $[8] = t8;
  } else {
    t8 = $[8];
  }
  const onDragLeave = t8;
  let t9;
  if ($[9] !== handleSelectedFiles) {
    t9 = () => {
      setDragOver(false);
      if (fileInputRef.current?.files) {
        handleSelectedFiles(Array.from(fileInputRef.current.files));
      }
    };
    $[9] = handleSelectedFiles;
    $[10] = t9;
  } else {
    t9 = $[10];
  }
  const onInputChange = t9;
  const t10 = variant === "success";
  const t11 = variant === "error";
  let t12;
  if ($[11] !== className || $[12] !== disabled || $[13] !== dragOver || $[14] !== t10 || $[15] !== t11) {
    t12 = classNames(className, styles.upload, {
      [styles.disabled]: disabled,
      [styles.dragOver]: dragOver,
      [styles.success]: t10,
      [styles.error]: t11
    });
    $[11] = className;
    $[12] = disabled;
    $[13] = dragOver;
    $[14] = t10;
    $[15] = t11;
    $[16] = t12;
  } else {
    t12 = $[16];
  }
  let t13;
  if ($[17] !== accept || $[18] !== disabled || $[19] !== multiple || $[20] !== onInputChange) {
    t13 = /*#__PURE__*/jsx("input", {
      onDragEnter: onDragEnter,
      onDragOver: onDragOver,
      onDragLeave: onDragLeave,
      disabled: disabled,
      ref: fileInputRef,
      "data-test": "ring-file-input",
      multiple: multiple,
      accept: accept,
      onChange: onInputChange,
      type: "file",
      autoComplete: "off",
      "aria-label": "file-picker",
      className: styles.invisibleFileInput
    });
    $[17] = accept;
    $[18] = disabled;
    $[19] = multiple;
    $[20] = onInputChange;
    $[21] = t13;
  } else {
    t13 = $[21];
  }
  let t14;
  if ($[22] !== renderIcon) {
    t14 = renderIcon();
    $[22] = renderIcon;
    $[23] = t14;
  } else {
    t14 = $[23];
  }
  let t15;
  if ($[24] !== children || $[25] !== t12 || $[26] !== t13 || $[27] !== t14) {
    t15 = /*#__PURE__*/jsxs("div", {
      className: t12,
      "data-test": "ring-upload",
      children: [t13, t14, children]
    });
    $[24] = children;
    $[25] = t12;
    $[26] = t13;
    $[27] = t14;
    $[28] = t15;
  } else {
    t15 = $[28];
  }
  return t15;
}
function _temp2(e) {
  return e.preventDefault();
}
function _temp() {
  return true;
}
const Upload = /*#__PURE__*/forwardRef(function Upload(props, ref) {
  const $ = c(4);
  if ($[0] !== "7d5a4ecc65f01d46fa3e08e16dbed39a1cd6f740a025c62cc404b96900b00084") {
    for (let $i = 0; $i < 4; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "7d5a4ecc65f01d46fa3e08e16dbed39a1cd6f740a025c62cc404b96900b00084";
  }
  let t0;
  if ($[1] !== props || $[2] !== ref) {
    t0 = /*#__PURE__*/jsx(UploadInner, {
      ...props,
      forwardedRef: ref
    });
    $[1] = props;
    $[2] = ref;
    $[3] = t0;
  } else {
    t0 = $[3];
  }
  return t0;
});

export { Upload, Upload as default };
