// @jetbrains/ring-ui-built/components/message/message.js
import { Component } from 'react';
import classNames from 'classnames';
import gift from '@jetbrains/icons/gift';
import Popup from '../popup/popup.js';
import { Directions } from '../popup/popup.consts.js';
import Icon from '../icon/icon.js';
import { Button } from '../button/button.js';
import { I18nContext } from '../i18n/i18n-context.js';
import { T as Theme, d as darkStyles, W as WithThemeClasses, a as ThemeProvider } from '../_helpers/theme.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/data-tests.js';
import '../tab-trap/tab-trap.js';
import 'react-compiler-runtime';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../_helpers/button.classes.js';
import '../i18n/i18n.js';

var styles = {"container":"ring-message-container","message":"ring-message-message","icon":"ring-message-icon","title":"ring-message-title ring-global-font-lower ring-global-font","description":"ring-message-description","gotIt":"ring-message-gotIt","tail":"ring-message-tail"};

const UNIT = 8;
const TAIL_SIZE = 11;
const getTailOffsets = offset => ({
  [Directions.BOTTOM_RIGHT]: {
    top: 0,
    left: offset + UNIT,
    transform: 'rotate(135deg)'
  },
  [Directions.BOTTOM_LEFT]: {
    top: 0,
    right: offset - UNIT - TAIL_SIZE,
    transform: 'rotate(135deg)'
  },
  [Directions.BOTTOM_CENTER]: {
    top: 0,
    left: offset + UNIT,
    transform: 'rotate(135deg)'
  },
  [Directions.TOP_RIGHT]: {
    bottom: -TAIL_SIZE,
    left: offset - UNIT,
    transform: 'rotate(-45deg)'
  },
  [Directions.TOP_LEFT]: {
    bottom: -TAIL_SIZE,
    right: offset + UNIT - TAIL_SIZE,
    transform: 'rotate(-45deg)'
  },
  [Directions.TOP_CENTER]: {
    bottom: -TAIL_SIZE,
    left: offset - UNIT,
    transform: 'rotate(-45deg)'
  },
  [Directions.RIGHT_TOP]: {
    bottom: offset + UNIT - TAIL_SIZE,
    left: 0,
    transform: 'rotate(45deg)'
  },
  [Directions.RIGHT_BOTTOM]: {
    top: offset - UNIT,
    left: 0,
    transform: 'rotate(45deg)'
  },
  [Directions.RIGHT_CENTER]: {
    top: offset - UNIT,
    left: 0,
    transform: 'rotate(45deg)'
  },
  [Directions.LEFT_TOP]: {
    bottom: offset - UNIT - TAIL_SIZE,
    right: -TAIL_SIZE,
    transform: 'rotate(-135deg)'
  },
  [Directions.LEFT_BOTTOM]: {
    top: offset + UNIT,
    right: -TAIL_SIZE,
    transform: 'rotate(-135deg)'
  },
  [Directions.LEFT_CENTER]: {
    top: offset + UNIT,
    right: -TAIL_SIZE,
    transform: 'rotate(-135deg)'
  }
});
/**
 * Displays a popup containing a message.
 */
class Message extends Component {
  static defaultProps = {
    icon: gift,
    directions: [Directions.TOP_RIGHT, Directions.TOP_LEFT, Directions.TOP_CENTER, Directions.BOTTOM_RIGHT, Directions.BOTTOM_LEFT, Directions.BOTTOM_CENTER, Directions.RIGHT_TOP, Directions.RIGHT_BOTTOM, Directions.RIGHT_CENTER, Directions.LEFT_TOP, Directions.LEFT_BOTTOM, Directions.LEFT_CENTER],
    theme: Theme.DARK
  };
  state = {};
  static Directions = Directions;
  static PopupProps = Popup.PopupProps;
  _onDirectionChange = direction => this.setState({
    direction
  });
  popup;
  node;
  popupRef = el => {
    this.popup = el;
    this.node = this.popup?.node;
  };
  getTailOffset() {
    const DEFAULT_OFFSET = 32;
    const {
      popupProps
    } = this.props;
    if (this.props.tailOffset) {
      return this.props.tailOffset;
    }
    const anchor = popupProps?.anchorElement || this.popup?.parent;
    if (!anchor) {
      return DEFAULT_OFFSET;
    }
    const offset = Math.floor(anchor.offsetWidth / 2);
    const isOpenedToRight = this.state.direction && [Directions.TOP_RIGHT, Directions.BOTTOM_RIGHT].includes(this.state.direction);
    if (popupProps?.left && isOpenedToRight) {
      return offset - popupProps?.left;
    }
    return offset;
  }
  render() {
    const {
      children,
      className,
      tailClassName,
      title,
      icon,
      popupProps,
      buttonProps,
      onClose,
      onDismiss,
      translations,
      theme
    } = this.props;
    const classes = classNames(styles.message, className, {
      [darkStyles.dark]: theme === Theme.DARK
    });
    const tailClasses = classNames(styles.tail, tailClassName);
    const popupDirections = this.props.direction ? [this.props.direction] : this.props.directions;
    const {
      direction
    } = this.state;
    return /*#__PURE__*/jsx(I18nContext.Consumer, {
      children: ({
        translate
      }) => /*#__PURE__*/jsx(WithThemeClasses, {
        theme: theme,
        children: themeClasses => {
          var _translations$gotIt, _translations$dismiss;
          return /*#__PURE__*/jsx(Popup, {
            ref: this.popupRef,
            hidden: false,
            directions: popupDirections,
            className: classNames(classes, themeClasses),
            offset: UNIT * 2,
            onDirectionChange: this._onDirectionChange,
            ...popupProps,
            children: /*#__PURE__*/jsxs(ThemeProvider, {
              theme: theme,
              passToPopups: true,
              children: [direction && /*#__PURE__*/jsx("div", {
                className: tailClasses,
                style: getTailOffsets(this.getTailOffset())[direction]
              }), icon && /*#__PURE__*/jsx(Icon, {
                className: styles.icon,
                glyph: icon
              }), title && /*#__PURE__*/jsx("h1", {
                "data-test": "rgMessageTitle",
                className: styles.title,
                children: title
              }), children && /*#__PURE__*/jsx("div", {
                className: styles.description,
                children: children
              }), (onClose || buttonProps) && /*#__PURE__*/jsx(Button, {
                className: styles.gotIt,
                onClick: onClose,
                primary: true,
                ...buttonProps,
                children: (_translations$gotIt = translations?.gotIt) !== null && _translations$gotIt !== void 0 ? _translations$gotIt : translate('gotIt')
              }), onDismiss && /*#__PURE__*/jsx(Button, {
                onClick: onDismiss,
                inline: true,
                children: (_translations$dismiss = translations?.dismiss) !== null && _translations$dismiss !== void 0 ? _translations$dismiss : translate('dismiss')
              })]
            })
          });
        }
      })
    });
  }
}

export { Message as default };
