// @jetbrains/ring-ui-built/components/content-layout/content-layout.js
import { Component, Children, cloneElement } from 'react';
import classNames from 'classnames';
import { s as styles, S as Sidebar } from '../_helpers/sidebar.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react-waypoint';

class ContentLayout extends Component {
  static defaultProps = {
    responsive: true
  };
  state = {
    contentNode: null
  };
  saveContentNode = contentNode => {
    this.setState({
      contentNode
    });
  };
  render() {
    const {
      children,
      className,
      contentClassName,
      responsive,
      ...restProps
    } = this.props;
    const classes = classNames(styles.contentLayout, className, {
      [styles.contentLayoutResponsive]: responsive
    });
    const contentClasses = classNames(styles.contentLayoutContent, contentClassName);
    const childrenArray = Children.toArray(children);
    const sidebarChild = childrenArray.filter(child => !!child && typeof child === 'object' && 'type' in child && child.type === Sidebar)[0];
    const sidebar = sidebarChild && /*#__PURE__*/cloneElement(sidebarChild, {
      contentNode: this.state.contentNode
    });
    const contentChildren = childrenArray.filter(child => child !== sidebarChild);
    return /*#__PURE__*/jsxs("div", {
      "data-test": "content-layout",
      ...restProps,
      className: classes,
      children: [sidebar, /*#__PURE__*/jsx("main", {
        className: contentClasses,
        ref: this.saveContentNode,
        children: contentChildren
      })]
    });
  }
}

export { Sidebar, ContentLayout as default };
