// @jetbrains/ring-ui-built/components/global/dom.js
/**
 * @name DOM
 */
const getStyles = window.getComputedStyle.bind(window);
function isMounted(node) {
  if (node === document) {
    return true;
  }
  return node instanceof Node && document.documentElement.contains(node.parentNode);
}
const rectStub = {
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  width: 0,
  height: 0
};
function getRect(node) {
  if (node instanceof Range || node && isMounted(node)) {
    const {
      top,
      right,
      bottom,
      left,
      width,
      height
    } = node.getBoundingClientRect();
    return {
      top,
      right,
      bottom,
      left,
      width,
      height
    };
  }
  return Object.assign({}, rectStub);
}
function getPixelRatio() {
  return 'devicePixelRatio' in window ? window.devicePixelRatio : 1;
}
function getWindowHeight() {
  return window.innerHeight;
}
function getWindowWidth() {
  return window.innerWidth;
}
function isNodeInVisiblePartOfPage(node) {
  const {
    top,
    bottom,
    left,
    right
  } = getRect(node);
  return !(bottom < 0 || right < 0 || getWindowHeight() - top < 0 || getWindowWidth() - left < 0);
}
function getDocumentScrollTop() {
  return document.documentElement && document.documentElement.scrollTop || document.body.scrollTop;
}
function getDocumentScrollLeft() {
  return document.documentElement && document.documentElement.scrollLeft || document.body.scrollLeft;
}
const applyMethodToClasses = method => (classList, classes = '') => {
  classes.split(/\s+/g).filter(className => !!className).forEach(className => classList[method](className));
};
const addClasses = applyMethodToClasses('add');
const removeClasses = applyMethodToClasses('remove');
const toggleClasses = (classList, classes) => Object.entries(classes).forEach(([className, on]) => classList.toggle(className, on));
function setRootStyleProperties(properties = {}) {
  const rootStyle = document.documentElement.style;
  Object.entries(properties).forEach(([key, value]) => {
    rootStyle.setProperty(key, value);
  });
}
function resetRootStyleProperties(properties = {}) {
  const rootStyle = document.documentElement.style;
  Object.keys(properties).forEach(key => rootStyle.removeProperty(key));
}
class Listeners {
  _all = new Set();
  add(el, event, handler, options) {
    el.addEventListener(event, handler, options);
    const dispatchFn = () => el.removeEventListener(event, handler, options);
    this._all.add(dispatchFn);
    return dispatchFn;
  }
  remove(fn) {
    fn();
    this._all.delete(fn);
  }
  removeAll() {
    this._all.forEach(fn => this.remove(fn));
  }
}
// Synthetic events from Combokeys#trigger are plain objects
function preventDefault(e) {
  if (e.preventDefault) {
    e.preventDefault();
  }
}

export { Listeners, addClasses, applyMethodToClasses, getDocumentScrollLeft, getDocumentScrollTop, getPixelRatio, getRect, getStyles, getWindowHeight, getWindowWidth, isMounted, isNodeInVisiblePartOfPage, preventDefault, removeClasses, resetRootStyleProperties, setRootStyleProperties, toggleClasses };
