// @jetbrains/ring-ui-built/components/global/rerender-hoc.js
import { c } from 'react-compiler-runtime';
import { forwardRef, Component } from 'react';
import { createComposedRef } from './compose-refs.js';
import { jsx } from 'react/jsx-runtime';
import 'memoize-one';

function rerenderHOC(ComposedComponent) {
  class Rerenderer extends Component {
    state = this.props.props;
    composedRef = createComposedRef();
    render() {
      const ref = this.composedRef(this.props.forwardedRef);
      return /*#__PURE__*/jsx(ComposedComponent, {
        ...this.state,
        ref: instance => ref(instance ? {
          ...instance,
          rerender: this.setState.bind(this)
        } : null)
      });
    }
  }
  return /*#__PURE__*/forwardRef(function RerendererForwardRef(props, ref) {
    const $ = c(4);
    if ($[0] !== "ac36aeaece54c7d00df3e23ff1d0deb675c0bc1bf4659806ffc7c8497029aeb7") {
      for (let $i = 0; $i < 4; $i += 1) {
        $[$i] = Symbol.for("react.memo_cache_sentinel");
      }
      $[0] = "ac36aeaece54c7d00df3e23ff1d0deb675c0bc1bf4659806ffc7c8497029aeb7";
    }
    let t0;
    if ($[1] !== props || $[2] !== ref) {
      t0 = /*#__PURE__*/jsx(Rerenderer, {
        props: props,
        forwardedRef: ref
      });
      $[1] = props;
      $[2] = ref;
      $[3] = t0;
    } else {
      t0 = $[3];
    }
    return t0;
  });
}

export { rerenderHOC as default };
