// @jetbrains/ring-ui-built/components/control-label/control-label.js
import { c } from 'react-compiler-runtime';
import 'react';
import classNames from 'classnames';
import { jsx } from 'react/jsx-runtime';

var styles = {"label":"ring-control-label-label","formLabel":"ring-control-label-formLabel","secondaryLabel":"ring-control-label-secondaryLabel","disabledLabel":"ring-control-label-disabledLabel"};

var LabelType;
(function (LabelType) {
  LabelType["SECONDARY"] = "secondary";
  LabelType["FORM"] = "form";
})(LabelType || (LabelType = {}));
const classNameByType = {
  [LabelType.SECONDARY]: styles.secondaryLabel,
  [LabelType.FORM]: styles.formLabel
};
const ControlLabel = t0 => {
  const $ = c(13);
  if ($[0] !== "bb4c1a3448a8ab687255dd795b9f172a7ee374fecb5dfc264c1d31d9ce69f171") {
    for (let $i = 0; $i < 13; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "bb4c1a3448a8ab687255dd795b9f172a7ee374fecb5dfc264c1d31d9ce69f171";
  }
  let children;
  let disabled;
  let rest;
  let t1;
  if ($[1] !== t0) {
    ({
      children,
      type: t1,
      disabled,
      ...rest
    } = t0);
    $[1] = t0;
    $[2] = children;
    $[3] = disabled;
    $[4] = rest;
    $[5] = t1;
  } else {
    children = $[2];
    disabled = $[3];
    rest = $[4];
    t1 = $[5];
  }
  const type = t1 === undefined ? LabelType.SECONDARY : t1;
  const t2 = classNameByType[type];
  let t3;
  if ($[6] !== disabled || $[7] !== t2) {
    t3 = classNames(styles.label, t2, {
      [styles.disabledLabel]: disabled
    });
    $[6] = disabled;
    $[7] = t2;
    $[8] = t3;
  } else {
    t3 = $[8];
  }
  let t4;
  if ($[9] !== children || $[10] !== rest || $[11] !== t3) {
    t4 = /*#__PURE__*/jsx("label", {
      className: t3,
      ...rest,
      children: children
    });
    $[9] = children;
    $[10] = rest;
    $[11] = t3;
    $[12] = t4;
  } else {
    t4 = $[12];
  }
  return t4;
};

export { ControlLabel, LabelType, ControlLabel as default };
