// @jetbrains/ring-ui-built/components/grid/col.js
import { Component } from 'react';
import classNames from 'classnames';
import { s as styles } from '../_helpers/grid.js';
import { jsx } from 'react/jsx-runtime';

const classMap = {
  xs: 'col-xs',
  sm: 'col-sm',
  md: 'col-md',
  lg: 'col-lg',
  xsOffset: 'col-xs-offset',
  smOffset: 'col-sm-offset',
  mdOffset: 'col-md-offset',
  lgOffset: 'col-lg-offset'
};
/**
 * Converts props like "xs=9 xsOffset={2}" to classes "col-xs-9 col-xs-offset-2"
 * @param {Object} props incoming props
 * @mockReturnValue {Array} result classnames
 */
function getClassNames(props) {
  return Object.entries(props).filter(([key]) => classMap[key]).map(([key, value]) => styles[Number.isInteger(value) ? `${classMap[key]}-${value}` : classMap[key]]);
}
class Col extends Component {
  render() {
    const {
      children,
      className,
      reverse,
      ...restProps
    } = this.props;
    const classes = classNames(styles.col, className, getClassNames(restProps), {
      [styles.reverse]: reverse
    });
    return /*#__PURE__*/jsx("div", {
      className: classes,
      "data-test": "ring-grid-column",
      children: children
    });
  }
}

export { Col as default };
