// @jetbrains/youtrack-apps-tools/dist/src/cli/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = run;
const i18n_1 = require("../../lib/i18n/i18n");
const exit_1 = require("../../lib/cli/exit");
const parseargv_1 = require("../../lib/cli/parseargv");
const list_1 = require("./list");
const download_1 = require("./download");
const upload_1 = require("./upload");
const resolve_1 = require("../../lib/net/resolve");
const validate_1 = require("./validate");
const options = {
    list: list_1.list,
    download: download_1.download,
    upload: upload_1.upload,
    validate: validate_1.validate,
};
async function run(argv = process.argv) {
    const args = (0, parseargv_1.parse)(argv);
    const { YOUTRACK_HOST, YOUTRACK_API_TOKEN } = process.env;
    const config = {
        host: args.host || YOUTRACK_HOST || null,
        token: args.token || YOUTRACK_API_TOKEN || null,
        output: args.output || null,
        overwrite: args.overwrite || null,
        manifest: args.manifest || null,
        schema: args.schema || null,
        open: args.open || null,
        cwd: process.cwd(),
    };
    if (args.version || args.v) {
        return printVersion();
    }
    const option = args._[0];
    switch (option) {
        case 'list':
        case 'download':
        case 'upload':
            await checkRequiredParams(['host', 'token'], args, async () => {
                const executable = options[option];
                await executable(config, args._.slice(1)[0]);
            });
            return;
        case 'validate':
            await options['validate'](config, args._.slice(1)[0]);
            return;
        case 'version':
            printVersion();
            return;
        default:
            printHelp();
            return;
    }
    function printHelp() {
        br();
        printLine((0, i18n_1.i18n)('list     --host --token                      '), (0, i18n_1.i18n)('View a list of installed apps'));
        printLine((0, i18n_1.i18n)('download <app> [--output, --overwrite]       '), (0, i18n_1.i18n)('Download an app'));
        printLine((0, i18n_1.i18n)('upload   <directory>                         '), (0, i18n_1.i18n)('Upload app to server'));
        printLine((0, i18n_1.i18n)('validate <directory> [--manifest, --schema]  '), (0, i18n_1.i18n)('Validate manifest'));
        br();
        console.log((0, i18n_1.i18n)('One can also provide host and token via environment variables $YOUTRACK_HOST and $YOUTRACK_API_TOKEN.'));
        function br() {
            console.log('');
        }
        function printLine(option, description) {
            console.log('    ' + option + '   ' + description);
        }
    }
    async function checkRequiredParams(required, args, fn) {
        function allParamsProvided(params, args) {
            return params.every(param => {
                if ((!args.hasOwnProperty(param) || !args[param]) && !config[param]) {
                    if (param === 'token') {
                        const createTokenUrl = `${(0, resolve_1.resolve)(config.host, 'users/me?tab=account-security').href}`;
                        (0, exit_1.exit)(new Error((0, i18n_1.i18n)(`Token is required. Please create one at ${createTokenUrl}`)));
                    }
                    else {
                        (0, exit_1.exit)(new Error((0, i18n_1.i18n)('Option "--' + param + '" is required')));
                    }
                    return false;
                }
                return true;
            });
        }
        if (allParamsProvided(required, args))
            await fn();
    }
    function printVersion() {
        console.log(require('../../package.json').version);
    }
}
