// @jetbrains/ring-ui-built/components/popup/popup.consts.js
var Directions;
(function (Directions) {
  Directions["BOTTOM_RIGHT"] = "BOTTOM_RIGHT";
  Directions["BOTTOM_LEFT"] = "BOTTOM_LEFT";
  Directions["BOTTOM_CENTER"] = "BOTTOM_CENTER";
  Directions["TOP_LEFT"] = "TOP_LEFT";
  Directions["TOP_RIGHT"] = "TOP_RIGHT";
  Directions["TOP_CENTER"] = "TOP_CENTER";
  Directions["RIGHT_TOP"] = "RIGHT_TOP";
  Directions["RIGHT_BOTTOM"] = "RIGHT_BOTTOM";
  Directions["RIGHT_CENTER"] = "RIGHT_CENTER";
  Directions["LEFT_TOP"] = "LEFT_TOP";
  Directions["LEFT_BOTTOM"] = "LEFT_BOTTOM";
  Directions["LEFT_CENTER"] = "LEFT_CENTER";
})(Directions || (Directions = {}));
/**
 * When positioning a popup, directions will be tried in the listed order.
 * @type {Array.<string>}
 */
const DEFAULT_DIRECTIONS = [Directions.BOTTOM_RIGHT, Directions.BOTTOM_LEFT, Directions.TOP_LEFT, Directions.TOP_RIGHT, Directions.RIGHT_TOP, Directions.RIGHT_BOTTOM, Directions.LEFT_TOP, Directions.LEFT_BOTTOM];
/**
 * @enum {number}
 */
var Dimension;
(function (Dimension) {
  Dimension[Dimension["MARGIN"] = 16] = "MARGIN";
  Dimension[Dimension["BORDER_WIDTH"] = 1] = "BORDER_WIDTH";
})(Dimension || (Dimension = {}));
var MinWidth;
(function (MinWidth) {
  MinWidth[MinWidth["TARGET"] = -1] = "TARGET";
})(MinWidth || (MinWidth = {}));
var MaxHeight;
(function (MaxHeight) {
  MaxHeight[MaxHeight["SCREEN"] = -1] = "SCREEN";
})(MaxHeight || (MaxHeight = {}));
var Display;
(function (Display) {
  Display[Display["HIDDEN"] = 0] = "HIDDEN";
  Display[Display["SHOWING"] = 1] = "SHOWING";
  Display[Display["SHOWN"] = 2] = "SHOWN";
})(Display || (Display = {}));

export { DEFAULT_DIRECTIONS, Dimension, Directions, Display, MaxHeight, MinWidth };
