// @jetbrains/ring-ui-built/components/contenteditable/contenteditable.js
import { c } from 'react-compiler-runtime';
import { Component } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { jsx } from 'react/jsx-runtime';

function noop() {}
class ContentEditableBase extends Component {
  static defaultProps = {
    disabled: false,
    tabIndex: 0,
    onInput: noop,
    onComponentUpdate: noop
  };
  shouldComponentUpdate(nextProps) {
    // eslint-disable-next-line no-underscore-dangle
    return nextProps.disabled !== this.props.disabled || nextProps.__html !== this.props.__html;
  }
  componentDidUpdate(prevProps) {
    this.props.onComponentUpdate(prevProps);
  }
  render() {
    const {
      __html,
      onComponentUpdate,
      disabled,
      tabIndex,
      inputRef,
      ...props
    } = this.props;
    return /*#__PURE__*/jsx("div", {
      ...props,
      ref: inputRef
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      ,
      disabled: disabled,
      role: "textbox",
      tabIndex: disabled ? undefined : tabIndex,
      contentEditable: !this.props.disabled,
      dangerouslySetInnerHTML: {
        __html
      }
    });
  }
}
const ContentEditable = t0 => {
  const $ = c(9);
  if ($[0] !== "a16faaa5b9814b3292a3bdccbd3ed2550092bea3ba98e71aa85f18aa005eb348") {
    for (let $i = 0; $i < 9; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "a16faaa5b9814b3292a3bdccbd3ed2550092bea3ba98e71aa85f18aa005eb348";
  }
  let children;
  let props;
  if ($[1] !== t0) {
    ({
      children,
      ...props
    } = t0);
    $[1] = t0;
    $[2] = children;
    $[3] = props;
  } else {
    children = $[2];
    props = $[3];
  }
  let t1;
  if ($[4] !== children) {
    t1 = renderToStaticMarkup(children);
    $[4] = children;
    $[5] = t1;
  } else {
    t1 = $[5];
  }
  let t2;
  if ($[6] !== props || $[7] !== t1) {
    t2 = /*#__PURE__*/jsx(ContentEditableBase, {
      ...props,
      __html: t1
    });
    $[6] = props;
    $[7] = t1;
    $[8] = t2;
  } else {
    t2 = $[8];
  }
  return t2;
};

export { ContentEditable as default };
