// @jetbrains/ring-ui-built/components/confirm/confirm.js
import { PureComponent } from 'react';
import Dialog from '../dialog/dialog.js';
import { Button } from '../button/button.js';
import 'classnames';
import { jsxs, jsx } from 'react/jsx-runtime';
import '../_helpers/island.js';
import HeaderWrapper from '../island/header.js';
import ContentWrapper from '../island/content.js';
import Panel from '../panel/panel.js';
import 'react-dom';
import '@jetbrains/icons/close';
import '../island/island.js';
import '../global/data-tests.js';
import '../island/adaptive-island-hoc.js';
import '../global/linear-function.js';
import 'react-compiler-runtime';
import 'element-resize-detector';
import '../global/schedule-raf.js';
import '../global/get-uid.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/dom.js';
import '../popup/popup.target.js';
import '../popup/popup.js';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../_helpers/dialog-body-scroll-preventer.js';
import 'scrollbar-width';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../_helpers/button.classes.js';

class Confirm extends PureComponent {
  static defaultProps = {
    text: null,
    description: null,
    show: false,
    rejectOnEsc: true,
    inProgress: false,
    cancelIsDefault: false,
    confirmLabel: 'OK',
    rejectLabel: 'Cancel',
    onConfirm: () => {},
    onReject: () => {}
  };
  onEscPress = () => {
    if (this.props.rejectOnEsc) {
      this.props.onReject();
    }
  };
  render() {
    const {
      show,
      className,
      inProgress,
      cancelIsDefault,
      text,
      description,
      confirmLabel,
      rejectLabel,
      onConfirm,
      onReject,
      native
    } = this.props;
    return /*#__PURE__*/jsxs(Dialog, {
      label: text || (typeof description === 'string' ? description : undefined),
      className: className,
      onEscPress: this.onEscPress,
      show: show,
      trapFocus: true,
      "data-test": "ring-confirm",
      native: native,
      children: [text && /*#__PURE__*/jsx(HeaderWrapper, {
        children: text
      }), description && /*#__PURE__*/jsx(ContentWrapper, {
        children: description
      }), /*#__PURE__*/jsxs(Panel, {
        children: [/*#__PURE__*/jsx(Button, {
          "data-test": "confirm-ok-button",
          primary: !cancelIsDefault,
          loader: inProgress,
          disabled: inProgress,
          onClick: onConfirm,
          children: confirmLabel
        }), /*#__PURE__*/jsx(Button, {
          "data-test": "confirm-reject-button",
          onClick: onReject,
          disabled: inProgress,
          primary: cancelIsDefault,
          children: rejectLabel
        })]
      })]
    });
  }
}

export { Confirm as default };
