// @jetbrains/ring-ui-built/components/button/button.js
import { PureComponent, createRef } from 'react';
import classNames from 'classnames';
import chevronDownIcon from '@jetbrains/icons/chevron-down';
import chevron12pxDown from '@jetbrains/icons/chevron-12px-down';
import deprecate from 'util-deprecate';
import Icon from '../icon/icon.js';
import ClickableLink from '../link/clickable-link.js';
import { ControlsHeightContext } from '../global/controls-height.js';
import { g as getButtonClasses, s as styles } from '../_helpers/button.classes.js';
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { Size } from '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../global/configuration.js';

const warnText = deprecate(() => {}, 'Button: "text" prop is deprecated and will be removed in 8.0. Use inline instead.');
function removeLinkProps(props) {
  const {
    download,
    href,
    hrefLang,
    media,
    ping,
    target,
    type,
    referrerPolicy,
    onConditionalClick,
    onPlainLeftClick,
    activeClassName,
    ...restProps
  } = props;
  return restProps;
}
/**
 * @name Button
 */
/**
 * A component for displaying variously styled buttons.
 */
class Button extends PureComponent {
  /**
   * @deprecated Use icons with appropriate intrinsic sizes instead
   */
  static IconSize = Size;
  static contextType = ControlsHeightContext;
  buttonRef = /*#__PURE__*/createRef();
  render() {
    var _ref;
    const {
      // Modifiers
      active,
      danger,
      delayed,
      loader,
      primary,
      success,
      error,
      secondary,
      ghost,
      short,
      text,
      dropdown,
      height,
      // Props
      icon,
      iconRight,
      iconSize,
      iconClassName,
      iconRightClassName,
      iconSuppressSizeWarning,
      className,
      children,
      inline,
      disabled,
      ...props
    } = this.props;
    const isInline = (_ref = inline !== null && inline !== void 0 ? inline : text) !== null && _ref !== void 0 ? _ref : !!icon;
    if (text) {
      warnText();
    }
    const classes = getButtonClasses({
      ...this.props,
      inline: isInline,
      height: height !== null && height !== void 0 ? height : typeof this.context === 'function' ? this.context() : this.context
    });
    const content = /*#__PURE__*/jsxs(Fragment, {
      children: [icon && /*#__PURE__*/jsx(Icon, {
        className: classNames(styles.icon, iconClassName),
        glyph: icon,
        size: iconSize,
        suppressSizeWarning: iconSuppressSizeWarning
      }), children, iconRight && /*#__PURE__*/jsx(Icon, {
        className: classNames(styles.iconRight, iconRightClassName),
        glyph: iconRight
      }), dropdown && /*#__PURE__*/jsx(Icon, {
        glyph: isInline ? chevron12pxDown : chevronDownIcon,
        className: styles.dropdownIcon
      })]
    });
    const isDisabled = disabled || loader || undefined;
    const commonProps = {
      ...(props.href !== undefined && isDisabled ? removeLinkProps(props) : props),
      className: classes,
      children: /*#__PURE__*/jsxs(Fragment, {
        children: [loader && !isInline && /*#__PURE__*/jsx("div", {
          className: styles.loaderBackground
        }), content]
      })
    };
    return 'href' in commonProps && commonProps.href !== undefined ? /*#__PURE__*/jsx(ClickableLink, {
      ...commonProps
    }) : /*#__PURE__*/jsx("button", {
      ref: this.buttonRef,
      type: "button",
      disabled: isDisabled,
      ...commonProps
    });
  }
}

export { Button, Size as IconSize, Button as default };
