// @jetbrains/ring-ui-built/components/header/header.js
import { Component } from 'react';
import classNames from 'classnames';
import { T as Theme, a as ThemeProvider } from '../_helpers/theme.js';
import { s as styles } from '../_helpers/header.js';
import { jsx } from 'react/jsx-runtime';
export { default as Logo } from './logo.js';
export { default as Tray } from './tray.js';
export { default as HeaderIcon } from './header-icon.js';
export { default as Profile } from './profile.js';
export { default as SmartProfile } from './smart-profile.js';
export { default as Services } from './services.js';
export { default as SmartServices } from './smart-services.js';
import 'react-compiler-runtime';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../avatar/avatar.js';
import '../global/url.js';
import '../global/dom.js';
import '../avatar/fallback-avatar.js';
import '../global/get-uid.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../dropdown-menu/dropdown-menu.js';
import '../list/list.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'memoize-one';
import '../global/data-tests.js';
import '../global/schedule-raf.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../_helpers/link.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../dropdown/dropdown.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import '../popup-menu/popup-menu.js';
import '../popup/popup.js';
import 'react-dom';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../i18n/i18n-context.js';
import '../i18n/i18n.js';
import '../alert-service/alert-service.js';
import 'react-dom/client';
import '../alert/alert.js';
import '@jetbrains/icons/exception';
import '@jetbrains/icons/checkmark';
import '@jetbrains/icons/warning';
import '@jetbrains/icons/close';
import '../loader-inline/loader-inline.js';
import '../alert/container.js';
import '../auth/auth.js';
import '../auth/window-flow.js';
import '../auth/response-parser.js';
import '../auth/down-notification.js';
import '../group/group.js';
import '../auth/auth-core.js';
import '../global/listeners.js';
import '../http/http.js';
import '../global/promise-with-timeout.js';
import '../auth/storage.js';
import '../storage/storage-local.js';
import '../auth/request-builder.js';
import 'simply-uuid';
import '../auth/background-flow.js';
import '../auth/token-validator.js';
import '@jetbrains/icons/services-20px';
import '../_helpers/services-link.js';

class Header extends Component {
  static defaultProps = {
    spaced: true,
    theme: Theme.DARK,
    vertical: false
  };
  render() {
    const {
      children,
      className,
      spaced,
      theme,
      vertical,
      ...restProps
    } = this.props;
    const classes = classNames(vertical ? styles.headerVertical : styles.header, className, {
      [styles.headerSpaced]: spaced && !vertical
    });
    const overrideOuterTheme = theme !== Theme.LIGHT;
    const header = /*#__PURE__*/jsx("header", {
      ...restProps,
      className: classes,
      children: children
    });
    if (overrideOuterTheme) {
      return /*#__PURE__*/jsx(ThemeProvider, {
        theme: theme,
        children: header
      });
    }
    return header;
  }
}

export { Header as default };
