// @jetbrains/ring-ui-built/components/header/profile.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import Avatar from '../avatar/avatar.js';
import { Button } from '../button/button.js';
import DropdownMenu from '../dropdown-menu/dropdown-menu.js';
import PopupMenu from '../popup-menu/popup-menu.js';
import { I18nContext } from '../i18n/i18n-context.js';
import { isTruthy } from '../global/typescript-utils.js';
import { s as styles } from '../_helpers/header.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { Size } from '../avatar/avatar-size.js';
import 'util-deprecate';
import '../global/url.js';
import '../global/dom.js';
import '../global/memoize.js';
import '../avatar/fallback-avatar.js';
import 'react-compiler-runtime';
import '../global/get-uid.js';
import '../_helpers/avatar-info.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../list/list.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'memoize-one';
import '../global/data-tests.js';
import '../global/schedule-raf.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../_helpers/link.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../dropdown/dropdown.js';
import '../_helpers/anchor.js';
import '../popup/popup.js';
import 'react-dom';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import '../i18n/i18n.js';

const rgItemType = PopupMenu.ListProps.Type.LINK;
class Profile extends PureComponent {
  static defaultProps = {
    closeOnSelect: true,
    renderPopupItems: items => items,
    size: Size.Size32,
    renderGuest: ({
      loading,
      onLogin,
      className,
      translations
    }) => /*#__PURE__*/jsx(I18nContext.Consumer, {
      children: ({
        translate
      }) => {
        var _translations$login;
        return /*#__PURE__*/jsx("div", {
          className: classNames(styles.profileEmpty, className),
          children: /*#__PURE__*/jsx(Button, {
            primary: true,
            "data-test": "ring-header-login-button",
            disabled: loading,
            loader: loading,
            onClick: onLogin,
            children: (_translations$login = translations?.login) !== null && _translations$login !== void 0 ? _translations$login : translate('login')
          })
        });
      }
    })
  };
  static contextType = I18nContext;
  static Size = Size;
  render() {
    var _translations$applyCh, _translations$login2, _translations$profile, _translations$switchU, _translations$logout;
    const {
      className,
      activeClassName,
      closeOnSelect,
      hasUpdates,
      onLogout,
      user,
      profileUrl,
      LinkComponent,
      onSwitchUser,
      renderPopupItems,
      onRevertPostponement,
      showApplyChangedUser,
      showName,
      showLogIn,
      showLogOut,
      showSwitchUser,
      renderGuest,
      translations,
      size,
      round,
      loading,
      onLogin,
      menuProps,
      onShow,
      onHide,
      ...props
    } = this.props;
    const {
      translate
    } = this.context;
    if (!user) {
      return /*#__PURE__*/jsx("div", {
        ...props,
        className: classNames(styles.profileEmpty, className),
        children: /*#__PURE__*/jsx(Avatar, {
          size: size,
          round: round
        })
      });
    }
    if (user.guest) {
      return renderGuest(this.props);
    }
    const avatarWrapper = classNames(styles.avatarWrapper, {
      [styles.hasUpdates]: hasUpdates
    });
    const anchor = /*#__PURE__*/jsxs("button", {
      type: "button",
      className: styles.anchorClassName,
      children: [/*#__PURE__*/jsx("span", {
        className: avatarWrapper,
        children: /*#__PURE__*/jsx(Avatar, {
          url: user.profile && user.profile.avatar && user.profile.avatar.url,
          size: size,
          round: round,
          username: user.name
        })
      }), showName && /*#__PURE__*/jsx("span", {
        children: user.name
      })]
    });
    const items = [showApplyChangedUser && {
      rgItemType,
      label: (_translations$applyCh = translations?.applyChangedUser) !== null && _translations$applyCh !== void 0 ? _translations$applyCh : translate('applyChangedUser'),
      className: styles.profileMenuItem,
      onClick: onRevertPostponement
    }, showLogIn && {
      rgItemType,
      label: (_translations$login2 = translations?.login) !== null && _translations$login2 !== void 0 ? _translations$login2 : translate('login'),
      className: styles.profileMenuItem,
      onClick: onRevertPostponement
    }, {
      rgItemType: PopupMenu.ListProps.Type.LINK,
      label: (_translations$profile = translations?.profile) !== null && _translations$profile !== void 0 ? _translations$profile : translate('profile'),
      target: '_self',
      // Full page reload in Angular
      href: profileUrl,
      LinkComponent
    }, showSwitchUser && {
      rgItemType,
      label: (_translations$switchU = translations?.switchUser) !== null && _translations$switchU !== void 0 ? _translations$switchU : translate('switchUser'),
      className: styles.profileMenuItem,
      onClick: onSwitchUser
    }, showLogOut && {
      rgItemType,
      label: (_translations$logout = translations?.logout) !== null && _translations$logout !== void 0 ? _translations$logout : translate('logout'),
      onClick: onLogout
    }].filter(isTruthy);
    return /*#__PURE__*/jsx(DropdownMenu, {
      ...props,
      onShow: onShow,
      onHide: onHide,
      title: user.name,
      anchor: anchor,
      data: renderPopupItems(items),
      "data-test": "ring-profile",
      className: classNames(styles.profile, className),
      activeClassName: activeClassName,
      menuProps: {
        closeOnSelect,
        left: -2,
        top: -8,
        sidePadding: 32,
        ...menuProps
      }
    });
  }
}

export { Profile as default };
