// @jetbrains/ring-ui-built/components/header/services.js
import { PureComponent } from 'react';
import servicesIcon from '@jetbrains/icons/services-20px';
import classNames from 'classnames';
import Dropdown from '../dropdown/dropdown.js';
import Popup from '../popup/popup.js';
import { T as Theme, d as darkStyles } from '../_helpers/theme.js';
import HeaderIcon from './header-icon.js';
import { S as ServicesLink, s as styles } from '../_helpers/services-link.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import '../global/data-tests.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import 'react-compiler-runtime';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/header.js';
import '../link/link.js';
import '../_helpers/link.js';

const makeAnchor = loading => {
  const anchor = ({
    active
  }) => /*#__PURE__*/jsx(HeaderIcon, {
    loader: loading,
    active: active,
    icon: servicesIcon,
    "aria-label": "Services"
  });
  return anchor;
};
class Services extends PureComponent {
  static sort = (a, b) => {
    const aApplicationName = a.applicationName || '';
    const bApplicationName = b.applicationName || '';
    return aApplicationName.localeCompare(bApplicationName) || a.name.localeCompare(b.name);
  };
  static defaultProps = {
    theme: Theme.DARK
  };
  static Link = ServicesLink;
  serviceIsActive = service => service.id === this.props.clientId;
  render() {
    const {
      clientId,
      loading,
      services,
      initShown,
      theme,
      ...props
    } = this.props;
    if (!services) {
      return /*#__PURE__*/jsx(HeaderIcon, {
        ...props,
        loader: loading,
        active: loading,
        icon: servicesIcon,
        "aria-label": "Services"
      });
    }
    const sortedServices = [...services].sort(Services.sort);
    const servicesWithIcons = sortedServices.filter(service => service.iconUrl && service.homeUrl);
    const servicesWithOutIcons = sortedServices.filter(service => !service.iconUrl && service.homeUrl);
    const separatorIsRequired = servicesWithIcons.length !== 0 && servicesWithOutIcons.length !== 0;
    return /*#__PURE__*/jsx(Dropdown, {
      ...props,
      anchor: makeAnchor(loading),
      initShown: initShown,
      children: /*#__PURE__*/jsxs(Popup, {
        className: classNames(styles.services, {
          [darkStyles.dark]: theme === Theme.DARK
        }),
        top: -3,
        children: [servicesWithIcons.map(service => {
          const isActive = this.serviceIsActive(service);
          return /*#__PURE__*/jsx(Services.Link, {
            active: isActive,
            className: isActive ? styles.activeItem : styles.item,
            service: service
          }, service.id);
        }), separatorIsRequired && /*#__PURE__*/jsx("div", {
          className: styles.line
        }, 'separator'), servicesWithOutIcons.map(service => {
          const isActive = this.serviceIsActive(service);
          return /*#__PURE__*/jsx(Services.Link, {
            active: isActive,
            className: isActive ? styles.activeItemStacked : styles.itemStacked,
            service: service
          }, service.id);
        })]
      })
    });
  }
}

export { Services as default };
