// @jetbrains/ring-ui-built/components/header/smart-profile.js
import { PureComponent } from 'react';
import alertService from '../alert-service/alert-service.js';
import Profile from './profile.js';
import '../auth/auth.js';
import { jsx } from 'react/jsx-runtime';
import { USER_CHANGED_EVENT, LOGOUT_POSTPONED_EVENT, USER_CHANGE_POSTPONED_EVENT } from '../auth/auth-core.js';
import 'react-dom/client';
import '../global/get-uid.js';
import '../alert/alert.js';
import 'classnames';
import '@jetbrains/icons/exception';
import '@jetbrains/icons/checkmark';
import '@jetbrains/icons/warning';
import '@jetbrains/icons/close';
import '../icon/icon.js';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import 'react-compiler-runtime';
import '../global/memoize.js';
import '../loader-inline/loader-inline.js';
import '../global/data-tests.js';
import '../global/dom.js';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../_helpers/theme.js';
import '../alert/container.js';
import 'react-dom';
import '../avatar/avatar.js';
import '../global/url.js';
import '../avatar/fallback-avatar.js';
import '../avatar/avatar-size.js';
import '../_helpers/avatar-info.js';
import '../dropdown-menu/dropdown-menu.js';
import '../list/list.js';
import 'react-virtualized/dist/es/List';
import 'react-virtualized/dist/es/AutoSizer';
import 'react-virtualized/dist/es/WindowScroller';
import 'react-virtualized/dist/es/CellMeasurer';
import 'memoize-one';
import '../global/schedule-raf.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../global/create-stateful-context.js';
import '../list/list-item.js';
import '../checkbox/checkbox.js';
import '@jetbrains/icons/checkmark-12px';
import '@jetbrains/icons/remove-12px';
import '../global/compose-refs.js';
import '../control-help/control-help.js';
import '../link/link.js';
import '../_helpers/link.js';
import '../list/consts.js';
import '../list/list.classes.js';
import '../_helpers/list.js';
import '../list/list-custom.js';
import '../global/get-event-key.js';
import '../list/list-title.js';
import '../list/list-separator.js';
import '../list/list-hint.js';
import '../dropdown/dropdown.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import '../popup-menu/popup-menu.js';
import '../popup/popup.js';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../i18n/i18n-context.js';
import '../i18n/i18n.js';
import '../_helpers/header.js';
import '../auth/window-flow.js';
import '../auth/response-parser.js';
import '../auth/down-notification.js';
import '../group/group.js';
import '../global/listeners.js';
import '../http/http.js';
import '../global/promise-with-timeout.js';
import '../auth/storage.js';
import '../storage/storage-local.js';
import '../auth/request-builder.js';
import 'simply-uuid';
import '../auth/background-flow.js';
import '../auth/token-validator.js';

const CERTIFICATE_MISMATCH_HEADER = 'x-client-certificate-token-mismatch';
class SmartProfile extends PureComponent {
  state = {
    user: null,
    size: Profile.defaultProps.size,
    isLogoutPostponed: false,
    isUserChangePostponed: false
  };
  componentDidMount() {
    this.requestUser();
  }
  static Size = Profile.Size;
  login = async () => {
    this.setState({
      loading: true
    });
    try {
      await this.props.auth.login();
    } catch (err) {
      // eslint-disable-next-line no-console
      console.debug('Profile login errored', err);
    } finally {
      this.setState({
        loading: false
      });
    }
  };
  logout = () => this.props.auth.logout();
  switchUser = () => this.props.auth.switchUser();
  onRevertPostponement = () => {
    if (this.state.isLogoutPostponed) {
      this.props.auth.login();
    }
    if (this.state.isUserChangePostponed) {
      this.props.auth.updateUser();
    }
  };
  async requestUser() {
    try {
      const {
        auth
      } = this.props;
      const user = await auth.requestUser();
      this.checkUserCertificateMismatch(user);
      this.setState({
        user
      });
      auth.addListener(USER_CHANGED_EVENT, newUser => {
        this.setState({
          user: newUser,
          isLogoutPostponed: false,
          isUserChangePostponed: false
        });
      });
      auth.addListener(LOGOUT_POSTPONED_EVENT, () => {
        this.setState({
          isLogoutPostponed: true
        });
      });
      auth.addListener(USER_CHANGE_POSTPONED_EVENT, () => {
        this.setState({
          isUserChangePostponed: true
        });
      });
    } catch (e) {
      // noop
    }
  }
  checkUserCertificateMismatch(user) {
    const {
      auth,
      translations
    } = this.props;
    const userMeta = auth.http.getMetaForResponse(user);
    if (userMeta?.headers?.has(CERTIFICATE_MISMATCH_HEADER)) {
      const message = translations?.certificateMismatch || `You are authenticated as ${user.login || user.name}. To authenticate with the client certificate for your account, log out, then click the "Log in with certificate" option on the login page.`;
      alertService.warning(message, 0);
    }
  }
  render() {
    const {
      user,
      loading,
      isLogoutPostponed,
      isUserChangePostponed
    } = this.state;
    const {
      auth,
      profileUrl,
      ...props
    } = this.props;
    const url = profileUrl || (user ? `${auth.config.serverUri}users/${user.id}` : '');
    return /*#__PURE__*/jsx(Profile, {
      onLogin: this.login,
      onLogout: this.logout,
      onSwitchUser: this.switchUser,
      loading: loading,
      user: user,
      profileUrl: url,
      showApplyChangedUser: isUserChangePostponed,
      showLogIn: isLogoutPostponed,
      showLogOut: !isLogoutPostponed,
      showSwitchUser: auth._canShowDialogs() && !isLogoutPostponed && !isUserChangePostponed,
      onRevertPostponement: this.onRevertPostponement,
      ...props
    });
  }
}

export { SmartProfile as default };
