// @jetbrains/ring-ui-built/components/header/smart-services.js
import { Component } from 'react';
import HTTP from '../http/http.js';
import Services from './services.js';
import { jsx } from 'react/jsx-runtime';
import '../global/url.js';
import '@jetbrains/icons/services-20px';
import 'classnames';
import '../dropdown/dropdown.js';
import '../global/data-tests.js';
import '../global/typescript-utils.js';
import '../_helpers/anchor.js';
import 'react-compiler-runtime';
import '../button/button.js';
import '@jetbrains/icons/chevron-down';
import '@jetbrains/icons/chevron-12px-down';
import 'util-deprecate';
import '../icon/icon.js';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../link/clickable-link.js';
import '../global/controls-height.js';
import '../global/configuration.js';
import '../_helpers/button.classes.js';
import '../popup/popup.js';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';
import './header-icon.js';
import '../_helpers/header.js';
import '../_helpers/services-link.js';
import '../link/link.js';
import '../_helpers/link.js';

function noop() {}
class SmartServices extends Component {
  state = {
    visible: true,
    loading: false,
    services: null
  };
  componentDidMount() {
    const {
      auth
    } = this.props;
    this.http = new HTTP(auth, auth.getAPIPath());
    this.getServices(SmartServices.countFields)?.then(services => {
      if (!services.length) {
        this.setState({
          visible: false
        });
      }
    }).catch(noop);
  }
  static allFields = 'id,name,applicationName,homeUrl,iconUrl';
  static countFields = 'key';
  http;
  stopLoading = () => {
    this.setState({
      loading: false
    });
  };
  getServicesContent = () => {
    this.setState({
      loading: true
    });
    this.getServices(SmartServices.allFields)?.then(services => {
      this.setState({
        services
      });
      this.stopLoading();
    }).catch(this.stopLoading);
  };
  getServices(fields) {
    return this.http?.get(`services/header?fields=${fields}`);
  }
  render() {
    const {
      services,
      visible,
      loading
    } = this.state;
    const {
      auth,
      ...props
    } = this.props;
    if (!visible) {
      return null;
    }
    return /*#__PURE__*/jsx(Services, {
      ...props,
      clientId: auth.config.clientId,
      initShown: true,
      loading: loading,
      onClick: this.getServicesContent,
      services: services
    });
  }
}

export { SmartServices as default };
