// @jetbrains/ring-ui-built/components/link/link.js
import { PureComponent } from 'react';
import classNames from 'classnames';
import joinDataTestAttributes from '../global/data-tests.js';
import ClickableLink from './clickable-link.js';
import { l as linkStyles } from '../_helpers/link.js';
import { jsx } from 'react/jsx-runtime';

function linkHOC(ComposedComponent) {
  const isCustom = typeof ComposedComponent !== 'string' && ComposedComponent !== ClickableLink;
  return class Link extends PureComponent {
    render() {
      const {
        active,
        inherit,
        pseudo,
        hover,
        className,
        'data-test': dataTest,
        href,
        children,
        onPlainLeftClick,
        onClick,
        ...restProps
      } = this.props;
      const useButton = pseudo || !isCustom && (href === null || href === undefined);
      const classes = classNames(linkStyles.link, className, {
        [linkStyles.active]: active,
        [linkStyles.inherit]: inherit,
        [linkStyles.hover]: hover,
        [linkStyles.pseudo]: useButton
      });
      let props = restProps;
      if (isCustom && !props.activeClassName) {
        props = {
          ...props,
          activeClassName: linkStyles.active
        };
      }
      if (useButton) {
        return /*#__PURE__*/jsx("button", {
          type: "button",
          ...props,
          className: classes,
          onClick: onClick || onPlainLeftClick,
          "data-test": joinDataTestAttributes('ring-link', dataTest),
          children: children
        });
      }
      return /*#__PURE__*/jsx(ComposedComponent, {
        ...props,
        href: href,
        className: classes,
        onClick: onClick,
        ...(typeof ComposedComponent !== 'string' ? {
          onPlainLeftClick
        } : {}),
        "data-test": joinDataTestAttributes('ring-link', dataTest),
        children: children
      });
    }
  };
}
var Link = linkHOC(ClickableLink);

export { Link as default, linkHOC };
