// @jetbrains/ring-ui-built/components/banner/banner.js
import { c } from 'react-compiler-runtime';
import React__default from 'react';
import classNames from 'classnames';
import infoIcon from '@jetbrains/icons/info-filled';
import exceptionIcon from '@jetbrains/icons/exception';
import successIcon from '@jetbrains/icons/success';
import warningIcon from '@jetbrains/icons/warning';
import closeIcon from '@jetbrains/icons/close';
import Icon from '../icon/icon.js';
import Link from '../link/link.js';
import Tooltip from '../tooltip/tooltip.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import 'util-deprecate';
import '../icon/icon.constants.js';
import '../_helpers/icon-svg.js';
import '../global/memoize.js';
import '../global/data-tests.js';
import '../link/clickable-link.js';
import '../_helpers/link.js';
import '../popup/popup.js';
import 'react-dom';
import '../global/get-uid.js';
import '../global/schedule-raf.js';
import '../global/dom.js';
import '../shortcuts/shortcuts.js';
import '../shortcuts/core.js';
import 'combokeys';
import '../global/sniffer.js';
import 'sniffr';
import '../tab-trap/tab-trap.js';
import '../global/configuration.js';
import '../popup/position.js';
import '../popup/popup.consts.js';
import '../popup/popup.target.js';
import '../popup/position-css.js';
import '../_helpers/theme.js';

var styles = {"banner":"ring-banner-banner","inline":"ring-banner-inline","info":"ring-banner-info","icon":"ring-banner-icon","error":"ring-banner-error","success":"ring-banner-success","warning":"ring-banner-warning","purple":"ring-banner-purple","grey":"ring-banner-grey","title":"ring-banner-title","description":"ring-banner-description","closeButtonWrapper":"ring-banner-closeButtonWrapper","closeButton":"ring-banner-closeButton"};

function Banner(props) {
  const $ = c(32);
  if ($[0] !== "df3c8f8e5b4a6ff238ad913306a06eab46faf0049973cb0f10cf90f321812a0d") {
    for (let $i = 0; $i < 32; $i += 1) {
      $[$i] = Symbol.for("react.memo_cache_sentinel");
    }
    $[0] = "df3c8f8e5b4a6ff238ad913306a06eab46faf0049973cb0f10cf90f321812a0d";
  }
  const {
    title,
    children,
    mode: t0,
    withIcon: t1,
    inline: t2,
    onClose,
    closeButtonTitle,
    className,
    translations: t3
  } = props;
  const mode = t0 === undefined ? "info" : t0;
  const withIcon = t1 === undefined ? false : t1;
  const inline = t2 === undefined ? false : t2;
  let t4;
  if ($[1] !== t3) {
    t4 = t3 === undefined ? {
      close: "Close"
    } : t3;
    $[1] = t3;
    $[2] = t4;
  } else {
    t4 = $[2];
  }
  const translations = t4;
  const dataTest = props["data-test"];
  let icon = infoIcon;
  if (mode === "error") {
    icon = exceptionIcon;
  } else {
    if (mode === "success") {
      icon = successIcon;
    } else {
      if (mode === "warning") {
        icon = warningIcon;
      }
    }
  }
  const t5 = mode === "info";
  const t6 = mode === "error";
  const t7 = mode === "success";
  const t8 = mode === "warning";
  const t9 = mode === "purple";
  const t10 = mode === "grey";
  let t11;
  if ($[3] !== className || $[4] !== inline || $[5] !== t10 || $[6] !== t5 || $[7] !== t6 || $[8] !== t7 || $[9] !== t8 || $[10] !== t9) {
    t11 = classNames({
      [styles.banner]: true,
      [styles.inline]: inline,
      [styles.info]: t5,
      [styles.error]: t6,
      [styles.success]: t7,
      [styles.warning]: t8,
      [styles.purple]: t9,
      [styles.grey]: t10
    }, className);
    $[3] = className;
    $[4] = inline;
    $[5] = t10;
    $[6] = t5;
    $[7] = t6;
    $[8] = t7;
    $[9] = t8;
    $[10] = t9;
    $[11] = t11;
  } else {
    t11 = $[11];
  }
  let t12;
  if ($[12] !== icon || $[13] !== withIcon) {
    t12 = withIcon && /*#__PURE__*/jsx(Icon, {
      glyph: icon,
      className: styles.icon
    });
    $[12] = icon;
    $[13] = withIcon;
    $[14] = t12;
  } else {
    t12 = $[14];
  }
  let t13;
  if ($[15] !== title) {
    t13 = title && /*#__PURE__*/jsx("div", {
      className: styles.title,
      children: title
    });
    $[15] = title;
    $[16] = t13;
  } else {
    t13 = $[16];
  }
  let t14;
  if ($[17] !== children) {
    t14 = /*#__PURE__*/jsx("div", {
      className: styles.description,
      children: children
    });
    $[17] = children;
    $[18] = t14;
  } else {
    t14 = $[18];
  }
  let t15;
  if ($[19] !== t13 || $[20] !== t14) {
    t15 = /*#__PURE__*/jsxs("div", {
      children: [t13, t14]
    });
    $[19] = t13;
    $[20] = t14;
    $[21] = t15;
  } else {
    t15 = $[21];
  }
  let t16;
  if ($[22] !== closeButtonTitle || $[23] !== onClose || $[24] !== translations) {
    t16 = onClose && /*#__PURE__*/jsx(Tooltip, {
      title: closeButtonTitle !== null && closeButtonTitle !== void 0 ? closeButtonTitle : translations.close,
      className: styles.closeButtonWrapper,
      children: /*#__PURE__*/jsx(Link, {
        className: styles.closeButton,
        pseudo: true,
        onClick: onClose,
        "aria-label": "Close",
        children: /*#__PURE__*/jsx(Icon, {
          glyph: closeIcon
        })
      })
    });
    $[22] = closeButtonTitle;
    $[23] = onClose;
    $[24] = translations;
    $[25] = t16;
  } else {
    t16 = $[25];
  }
  let t17;
  if ($[26] !== dataTest || $[27] !== t11 || $[28] !== t12 || $[29] !== t15 || $[30] !== t16) {
    t17 = /*#__PURE__*/jsxs("div", {
      className: t11,
      "data-test": dataTest,
      children: [t12, t15, t16]
    });
    $[26] = dataTest;
    $[27] = t11;
    $[28] = t12;
    $[29] = t15;
    $[30] = t16;
    $[31] = t17;
  } else {
    t17 = $[31];
  }
  return t17;
}
var banner = /*#__PURE__*/React__default.memo(Banner);

export { Banner, banner as default };
